{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Define Event\n",
    "\n",
    "Before we can analyze the event, we need to identify the boundaries of the event. We will accomplish this by\n",
    "\n",
    "1. Defining the values to be classified as gains and losses\n",
    "2. Creating a counts table defining the various events\n",
    "3. Defining the porportion of patients with gain or loss to be considered significant\n",
    "4. Identifying regions of gain and loss\n",
    "5. Identifying regions where all cancer types meet criteria for gain or loss event."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cptac warning: Your version of cptac (0.8.5) is out-of-date. Latest is 0.8.6. Please run 'pip install --upgrade cptac' to update it. (C:\\Users\\chels\\anaconda3\\envs\\notebooks\\lib\\site-packages\\ipykernel_launcher.py, line 4)\n"
     ]
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import cptac\n",
    "import numpy as np\n",
    "import cnvutils"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 1: Define Parameters\n",
    "\n",
    "These are the parameters that must be set for each analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The porportion of patients with gain or loss of a given gene to be considered\n",
    "PATIENT_CUTOFF = 0.15\n",
    "# The chromosome to be analyzed (should be a string)\n",
    "CHROMOSOME = '5'\n",
    "# The arm to be analyzed. Options are: 'p', 'q' or 'both'.\n",
    "ARM = 'both'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking that lscc index is up-to-date... \r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cptac warning: The HNSCC data is currently strictly reserved for CPTAC investigators. Otherwise, you are not authorized to access these data. Additionally, even after these data become publicly available, they will be subject to a publication embargo (see https://proteomics.cancer.gov/data-portal/about/data-use-agreement or enter cptac.embargo() to open the webpage for more details). (C:\\Users\\chels\\anaconda3\\envs\\notebooks\\lib\\site-packages\\ipykernel_launcher.py, line 9)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking that luad index is up-to-date...\r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cptac warning: The LSCC data is currently strictly reserved for CPTAC investigators. Otherwise, you are not authorized to access these data. Additionally, even after these data become publicly available, they will be subject to a publication embargo (see https://proteomics.cancer.gov/data-portal/about/data-use-agreement or enter cptac.embargo() to open the webpage for more details). (C:\\Users\\chels\\anaconda3\\envs\\notebooks\\lib\\site-packages\\ipykernel_launcher.py, line 10)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                            \r"
     ]
    }
   ],
   "source": [
    "# Here you will need to load in all the cancer types to be considered. We have listed all for convenience, simply comment\n",
    "# out the cancer types you do not want to consider.\n",
    "cnv = {\n",
    "    'BRCA': cptac.Brca().get_CNV(),\n",
    "#     'CCRCC': cptac.Ccrcc().get_CNV(),\n",
    "#     'COLON': cptac.Colon().get_CNV(),\n",
    "#     'ENDO': cptac.Endometrial().get_CNV(),\n",
    "#     'GBM': cptac.Gbm().get_CNV(),\n",
    "    'HNSCC': cptac.Hnscc().get_CNV(),\n",
    "    'LSCC': cptac.Lscc().get_CNV(),\n",
    "    'LUAD': cptac.Luad().get_CNV(),\n",
    "    'OVARIAN': cptac.Ovarian().get_CNV()\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 2: Determine Cutoffs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\chels\\anaconda3\\envs\\notebooks\\lib\\site-packages\\IPython\\core\\interactiveshell.py:3338: DtypeWarning: Columns (2) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  if (await self.run_code(code, result,  async_=asy)):\n"
     ]
    }
   ],
   "source": [
    "counts = cnvutils.get_counts_table()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "cutoffs = dict()\n",
    "for cancer_type in cnv.keys():\n",
    "    df = cnv[cancer_type]\n",
    "    cutoffs[cancer_type] = len(df) * PATIENT_CUTOFF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 3: Subset Event"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = counts[(counts.chromosome == CHROMOSOME)]\n",
    "if ARM != 'both':\n",
    "    counts = counts[counts.arm == ARM]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find Loss Regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "976\n",
      "2853\n",
      "1003\n",
      "818\n",
      "759\n"
     ]
    }
   ],
   "source": [
    "df = counts\n",
    "loss_event_locations = dict()\n",
    "for cancer in cnv.keys():\n",
    "    df_loss = df[(df.variable == 'loss') & (df.cancer == cancer)].sort_values('start_bp')\n",
    "    print(len(df_loss))\n",
    "    values = list(df_loss.value)\n",
    "    loss_events = list()\n",
    "    start = None\n",
    "    for i in range(0, len(values)):\n",
    "        val = values[i]\n",
    "        if val > cutoffs[cancer]:\n",
    "            if start is None:\n",
    "                start = i\n",
    "        else:\n",
    "            if start is not None:\n",
    "                loss_events.append((start, i))\n",
    "                start = None\n",
    "    if start is not None:\n",
    "        loss_events.append((start, len(values)-1))\n",
    "    event_locations = list()\n",
    "    for event in loss_events:\n",
    "        start_bp = df_loss.iloc[event[0]].start_bp\n",
    "        end_bp = df_loss.iloc[event[1]].start_bp\n",
    "        event_locations.append((start_bp, end_bp-start_bp))\n",
    "    loss_event_locations[cancer] = event_locations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "loss_event_patients = list()\n",
    "for cancer in loss_event_locations.keys():\n",
    "    events = loss_event_locations[cancer]\n",
    "    for event in events:\n",
    "        start = event[0]\n",
    "        end = event[0] + event[1]\n",
    "        loss_event_patients.append((start, 1))\n",
    "        loss_event_patients.append((end, 0))\n",
    "#     patients += list(gain_event_locations[cancer])\n",
    "loss_event_patients.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "count = 0\n",
    "current_bp = 0\n",
    "start = list()\n",
    "end = list()\n",
    "size = list()\n",
    "total = list()\n",
    "for patient in loss_event_patients:\n",
    "    if patient[0] != current_bp:\n",
    "        start.append(current_bp)\n",
    "        end.append(patient[0])\n",
    "        size.append(patient[0]-current_bp)\n",
    "        total.append(count)\n",
    "        current_bp = patient[0]\n",
    "    if patient[1] == 1:\n",
    "        count += 1\n",
    "    else:\n",
    "        count -= 1\n",
    "loss_event_data = pd.DataFrame({'start': start, 'end': end, 'counts': total, 'length': size})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>counts</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>38474668.0</td>\n",
       "      <td>0</td>\n",
       "      <td>38474668.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>38474668.0</td>\n",
       "      <td>38937920.0</td>\n",
       "      <td>1</td>\n",
       "      <td>463252.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>38937920.0</td>\n",
       "      <td>45254948.0</td>\n",
       "      <td>0</td>\n",
       "      <td>6317028.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>45254948.0</td>\n",
       "      <td>50396192.0</td>\n",
       "      <td>1</td>\n",
       "      <td>5141244.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>50396192.0</td>\n",
       "      <td>51383448.0</td>\n",
       "      <td>3</td>\n",
       "      <td>987256.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>51383448.0</td>\n",
       "      <td>86617928.0</td>\n",
       "      <td>4</td>\n",
       "      <td>35234480.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>86617928.0</td>\n",
       "      <td>118836074.0</td>\n",
       "      <td>5</td>\n",
       "      <td>32218146.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>118836074.0</td>\n",
       "      <td>148169733.0</td>\n",
       "      <td>4</td>\n",
       "      <td>29333659.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>148169733.0</td>\n",
       "      <td>149271859.0</td>\n",
       "      <td>3</td>\n",
       "      <td>1102126.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>149271859.0</td>\n",
       "      <td>150660874.0</td>\n",
       "      <td>4</td>\n",
       "      <td>1389015.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>150660874.0</td>\n",
       "      <td>151212150.0</td>\n",
       "      <td>3</td>\n",
       "      <td>551276.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>151212150.0</td>\n",
       "      <td>154445997.0</td>\n",
       "      <td>4</td>\n",
       "      <td>3233847.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>154445997.0</td>\n",
       "      <td>154712843.0</td>\n",
       "      <td>3</td>\n",
       "      <td>266846.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>154712843.0</td>\n",
       "      <td>156919292.0</td>\n",
       "      <td>4</td>\n",
       "      <td>2206449.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>156919292.0</td>\n",
       "      <td>157671533.0</td>\n",
       "      <td>3</td>\n",
       "      <td>752241.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>157671533.0</td>\n",
       "      <td>158695916.0</td>\n",
       "      <td>4</td>\n",
       "      <td>1024383.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>158695916.0</td>\n",
       "      <td>159314780.0</td>\n",
       "      <td>3</td>\n",
       "      <td>618864.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>159314780.0</td>\n",
       "      <td>159318042.0</td>\n",
       "      <td>4</td>\n",
       "      <td>3262.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>159318042.0</td>\n",
       "      <td>159865080.0</td>\n",
       "      <td>3</td>\n",
       "      <td>547038.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>159865080.0</td>\n",
       "      <td>161288429.0</td>\n",
       "      <td>4</td>\n",
       "      <td>1423349.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>161288429.0</td>\n",
       "      <td>171861549.0</td>\n",
       "      <td>3</td>\n",
       "      <td>10573120.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>171861549.0</td>\n",
       "      <td>177301198.0</td>\n",
       "      <td>4</td>\n",
       "      <td>5439649.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>177301198.0</td>\n",
       "      <td>177707981.0</td>\n",
       "      <td>3</td>\n",
       "      <td>406783.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>177707981.0</td>\n",
       "      <td>180988846.0</td>\n",
       "      <td>4</td>\n",
       "      <td>3280865.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>180988846.0</td>\n",
       "      <td>181040225.0</td>\n",
       "      <td>5</td>\n",
       "      <td>51379.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>181040225.0</td>\n",
       "      <td>181254417.0</td>\n",
       "      <td>4</td>\n",
       "      <td>214192.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>181254417.0</td>\n",
       "      <td>181261169.0</td>\n",
       "      <td>3</td>\n",
       "      <td>6752.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>181261169.0</td>\n",
       "      <td>181291673.0</td>\n",
       "      <td>2</td>\n",
       "      <td>30504.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>181291673.0</td>\n",
       "      <td>181367268.0</td>\n",
       "      <td>1</td>\n",
       "      <td>75595.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          start          end  counts      length\n",
       "0           0.0   38474668.0       0  38474668.0\n",
       "1    38474668.0   38937920.0       1    463252.0\n",
       "2    38937920.0   45254948.0       0   6317028.0\n",
       "3    45254948.0   50396192.0       1   5141244.0\n",
       "4    50396192.0   51383448.0       3    987256.0\n",
       "5    51383448.0   86617928.0       4  35234480.0\n",
       "6    86617928.0  118836074.0       5  32218146.0\n",
       "7   118836074.0  148169733.0       4  29333659.0\n",
       "8   148169733.0  149271859.0       3   1102126.0\n",
       "9   149271859.0  150660874.0       4   1389015.0\n",
       "10  150660874.0  151212150.0       3    551276.0\n",
       "11  151212150.0  154445997.0       4   3233847.0\n",
       "12  154445997.0  154712843.0       3    266846.0\n",
       "13  154712843.0  156919292.0       4   2206449.0\n",
       "14  156919292.0  157671533.0       3    752241.0\n",
       "15  157671533.0  158695916.0       4   1024383.0\n",
       "16  158695916.0  159314780.0       3    618864.0\n",
       "17  159314780.0  159318042.0       4      3262.0\n",
       "18  159318042.0  159865080.0       3    547038.0\n",
       "19  159865080.0  161288429.0       4   1423349.0\n",
       "20  161288429.0  171861549.0       3  10573120.0\n",
       "21  171861549.0  177301198.0       4   5439649.0\n",
       "22  177301198.0  177707981.0       3    406783.0\n",
       "23  177707981.0  180988846.0       4   3280865.0\n",
       "24  180988846.0  181040225.0       5     51379.0\n",
       "25  181040225.0  181254417.0       4    214192.0\n",
       "26  181254417.0  181261169.0       3      6752.0\n",
       "27  181261169.0  181291673.0       2     30504.0\n",
       "28  181291673.0  181367268.0       1     75595.0"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loss_event_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find Gain Regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = counts\n",
    "gain_event_locations = dict()\n",
    "for cancer in cnv.keys():\n",
    "    df_gain = df[(df.variable == 'gain') & (df.cancer == cancer)].sort_values('start_bp')\n",
    "    values = list(df_gain.value)\n",
    "    gain_events = list()\n",
    "    start = None\n",
    "    for i in range(0, len(values)):\n",
    "        val = values[i]\n",
    "        if val > cutoffs[cancer]:\n",
    "            if start is None:\n",
    "                start = i\n",
    "        else:\n",
    "            if start is not None:\n",
    "                gain_events.append((start, i))\n",
    "                start = None\n",
    "    if start is not None:\n",
    "        gain_events.append((start, len(values)-1))\n",
    "    event_locations = list()\n",
    "    for event in gain_events:\n",
    "        start_bp = df_gain.iloc[event[0]].start_bp\n",
    "        end_bp = df_gain.iloc[event[1]].start_bp\n",
    "        \n",
    "        event_locations.append((start_bp, end_bp-start_bp))\n",
    "    gain_event_locations[cancer] = event_locations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "gain_event_patients = list()\n",
    "for cancer in gain_event_locations.keys():\n",
    "    events = gain_event_locations[cancer]\n",
    "    for event in events:\n",
    "        start = event[0]\n",
    "        end = event[0] + event[1]\n",
    "        gain_event_patients.append((start, 1))\n",
    "        gain_event_patients.append((end, 0))\n",
    "#     patients += list(gain_event_locations[cancer])\n",
    "gain_event_patients.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "count = 0\n",
    "current_bp = 0\n",
    "# results = list()\n",
    "start = list()\n",
    "end = list()\n",
    "size = list()\n",
    "total = list()\n",
    "for patient in gain_event_patients:\n",
    "    if patient[0] != current_bp:\n",
    "        start.append(current_bp)\n",
    "        end.append(patient[0])\n",
    "        size.append(patient[0]-current_bp)\n",
    "        total.append(count)\n",
    "#         results.append((current_bp, patient[0], count))\n",
    "        current_bp = patient[0]\n",
    "    if patient[1] == 1:\n",
    "        count += 1\n",
    "    else:\n",
    "        count -= 1\n",
    "gain_event_data = pd.DataFrame({'start': start, 'end': end, 'counts': total, 'length': size})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>counts</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>58198.0</td>\n",
       "      <td>0</td>\n",
       "      <td>58198.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>58198.0</td>\n",
       "      <td>92151.0</td>\n",
       "      <td>1</td>\n",
       "      <td>33953.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>92151.0</td>\n",
       "      <td>50396192.0</td>\n",
       "      <td>5</td>\n",
       "      <td>50304041.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>50396192.0</td>\n",
       "      <td>50665899.0</td>\n",
       "      <td>3</td>\n",
       "      <td>269707.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>50665899.0</td>\n",
       "      <td>51383448.0</td>\n",
       "      <td>2</td>\n",
       "      <td>717549.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>51383448.0</td>\n",
       "      <td>52787916.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1404468.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>52787916.0</td>\n",
       "      <td>138561043.0</td>\n",
       "      <td>0</td>\n",
       "      <td>85773127.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>138561043.0</td>\n",
       "      <td>147585438.0</td>\n",
       "      <td>0</td>\n",
       "      <td>9024395.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>147585438.0</td>\n",
       "      <td>148025683.0</td>\n",
       "      <td>1</td>\n",
       "      <td>440245.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>148025683.0</td>\n",
       "      <td>148169733.0</td>\n",
       "      <td>2</td>\n",
       "      <td>144050.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>148169733.0</td>\n",
       "      <td>148202794.0</td>\n",
       "      <td>1</td>\n",
       "      <td>33061.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>148202794.0</td>\n",
       "      <td>149062313.0</td>\n",
       "      <td>2</td>\n",
       "      <td>859519.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>149062313.0</td>\n",
       "      <td>151276762.0</td>\n",
       "      <td>1</td>\n",
       "      <td>2214449.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>151276762.0</td>\n",
       "      <td>151661096.0</td>\n",
       "      <td>0</td>\n",
       "      <td>384334.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>151661096.0</td>\n",
       "      <td>152391532.0</td>\n",
       "      <td>1</td>\n",
       "      <td>730436.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>152391532.0</td>\n",
       "      <td>153489615.0</td>\n",
       "      <td>2</td>\n",
       "      <td>1098083.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>153489615.0</td>\n",
       "      <td>154712843.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1223228.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>154712843.0</td>\n",
       "      <td>156919292.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2206449.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>156919292.0</td>\n",
       "      <td>157266079.0</td>\n",
       "      <td>1</td>\n",
       "      <td>346787.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>157266079.0</td>\n",
       "      <td>181254417.0</td>\n",
       "      <td>2</td>\n",
       "      <td>23988338.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>181254417.0</td>\n",
       "      <td>181367268.0</td>\n",
       "      <td>1</td>\n",
       "      <td>112851.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          start          end  counts      length\n",
       "0           0.0      58198.0       0     58198.0\n",
       "1       58198.0      92151.0       1     33953.0\n",
       "2       92151.0   50396192.0       5  50304041.0\n",
       "3    50396192.0   50665899.0       3    269707.0\n",
       "4    50665899.0   51383448.0       2    717549.0\n",
       "5    51383448.0   52787916.0       1   1404468.0\n",
       "6    52787916.0  138561043.0       0  85773127.0\n",
       "7   138561043.0  147585438.0       0   9024395.0\n",
       "8   147585438.0  148025683.0       1    440245.0\n",
       "9   148025683.0  148169733.0       2    144050.0\n",
       "10  148169733.0  148202794.0       1     33061.0\n",
       "11  148202794.0  149062313.0       2    859519.0\n",
       "12  149062313.0  151276762.0       1   2214449.0\n",
       "13  151276762.0  151661096.0       0    384334.0\n",
       "14  151661096.0  152391532.0       1    730436.0\n",
       "15  152391532.0  153489615.0       2   1098083.0\n",
       "16  153489615.0  154712843.0       1   1223228.0\n",
       "17  154712843.0  156919292.0       2   2206449.0\n",
       "18  156919292.0  157266079.0       1    346787.0\n",
       "19  157266079.0  181254417.0       2  23988338.0\n",
       "20  181254417.0  181367268.0       1    112851.0"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gain_event_data"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
