{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Define Event\n",
    "\n",
    "Before we can analyze the event, we need to identify the boundaries of the event. We will accomplish this by\n",
    "\n",
    "1. Defining the values to be classified as gains and losses\n",
    "2. Creating a counts table defining the various events\n",
    "3. Defining the porportion of patients with gain or loss to be considered significant\n",
    "4. Identifying regions of gain and loss\n",
    "5. Identifying regions where all cancer types meet criteria for gain or loss event."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cptac warning: Your version of cptac (0.8.5) is out-of-date. Latest is 0.8.6. Please run 'pip install --upgrade cptac' to update it. (C:\\Users\\chels\\anaconda3\\envs\\notebooks\\lib\\site-packages\\ipykernel_launcher.py, line 4)\n"
     ]
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import cptac\n",
    "import numpy as np\n",
    "import cnvutils"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 1: Define Parameters\n",
    "\n",
    "These are the parameters that must be set for each analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # A list of cancer types to look at (should be all capital letter ex: BRCA)\n",
    "# CANCER_TYPES = ['BRCA', 'COLON', 'HNSCC', 'LSCC', 'LUAD', 'OVARIAN']\n",
    "# The porportion of patients with gain or loss of a given gene to be considered\n",
    "PATIENT_CUTOFF = 0.2\n",
    "# The chromosome to be analyzed (should be a string)\n",
    "CHROMOSOME = '20'\n",
    "# The arm to be analyzed. Options are: 'p', 'q' or 'both'.\n",
    "ARM = 'both'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking that hnscc index is up-to-date...      \r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cptac warning: The GBM dataset is under publication embargo until March 01, 2021. CPTAC is a community resource project and data are made available rapidly after generation for community research use. The embargo allows exploring and utilizing the data, but analysis may not be published until after the embargo date. Please see https://proteomics.cancer.gov/data-portal/about/data-use-agreement or enter cptac.embargo() to open the webpage for more details. (C:\\Users\\chels\\anaconda3\\envs\\notebooks\\lib\\site-packages\\ipykernel_launcher.py, line 8)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking that lscc index is up-to-date... \r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cptac warning: The HNSCC data is currently strictly reserved for CPTAC investigators. Otherwise, you are not authorized to access these data. Additionally, even after these data become publicly available, they will be subject to a publication embargo (see https://proteomics.cancer.gov/data-portal/about/data-use-agreement or enter cptac.embargo() to open the webpage for more details). (C:\\Users\\chels\\anaconda3\\envs\\notebooks\\lib\\site-packages\\ipykernel_launcher.py, line 9)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking that luad index is up-to-date...\r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cptac warning: The LSCC data is currently strictly reserved for CPTAC investigators. Otherwise, you are not authorized to access these data. Additionally, even after these data become publicly available, they will be subject to a publication embargo (see https://proteomics.cancer.gov/data-portal/about/data-use-agreement or enter cptac.embargo() to open the webpage for more details). (C:\\Users\\chels\\anaconda3\\envs\\notebooks\\lib\\site-packages\\ipykernel_launcher.py, line 10)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                            \r"
     ]
    }
   ],
   "source": [
    "# Here you will need to load in all the cancer types to be considered. We have listed all for convenience, simply comment\n",
    "# out the cancer types you do not want to consider.\n",
    "cnv = {\n",
    "    'BRCA': cptac.Brca().get_CNV(),\n",
    "    'CCRCC': cptac.Ccrcc().get_CNV(),\n",
    "    'COLON': cptac.Colon().get_CNV(),\n",
    "    'ENDO': cptac.Endometrial().get_CNV(),\n",
    "    'GBM': cptac.Gbm().get_CNV(),\n",
    "    'HNSCC': cptac.Hnscc().get_CNV(),\n",
    "    'LSCC': cptac.Lscc().get_CNV(),\n",
    "    'LUAD': cptac.Luad().get_CNV(),\n",
    "    'OVARIAN': cptac.Ovarian().get_CNV()\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 2: Determine Cutoffs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\chels\\anaconda3\\envs\\notebooks\\lib\\site-packages\\IPython\\core\\interactiveshell.py:3338: DtypeWarning: Columns (2) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  if (await self.run_code(code, result,  async_=asy)):\n"
     ]
    }
   ],
   "source": [
    "counts = cnvutils.get_counts_table()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "cutoffs = dict()\n",
    "for cancer_type in cnv.keys():\n",
    "    df = cnv[cancer_type]\n",
    "    cutoffs[cancer_type] = len(df) * PATIENT_CUTOFF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 3: Subset Event"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# counts.cancer.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = counts[(counts.chromosome == CHROMOSOME)]\n",
    "if ARM != 'both':\n",
    "    counts = counts[counts.arm == ARM]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find Loss Regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "645\n",
      "514\n",
      "614\n",
      "723\n",
      "540\n",
      "1398\n",
      "675\n",
      "514\n",
      "495\n"
     ]
    }
   ],
   "source": [
    "df = counts\n",
    "loss_event_locations = dict()\n",
    "for cancer in cnv.keys():\n",
    "    df_loss = df[(df.variable == 'loss') & (df.cancer == cancer)].sort_values('start_bp')\n",
    "    print(len(df_loss))\n",
    "    values = list(df_loss.value)\n",
    "    loss_events = list()\n",
    "    start = None\n",
    "    for i in range(0, len(values)):\n",
    "        val = values[i]\n",
    "        if val > cutoffs[cancer]:\n",
    "            if start is None:\n",
    "                start = i\n",
    "        else:\n",
    "            if start is not None:\n",
    "                loss_events.append((start, i))\n",
    "                start = None\n",
    "    if start is not None:\n",
    "        loss_events.append((start, len(values)-1))\n",
    "    event_locations = list()\n",
    "    for event in loss_events:\n",
    "        start_bp = df_loss.iloc[event[0]].start_bp\n",
    "        end_bp = df_loss.iloc[event[1]].start_bp\n",
    "        event_locations.append((start_bp, end_bp-start_bp))\n",
    "    loss_event_locations[cancer] = event_locations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loss_event_locations['LUAD']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "loss_event_patients = list()\n",
    "for cancer in loss_event_locations.keys():\n",
    "    events = loss_event_locations[cancer]\n",
    "    for event in events:\n",
    "        start = event[0]\n",
    "        end = event[0] + event[1]\n",
    "        loss_event_patients.append((start, 1))\n",
    "        loss_event_patients.append((end, 0))\n",
    "#     patients += list(gain_event_locations[cancer])\n",
    "loss_event_patients.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Database_ID</th>\n",
       "      <th>start_bp</th>\n",
       "      <th>end_bp</th>\n",
       "      <th>cancer</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>arm</th>\n",
       "      <th>variable</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>AAR2</td>\n",
       "      <td>ENSG00000131043.7</td>\n",
       "      <td>36236459.0</td>\n",
       "      <td>36270918.0</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>20</td>\n",
       "      <td>q</td>\n",
       "      <td>gain</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>ABHD12</td>\n",
       "      <td>ENSG00000100997.14</td>\n",
       "      <td>25294742.0</td>\n",
       "      <td>25390835.0</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>20</td>\n",
       "      <td>p</td>\n",
       "      <td>gain</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>ABHD16B</td>\n",
       "      <td>ENSG00000183260.5</td>\n",
       "      <td>63861498.0</td>\n",
       "      <td>63862988.0</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>20</td>\n",
       "      <td>q</td>\n",
       "      <td>gain</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>ACOT8</td>\n",
       "      <td>ENSG00000101473.12</td>\n",
       "      <td>45841721.0</td>\n",
       "      <td>45857405.0</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>20</td>\n",
       "      <td>q</td>\n",
       "      <td>gain</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>ACSS1</td>\n",
       "      <td>ENSG00000154930.10</td>\n",
       "      <td>25006230.0</td>\n",
       "      <td>25058980.0</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>20</td>\n",
       "      <td>p</td>\n",
       "      <td>gain</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>476428</th>\n",
       "      <td>Z98752.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>43549389.0</td>\n",
       "      <td>43550949.0</td>\n",
       "      <td>HNSCC</td>\n",
       "      <td>20</td>\n",
       "      <td>q</td>\n",
       "      <td>loss</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>476429</th>\n",
       "      <td>Z98752.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>43523261.0</td>\n",
       "      <td>43523572.0</td>\n",
       "      <td>HNSCC</td>\n",
       "      <td>20</td>\n",
       "      <td>q</td>\n",
       "      <td>loss</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>476430</th>\n",
       "      <td>Z98752.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>43590009.0</td>\n",
       "      <td>43590596.0</td>\n",
       "      <td>HNSCC</td>\n",
       "      <td>20</td>\n",
       "      <td>q</td>\n",
       "      <td>loss</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>476555</th>\n",
       "      <td>ZNF840P</td>\n",
       "      <td>NaN</td>\n",
       "      <td>46484461.0</td>\n",
       "      <td>46492640.0</td>\n",
       "      <td>HNSCC</td>\n",
       "      <td>20</td>\n",
       "      <td>q</td>\n",
       "      <td>loss</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>476561</th>\n",
       "      <td>ZNF877P</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21265854.0</td>\n",
       "      <td>21266465.0</td>\n",
       "      <td>HNSCC</td>\n",
       "      <td>20</td>\n",
       "      <td>p</td>\n",
       "      <td>loss</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12236 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Name         Database_ID    start_bp      end_bp cancer  \\\n",
       "31          AAR2   ENSG00000131043.7  36236459.0  36270918.0   BRCA   \n",
       "98        ABHD12  ENSG00000100997.14  25294742.0  25390835.0   BRCA   \n",
       "107      ABHD16B   ENSG00000183260.5  63861498.0  63862988.0   BRCA   \n",
       "202        ACOT8  ENSG00000101473.12  45841721.0  45857405.0   BRCA   \n",
       "232        ACSS1  ENSG00000154930.10  25006230.0  25058980.0   BRCA   \n",
       "...          ...                 ...         ...         ...    ...   \n",
       "476428  Z98752.1                 NaN  43549389.0  43550949.0  HNSCC   \n",
       "476429  Z98752.2                 NaN  43523261.0  43523572.0  HNSCC   \n",
       "476430  Z98752.4                 NaN  43590009.0  43590596.0  HNSCC   \n",
       "476555   ZNF840P                 NaN  46484461.0  46492640.0  HNSCC   \n",
       "476561   ZNF877P                 NaN  21265854.0  21266465.0  HNSCC   \n",
       "\n",
       "       chromosome arm variable  value  \n",
       "31             20   q     gain     45  \n",
       "98             20   p     gain     34  \n",
       "107            20   q     gain     47  \n",
       "202            20   q     gain     46  \n",
       "232            20   p     gain     39  \n",
       "...           ...  ..      ...    ...  \n",
       "476428         20   q     loss      4  \n",
       "476429         20   q     loss      4  \n",
       "476430         20   q     loss      4  \n",
       "476555         20   q     loss      4  \n",
       "476561         20   p     loss      6  \n",
       "\n",
       "[12236 rows x 9 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "count = 0\n",
    "current_bp = 0\n",
    "start = list()\n",
    "end = list()\n",
    "size = list()\n",
    "total = list()\n",
    "for patient in loss_event_patients:\n",
    "    if patient[0] != current_bp:\n",
    "        start.append(current_bp)\n",
    "        end.append(patient[0])\n",
    "        size.append(patient[0]-current_bp)\n",
    "        total.append(count)\n",
    "        current_bp = patient[0]\n",
    "    if patient[1] == 1:\n",
    "        count += 1\n",
    "    else:\n",
    "        count -= 1\n",
    "loss_event_data = pd.DataFrame({'start': start, 'end': end, 'counts': total, 'length': size})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>counts</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>87250.0</td>\n",
       "      <td>0</td>\n",
       "      <td>87250.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>87250.0</td>\n",
       "      <td>869900.0</td>\n",
       "      <td>1</td>\n",
       "      <td>782650.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>869900.0</td>\n",
       "      <td>2101827.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1231927.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2101827.0</td>\n",
       "      <td>2655067.0</td>\n",
       "      <td>1</td>\n",
       "      <td>553240.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2655067.0</td>\n",
       "      <td>2656624.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1557.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2656624.0</td>\n",
       "      <td>6446723.0</td>\n",
       "      <td>1</td>\n",
       "      <td>3790099.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6446723.0</td>\n",
       "      <td>13389392.0</td>\n",
       "      <td>0</td>\n",
       "      <td>6942669.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>13389392.0</td>\n",
       "      <td>14322985.0</td>\n",
       "      <td>1</td>\n",
       "      <td>933593.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>14322985.0</td>\n",
       "      <td>14884250.0</td>\n",
       "      <td>0</td>\n",
       "      <td>561265.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>14884250.0</td>\n",
       "      <td>16272104.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1387854.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>16272104.0</td>\n",
       "      <td>38516563.0</td>\n",
       "      <td>0</td>\n",
       "      <td>22244459.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>38516563.0</td>\n",
       "      <td>38581195.0</td>\n",
       "      <td>1</td>\n",
       "      <td>64632.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         start         end  counts      length\n",
       "0          0.0     87250.0       0     87250.0\n",
       "1      87250.0    869900.0       1    782650.0\n",
       "2     869900.0   2101827.0       0   1231927.0\n",
       "3    2101827.0   2655067.0       1    553240.0\n",
       "4    2655067.0   2656624.0       1      1557.0\n",
       "5    2656624.0   6446723.0       1   3790099.0\n",
       "6    6446723.0  13389392.0       0   6942669.0\n",
       "7   13389392.0  14322985.0       1    933593.0\n",
       "8   14322985.0  14884250.0       0    561265.0\n",
       "9   14884250.0  16272104.0       1   1387854.0\n",
       "10  16272104.0  38516563.0       0  22244459.0\n",
       "11  38516563.0  38581195.0       1     64632.0"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loss_event_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find Gain Regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = counts\n",
    "gain_event_locations = dict()\n",
    "for cancer in cnv.keys():\n",
    "    df_gain = df[(df.variable == 'gain') & (df.cancer == cancer)].sort_values('start_bp')\n",
    "    values = list(df_gain.value)\n",
    "    gain_events = list()\n",
    "    start = None\n",
    "    for i in range(0, len(values)):\n",
    "        val = values[i]\n",
    "        if val > cutoffs[cancer]:\n",
    "            if start is None:\n",
    "                start = i\n",
    "        else:\n",
    "            if start is not None:\n",
    "                gain_events.append((start, i))\n",
    "                start = None\n",
    "    if start is not None:\n",
    "        gain_events.append((start, len(values)-1))\n",
    "    event_locations = list()\n",
    "    for event in gain_events:\n",
    "        start_bp = df_gain.iloc[event[0]].start_bp\n",
    "        end_bp = df_gain.iloc[event[1]].start_bp\n",
    "        \n",
    "        event_locations.append((start_bp, end_bp-start_bp))\n",
    "    gain_event_locations[cancer] = event_locations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "gain_event_patients = list()\n",
    "for cancer in gain_event_locations.keys():\n",
    "    events = gain_event_locations[cancer]\n",
    "    for event in events:\n",
    "        start = event[0]\n",
    "        end = event[0] + event[1]\n",
    "        gain_event_patients.append((start, 1))\n",
    "        gain_event_patients.append((end, 0))\n",
    "#     patients += list(gain_event_locations[cancer])\n",
    "gain_event_patients.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "count = 0\n",
    "current_bp = 0\n",
    "# results = list()\n",
    "start = list()\n",
    "end = list()\n",
    "size = list()\n",
    "total = list()\n",
    "for patient in gain_event_patients:\n",
    "    if patient[0] != current_bp:\n",
    "        start.append(current_bp)\n",
    "        end.append(patient[0])\n",
    "        size.append(patient[0]-current_bp)\n",
    "        total.append(count)\n",
    "#         results.append((current_bp, patient[0], count))\n",
    "        current_bp = patient[0]\n",
    "    if patient[1] == 1:\n",
    "        count += 1\n",
    "    else:\n",
    "        count -= 1\n",
    "gain_event_data = pd.DataFrame({'start': start, 'end': end, 'counts': total, 'length': size})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>counts</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>87250.0</td>\n",
       "      <td>0</td>\n",
       "      <td>87250.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>87250.0</td>\n",
       "      <td>2655067.0</td>\n",
       "      <td>5</td>\n",
       "      <td>2567817.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2655067.0</td>\n",
       "      <td>2656097.0</td>\n",
       "      <td>4</td>\n",
       "      <td>1030.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2656097.0</td>\n",
       "      <td>2656624.0</td>\n",
       "      <td>5</td>\n",
       "      <td>527.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2656624.0</td>\n",
       "      <td>31563166.0</td>\n",
       "      <td>5</td>\n",
       "      <td>28906542.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>31563166.0</td>\n",
       "      <td>34516409.0</td>\n",
       "      <td>6</td>\n",
       "      <td>2953243.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>34516409.0</td>\n",
       "      <td>35002404.0</td>\n",
       "      <td>5</td>\n",
       "      <td>485995.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>35002404.0</td>\n",
       "      <td>35030317.0</td>\n",
       "      <td>6</td>\n",
       "      <td>27913.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>35030317.0</td>\n",
       "      <td>35115364.0</td>\n",
       "      <td>5</td>\n",
       "      <td>85047.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>35115364.0</td>\n",
       "      <td>35195306.0</td>\n",
       "      <td>6</td>\n",
       "      <td>79942.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>35195306.0</td>\n",
       "      <td>35201745.0</td>\n",
       "      <td>5</td>\n",
       "      <td>6439.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>35201745.0</td>\n",
       "      <td>35302566.0</td>\n",
       "      <td>6</td>\n",
       "      <td>100821.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>35302566.0</td>\n",
       "      <td>38516563.0</td>\n",
       "      <td>5</td>\n",
       "      <td>3213997.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>38516563.0</td>\n",
       "      <td>38581195.0</td>\n",
       "      <td>4</td>\n",
       "      <td>64632.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>38581195.0</td>\n",
       "      <td>53567065.0</td>\n",
       "      <td>5</td>\n",
       "      <td>14985870.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>53567065.0</td>\n",
       "      <td>55997357.0</td>\n",
       "      <td>6</td>\n",
       "      <td>2430292.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>55997357.0</td>\n",
       "      <td>57168753.0</td>\n",
       "      <td>5</td>\n",
       "      <td>1171396.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>57168753.0</td>\n",
       "      <td>64255695.0</td>\n",
       "      <td>6</td>\n",
       "      <td>7086942.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>64255695.0</td>\n",
       "      <td>64290385.0</td>\n",
       "      <td>2</td>\n",
       "      <td>34690.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         start         end  counts      length\n",
       "0          0.0     87250.0       0     87250.0\n",
       "1      87250.0   2655067.0       5   2567817.0\n",
       "2    2655067.0   2656097.0       4      1030.0\n",
       "3    2656097.0   2656624.0       5       527.0\n",
       "4    2656624.0  31563166.0       5  28906542.0\n",
       "5   31563166.0  34516409.0       6   2953243.0\n",
       "6   34516409.0  35002404.0       5    485995.0\n",
       "7   35002404.0  35030317.0       6     27913.0\n",
       "8   35030317.0  35115364.0       5     85047.0\n",
       "9   35115364.0  35195306.0       6     79942.0\n",
       "10  35195306.0  35201745.0       5      6439.0\n",
       "11  35201745.0  35302566.0       6    100821.0\n",
       "12  35302566.0  38516563.0       5   3213997.0\n",
       "13  38516563.0  38581195.0       4     64632.0\n",
       "14  38581195.0  53567065.0       5  14985870.0\n",
       "15  53567065.0  55997357.0       6   2430292.0\n",
       "16  55997357.0  57168753.0       5   1171396.0\n",
       "17  57168753.0  64255695.0       6   7086942.0\n",
       "18  64255695.0  64290385.0       2     34690.0"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gain_event_data"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
