{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check for confounding variables\n",
    "\n",
    "This notebook uses chi-squared tests to look for clinical variables that are associated with having a chromosome event or not."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Get clinical tables\n",
    "- Get event tables\n",
    "- Binarize clinical columns as needed\n",
    "- For each binary column in the clinical table, make a contingency table of that column and the event table\n",
    "- Run chi squared test and save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import cptac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "dss = {\n",
    "    \"brca\": cptac.Brca,\n",
    "#     \"ccrcc\": cptac.Ccrcc,\n",
    "    \"colon\": cptac.Colon,\n",
    "#     \"endometrial\": cptac.Endometrial,\n",
    "#     \"gbm\": cptac.Gbm,\n",
    "    \"hnscc\": cptac.Hnscc,\n",
    "    \"lscc\": cptac.Lscc,\n",
    "    \"luad\": cptac.Luad,\n",
    "    \"ovarian\": cptac.Ovarian\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "clins = {}\n",
    "events = {}\n",
    "\n",
    "def find_confounding_vars_cancer_type(cancer_type):\n",
    "    \n",
    "    # Load the dataset\n",
    "    ds = dss[cancer_type]()\n",
    "    \n",
    "    # Get the clinical table\n",
    "    clin = ds.get_clinical()\n",
    "    \n",
    "    # Get the event table\n",
    "    event = pd.read_csv(f\"{cancer_type}_has_event.tsv\", sep=\"\\t\", index_col=0)\n",
    "    \n",
    "    clins[cancer_type] = clin\n",
    "    events[cancer_type] = event"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking that lscc index is up-to-date... \r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cptac warning: The HNSCC data is currently strictly reserved for CPTAC investigators. Otherwise, you are not authorized to access these data. Additionally, even after these data become publicly available, they will be subject to a publication embargo (see https://proteomics.cancer.gov/data-portal/about/data-use-agreement or enter cptac.embargo() to open the webpage for more details). (/home/caleb/anaconda3/envs/dev/lib/python3.7/site-packages/ipykernel_launcher.py, line 7)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking that luad index is up-to-date...\r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cptac warning: The LSCC data is currently strictly reserved for CPTAC investigators. Otherwise, you are not authorized to access these data. Additionally, even after these data become publicly available, they will be subject to a publication embargo (see https://proteomics.cancer.gov/data-portal/about/data-use-agreement or enter cptac.embargo() to open the webpage for more details). (/home/caleb/anaconda3/envs/dev/lib/python3.7/site-packages/ipykernel_launcher.py, line 7)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                            \r"
     ]
    }
   ],
   "source": [
    "for cancer_type in dss.keys():\n",
    "    find_confounding_vars_cancer_type(cancer_type)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Clinical vars to use\n",
    "- Age\n",
    "- Gender\n",
    "- Race\n",
    "- Tumor stage/grade\n",
    "- Histology/subtype\n",
    "- TP53 and other mutation status for any that have like a 10% (maybe 5%) or greater frequency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">brca</th>\n",
       "      <th>age</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gender</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>race</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tumor_stage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>subtype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">colon</th>\n",
       "      <th>age</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gender</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>race</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tumor_stage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>subtype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">hnscc</th>\n",
       "      <th>age</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gender</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>race</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tumor_stage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>subtype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">lscc</th>\n",
       "      <th>age</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gender</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>race</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tumor_stage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>subtype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">luad</th>\n",
       "      <th>age</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gender</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>race</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tumor_stage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>subtype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">ovarian</th>\n",
       "      <th>age</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gender</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>race</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tumor_stage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>subtype</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [(brca, age), (brca, gender), (brca, race), (brca, tumor_stage), (brca, subtype), (colon, age), (colon, gender), (colon, race), (colon, tumor_stage), (colon, subtype), (hnscc, age), (hnscc, gender), (hnscc, race), (hnscc, tumor_stage), (hnscc, subtype), (lscc, age), (lscc, gender), (lscc, race), (lscc, tumor_stage), (lscc, subtype), (luad, age), (luad, gender), (luad, race), (luad, tumor_stage), (luad, subtype), (ovarian, age), (ovarian, gender), (ovarian, race), (ovarian, tumor_stage), (ovarian, subtype)]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cancer_types = [\n",
    "    \"brca\",\n",
    "    \"colon\",\n",
    "    \"hnscc\",\n",
    "    \"lscc\",\n",
    "    \"luad\",\n",
    "    \"ovarian\"\n",
    "]\n",
    "\n",
    "clinical_vars = [\n",
    "    \"age\",\n",
    "    \"gender\",\n",
    "    \"race\",\n",
    "    \"tumor_stage\",\n",
    "    \"subtype\",\n",
    "    # And mutation statuses for frequently mutated genes\n",
    "]\n",
    "\n",
    "cancer_clinical_vars = pd.DataFrame(index=pd.MultiIndex.from_product([cancer_types, clinical_vars]))\n",
    "cancer_clinical_vars"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------------------------------------------------------------------------\n",
      "brca\n",
      "\n",
      "Replicate_Measurement_IDs\n",
      "Sample_Tumor_Normal\n",
      "Age.in.Month\n",
      "Gender\n",
      "Race\n",
      "Human.Readable.Label\n",
      "Experiment\n",
      "Channel\n",
      "Stage\n",
      "PAM50\n",
      "NMF.v2.1\n",
      "ER\n",
      "PR\n",
      "ER.IHC.Score\n",
      "PR.IHC.Score\n",
      "Coring.or.Excision\n",
      "Ischemia.Time.in.Minutes\n",
      "Ischemia.Decade\n",
      "Necrosis\n",
      "Tumor.Cellularity\n",
      "Total.Cellularity\n",
      "In.CR\n",
      "QC.status\n",
      "\n",
      "------------------------------------------------------------------------\n",
      "colon\n",
      "\n",
      "Sample_Tumor_Normal\n",
      "Age\n",
      "CEA\n",
      "Gender\n",
      "Lymphatic_Invasion\n",
      "Mucinous\n",
      "Perineural_Invasion\n",
      "Polyps_History\n",
      "Polyps_Present\n",
      "Stage\n",
      "Subsite\n",
      "Synchronous_Tumors\n",
      "Tumor.Status\n",
      "Vascular_Invasion\n",
      "Vital.Status\n",
      "pathalogy_N_stage\n",
      "pathalogy_T_stage\n",
      "\n",
      "------------------------------------------------------------------------\n",
      "hnscc\n",
      "\n",
      "Sample_Tumor_Normal\n",
      "Cored_Sample\n",
      "P16\n",
      "age\n",
      "alcohol_consum\n",
      "clinic_staging_dist_metas\n",
      "country\n",
      "follow_up_days\n",
      "follow_up_is_contact\n",
      "follow_up_vital_status\n",
      "gender\n",
      "histologic_grade\n",
      "histologic_type\n",
      "num_pack_years_sm\n",
      "num_smoke_per_day\n",
      "num_yrs_alc_con\n",
      "patho_staging_curated\n",
      "patho_staging_orignial\n",
      "patho_staging_pn\n",
      "patho_staging_pt\n",
      "smoke_age_start\n",
      "smoke_age_stop\n",
      "smoking_history\n",
      "smoking_inferred_binary\n",
      "smoking_second_hand\n",
      "tumor_focality\n",
      "tumor_necrosis\n",
      "tumor_site_curated\n",
      "tumor_site_original\n",
      "tumor_size_cm\n",
      "\n",
      "------------------------------------------------------------------------\n",
      "lscc\n",
      "\n",
      "Sample_Tumor_Normal\n",
      "Smoking.History\n",
      "Stage\n",
      "Country.of.Origin\n",
      "Age\n",
      "Gender\n",
      "Ethnicity\n",
      "Cigarettes.per.Day\n",
      "Pack.Years.Smoked\n",
      "Secondhand.Smoke\n",
      "\n",
      "------------------------------------------------------------------------\n",
      "luad\n",
      "\n",
      "Sample.IDs\n",
      "Sample_Tumor_Normal\n",
      "Smoking.Status\n",
      "Stage\n",
      "Region.of.Origin\n",
      "Country.of.Origin\n",
      "Age\n",
      "Gender\n",
      "Ethnicity\n",
      "Height.cm\n",
      "Weight.kg\n",
      "BMI\n",
      "Cigarettes.per.Day\n",
      "Pack.Years.Smoked\n",
      "Smoking.History\n",
      "Secondhand.Smoke\n",
      "\n",
      "------------------------------------------------------------------------\n",
      "ovarian\n",
      "\n",
      "Sample_Tumor_Normal\n",
      "Participant_Procurement_Age\n",
      "Participant_Gender\n",
      "Participant_Race\n",
      "Participant_Ethnicity\n",
      "Participant_Jewish_Heritage\n",
      "Participant_History_Malignancy\n",
      "Participant_History_Chemotherapy\n",
      "Participant_History_Neo-adjuvant_Treatment\n",
      "Participant_History_Radiation_Therapy\n",
      "Participant_History_Hormonal_Therapy\n",
      "Aliquots_Plasma\n",
      "Blood_Collection_Time\n",
      "Blood_Collection_Method\n",
      "Anesthesia_Time\n",
      "Tumor_Surgical_Devascularized_Time\n",
      "Tumor_Sample_Number\n",
      "Tumor_Sample_1_Weight\n",
      "Tumor_Sample_1_LN2_Time\n",
      "Tumor_Sample_1_Ischemia_Time\n",
      "Tumor_Sample_2_Weight\n",
      "Tumor_Sample_2_LN2_Time\n",
      "Tumor_Sample_2_Ischemia_Time\n",
      "Tumor_Sample_3_Weight\n",
      "Tumor_Sample_3_LN2_Time\n",
      "Tumor_Sample_3_Ischemia_Time\n",
      "Tumor_Sample_4_Weight\n",
      "Tumor_Sample_4_LN2_Time\n",
      "Tumor_Sample_4_Ischemia_Time\n",
      "Tumor_Sample_5_Weight\n",
      "Tumor_Sample_5_LN2_Time\n",
      "Tumor_Sample_5_Ischemia_Time\n",
      "Normal_Sample_Number\n",
      "Normal_Sample_1_Surgical_Devascularized_Time\n",
      "Normal_Sample_1_Weight\n",
      "Normal_Sample_1_LN2_Time\n",
      "Normal_Sample_1_Ischemia_Time\n",
      "Normal_Sample_2_Surgical_Devascularized_Time\n",
      "Normal_Sample_2_Weight\n",
      "Normal_Sample_2_LN2_Time\n",
      "Normal_Sample_2_Ischemia_Time\n",
      "Normal_Sample_3_Surgical_Devascularized_Time\n",
      "Normal_Sample_3_Weight\n",
      "Normal_Sample_3_LN2_Time\n",
      "Normal_Sample_3_Ischemia_Time\n",
      "Normal_Sample_4_Surgical_Devascularized_Time\n",
      "Normal_Sample_4_Weight\n",
      "Normal_Sample_4_LN2_Time\n",
      "Normal_Sample_4_Ischemia_Time\n",
      "Normal_Sample_5_Surgical_Devascularized_Time\n",
      "Normal_Sample_5_Weight\n",
      "Normal_Sample_5_LN2_Time\n",
      "Normal_Sample_5_Ischemia_Time\n",
      "Origin_Site_Disease\n",
      "Anatomic_Site_Tumor\n",
      "Anatomic_Lateral_Position_Tumor\n",
      "Histological_Subtype\n",
      "Method_of_Pathologic_Diagnosis\n",
      "Tumor_Stage_Ovary_FIGO\n",
      "Tumor_Grade\n",
      "Tumor_Residual_Disease_Max_Diameter\n",
      "Days_Between_Collection_And_Last_Contact\n",
      "Vital_Status\n",
      "Days_Between_Collection_And_Death\n",
      "Tumor_Status\n",
      "Review_Of_Initial_Pathological_Findings\n",
      "Pathology_Review_Consistent_With_Diagnosis\n",
      "Adjuvant_Radiation_Therapy\n",
      "Adjuvant_Pharmaceutical_Therapy\n",
      "Adjuvant_Immunotherapy\n",
      "Adjuvant_Hormone_Therapy\n",
      "Adjuvant_Targeted_Molecular_Therapy\n",
      "Response_After_Surgery_And_Adjuvant_Therapies\n",
      "New_Tumor_Event_After_Initial_Treatment\n",
      "New_Tumor_Event_Type\n",
      "New_Tumor_Event_Site\n",
      "Other_New_Tumor_Event_Site\n",
      "Days_Between_Collection_And_New_Tumor_Event\n",
      "New_Tumor_Event_Diagnosis\n",
      "New_Tumor_Event_Surgery\n",
      "Days_Between_Collection_And_New_Tumor_Event_Surgery\n",
      "New_Tumor_Event_Chemotherapy\n",
      "New_Tumor_Event_Immunotherapy\n",
      "New_Tumor_Event_Hormone_Therapy\n",
      "New_Tumor_Event_Targeted_Molecular_Therapy\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for cancer in clins.keys():\n",
    "    df = clins[cancer]\n",
    "    print(\"------------------------------------------------------------------------\")\n",
    "    print(cancer)\n",
    "    print()\n",
    "    [print(col) for col in df.columns]\n",
    "    print()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
