{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e5ca066d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import cnvutils\n",
    "\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "from cnvutils.filenames import (\n",
    "    get_ttest_results_path,\n",
    ")\n",
    "\n",
    "from cnvutils.constants import (\n",
    "    SIG_CUTOFF,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "aa69b3b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def _get_protein_n(ttest_df):\n",
    "    return ttest_df.\\\n",
    "    notna().\\\n",
    "    any(axis=1).\\\n",
    "    reset_index(drop=False).\\\n",
    "    pivot(index=\"protein\", columns=\"group\", values=0).\\\n",
    "    any(axis=1).\\\n",
    "    sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4831814a",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>has_uniq</th>\n",
       "      <th>not_has_uniq</th>\n",
       "      <th>shared</th>\n",
       "      <th>neither</th>\n",
       "      <th>n</th>\n",
       "      <th>has_uniq_prop</th>\n",
       "      <th>not_has_prop</th>\n",
       "      <th>shared_prop</th>\n",
       "      <th>neither_prop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>all</th>\n",
       "      <td>2</td>\n",
       "      <td>20</td>\n",
       "      <td>112</td>\n",
       "      <td>8</td>\n",
       "      <td>142</td>\n",
       "      <td>0.014085</td>\n",
       "      <td>0.140845</td>\n",
       "      <td>0.788732</td>\n",
       "      <td>0.056338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coad</th>\n",
       "      <td>11</td>\n",
       "      <td>18</td>\n",
       "      <td>39</td>\n",
       "      <td>74</td>\n",
       "      <td>91</td>\n",
       "      <td>0.120879</td>\n",
       "      <td>0.197802</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.813187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hnscc</th>\n",
       "      <td>11</td>\n",
       "      <td>24</td>\n",
       "      <td>50</td>\n",
       "      <td>57</td>\n",
       "      <td>122</td>\n",
       "      <td>0.090164</td>\n",
       "      <td>0.196721</td>\n",
       "      <td>0.409836</td>\n",
       "      <td>0.467213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lscc</th>\n",
       "      <td>13</td>\n",
       "      <td>16</td>\n",
       "      <td>78</td>\n",
       "      <td>35</td>\n",
       "      <td>129</td>\n",
       "      <td>0.100775</td>\n",
       "      <td>0.124031</td>\n",
       "      <td>0.604651</td>\n",
       "      <td>0.271318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>luad</th>\n",
       "      <td>14</td>\n",
       "      <td>38</td>\n",
       "      <td>59</td>\n",
       "      <td>31</td>\n",
       "      <td>129</td>\n",
       "      <td>0.108527</td>\n",
       "      <td>0.294574</td>\n",
       "      <td>0.457364</td>\n",
       "      <td>0.240310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ov</th>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>80</td>\n",
       "      <td>0.187500</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.587500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pdac</th>\n",
       "      <td>1</td>\n",
       "      <td>61</td>\n",
       "      <td>14</td>\n",
       "      <td>66</td>\n",
       "      <td>120</td>\n",
       "      <td>0.008333</td>\n",
       "      <td>0.508333</td>\n",
       "      <td>0.116667</td>\n",
       "      <td>0.550000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       has_uniq  not_has_uniq  shared  neither    n  has_uniq_prop  \\\n",
       "all           2            20     112        8  142       0.014085   \n",
       "coad         11            18      39       74   91       0.120879   \n",
       "hnscc        11            24      50       57  122       0.090164   \n",
       "lscc         13            16      78       35  129       0.100775   \n",
       "luad         14            38      59       31  129       0.108527   \n",
       "ov           15             0       0      127   80       0.187500   \n",
       "pdac          1            61      14       66  120       0.008333   \n",
       "\n",
       "       not_has_prop  shared_prop  neither_prop  \n",
       "all        0.140845     0.788732      0.056338  \n",
       "coad       0.197802     0.428571      0.813187  \n",
       "hnscc      0.196721     0.409836      0.467213  \n",
       "lscc       0.124031     0.604651      0.271318  \n",
       "luad       0.294574     0.457364      0.240310  \n",
       "ov         0.000000     0.000000      1.587500  \n",
       "pdac       0.508333     0.116667      0.550000  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def groups_overlaps(\n",
    "    chromosome,\n",
    "    arm,\n",
    "    gain_or_loss,\n",
    "    cis_or_trans,\n",
    "    proteomics_or_transcriptomics,\n",
    "    comparison,\n",
    "    source,\n",
    "    level=None,\n",
    "    data_dir=os.path.join(os.getcwd(), \"..\", \"data\"),\n",
    "):\n",
    "    \n",
    "    if comparison == \"has_event\":\n",
    "        comparison_name = \"has_vs_not_has_event\"\n",
    "        groups = [\"tumor\", \"normal\"]\n",
    "        \n",
    "    elif comparison == \"tumor\":\n",
    "        comparison_name = \"tumor_vs_normal\"\n",
    "        groups = [\"has_event\", \"not_has_event\"]\n",
    "\n",
    "    all_ttest = pd.DataFrame()\n",
    "    for group in groups:\n",
    "        \n",
    "        group_path = get_ttest_results_path(\n",
    "            data_dir=data_dir,\n",
    "            source=source,\n",
    "            level=level,\n",
    "            chromosome=chromosome,\n",
    "            arm=arm,\n",
    "            gain_or_loss=gain_or_loss,\n",
    "            cis_or_trans=cis_or_trans,\n",
    "            proteomics_or_transcriptomics=proteomics_or_transcriptomics,\n",
    "            group=group,\n",
    "            comparison_name=comparison_name,\n",
    "        )\n",
    "        \n",
    "        group_ttest = pd.\\\n",
    "        read_csv(group_path, sep=\"\\t\").\\\n",
    "        dropna(subset=\"adj_p\")\n",
    "        \n",
    "        # We're going to combine adj_p and change for isoforms, weighted by sample\n",
    "        # size (n) for each isoform\n",
    "        \n",
    "        # 1. Calculate sample sizes for each isoform\n",
    "        group_ttest = group_ttest.assign(\n",
    "            n=group_ttest[\"tumor_sample_size\"] + group_ttest[\"not_tumor_sample_size\"],\n",
    "        ).\\\n",
    "        drop(columns=[\"tumor_sample_size\", \"not_tumor_sample_size\"])\n",
    "        \n",
    "        # 2. Calculate the overall sample size for all isoforms in each protein\n",
    "        group_ns = group_ttest.\\\n",
    "        groupby([\"cancer_type\", \"protein\"]).\\\n",
    "        apply(lambda x: x[\"n\"].sum()).\\\n",
    "        rename(\"group_n\")\n",
    "        \n",
    "        group_ttest = group_ttest.merge(\n",
    "            right=group_ns,\n",
    "            on=[\"protein\", \"cancer_type\"],\n",
    "            how=\"outer\",\n",
    "        )\n",
    "        \n",
    "        # 3. Calculate weights by dividing the sample size for each isoform by the \n",
    "        # overall sample size for the protein. Note that in this way, the weights within\n",
    "        # each protein sum to 1.\n",
    "        group_ttest = group_ttest.assign(weight=group_ttest[\"n\"] / group_ttest[\"group_n\"])\n",
    "        \n",
    "        # 4. Multiply the weights by the adj_p and change columns\n",
    "        group_ttest = group_ttest.assign(\n",
    "            adj_p=group_ttest[\"adj_p\"] * group_ttest[\"weight\"],\n",
    "            change=group_ttest[\"change\"] * group_ttest[\"weight\"],\n",
    "        )\n",
    "        \n",
    "        # 5. Sum the weighted adj_p and change columns to get the mean for isoforms\n",
    "        # Proteins with only one form will be unaffected \n",
    "        # Because the weights were normalized to sum to 1 within each protein, we\n",
    "        # don't have to divide the sums by the number of isoforms per protein\n",
    "        group_ttest = group_ttest.\\\n",
    "        groupby([\"protein\", \"cancer_type\"])[[\"adj_p\", \"change\"]].\\\n",
    "        sum().\\\n",
    "        reset_index(drop=False)\n",
    "        \n",
    "        # To capture whether a protein is up- or down-regulated, multiply p\n",
    "        # values by the sign of the associated change\n",
    "        group_ttest = group_ttest.assign(\n",
    "            sign_adj_p=group_ttest[\"adj_p\"] * np.sign(group_ttest[\"change\"])\n",
    "        )\n",
    "\n",
    "        # Pivot the table to have one row per protein and one column per cancer\n",
    "        group_ttest = group_ttest.pivot(\n",
    "            index=\"protein\",\n",
    "            columns=\"cancer_type\",\n",
    "            values=\"sign_adj_p\",\n",
    "        ).\\\n",
    "        reset_index(drop=False).\\\n",
    "        assign(group=group)\n",
    "        \n",
    "        group_ttest.columns.name = None\n",
    "        \n",
    "        all_ttest = pd.concat([all_ttest, group_ttest])\n",
    "        \n",
    "    all_ttest = all_ttest.\\\n",
    "    reset_index(drop=True).\\\n",
    "    set_index([\"group\", \"protein\"])\n",
    "\n",
    "    # Count how many proteins only show up in has_event or not_has_event, across\n",
    "    # all cancer types aggregated together\n",
    "    all_sig = ((\n",
    "            (all_ttest <= SIG_CUTOFF) &\n",
    "            (all_ttest >= -SIG_CUTOFF)\n",
    "        ).\\\n",
    "        sum(axis=1).\\\n",
    "        rename(\"sig_count\").\\\n",
    "        reset_index(drop=False).\\\n",
    "        pivot(\n",
    "            index=\"protein\",\n",
    "            columns=\"group\",\n",
    "            values=\"sig_count\",\n",
    "    ) > 0)\n",
    "    \n",
    "    has_uniq = all_sig[\"has_event\"] & ~all_sig[\"not_has_event\"]\n",
    "    not_has_uniq = ~all_sig[\"has_event\"] & all_sig[\"not_has_event\"]\n",
    "    shared = all_sig[\"has_event\"] & all_sig[\"not_has_event\"]\n",
    "    neither = ~all_sig[\"has_event\"] & ~all_sig[\"not_has_event\"]\n",
    "    \n",
    "    all_counts = pd.DataFrame({\n",
    "        \"has_uniq\": [has_uniq.sum()],\n",
    "        \"not_has_uniq\": [not_has_uniq.sum()],\n",
    "        \"shared\": shared.sum(),\n",
    "        \"neither\": neither.sum(),\n",
    "        \"n\": _get_protein_n(all_ttest),\n",
    "    }, index=[\"all\"])\n",
    "    \n",
    "    # Do that for each cancer type\n",
    "    for cancer_type in all_ttest.columns:\n",
    "        \n",
    "        cancer_ttest = all_ttest[[cancer_type]]\n",
    "#         return cancer_ttest\n",
    "        cancer_sig = (\n",
    "            (cancer_ttest <= SIG_CUTOFF) &\n",
    "            (cancer_ttest >= -SIG_CUTOFF)\n",
    "        ).\\\n",
    "        dropna().\\\n",
    "        reset_index(drop=False).\\\n",
    "        pivot(\n",
    "            index=\"protein\",\n",
    "            columns=\"group\",\n",
    "            values=cancer_type,\n",
    "        ).\\\n",
    "        fillna(False)\n",
    "        \n",
    "        has_uniq = cancer_sig[\"has_event\"] & ~cancer_sig[\"not_has_event\"]\n",
    "        not_has_uniq = ~cancer_sig[\"has_event\"] & cancer_sig[\"not_has_event\"]\n",
    "        shared = cancer_sig[\"has_event\"] & cancer_sig[\"not_has_event\"]\n",
    "        neither = ~cancer_sig[\"has_event\"] & ~cancer_sig[\"not_has_event\"]\n",
    "\n",
    "        cancer_counts = pd.DataFrame({\n",
    "            \"has_uniq\": [has_uniq.sum()],\n",
    "            \"not_has_uniq\": [not_has_uniq.sum()],\n",
    "            \"shared\": shared.sum(),\n",
    "            \"neither\": neither.sum(),\n",
    "            \"n\": _get_protein_n(cancer_ttest),\n",
    "        }, index=[cancer_type])\n",
    "        \n",
    "        all_counts = pd.concat([all_counts, cancer_counts])\n",
    "        \n",
    "    all_counts = all_counts.assign(\n",
    "        has_uniq_prop=all_counts[\"has_uniq\"] / all_counts[\"n\"],\n",
    "        not_has_prop=all_counts[\"not_has_uniq\"] / all_counts[\"n\"],\n",
    "        shared_prop=all_counts[\"shared\"] / all_counts[\"n\"],\n",
    "        neither_prop=all_counts[\"neither\"] / all_counts[\"n\"],\n",
    "    )\n",
    "    \n",
    "    return all_counts\n",
    "\n",
    "res = groups_overlaps(\n",
    "    chromosome=8,\n",
    "    arm=\"p\",\n",
    "    gain_or_loss=\"loss\",\n",
    "    cis_or_trans=\"cis\",\n",
    "    proteomics_or_transcriptomics=\"proteomics\",\n",
    "    comparison=\"tumor\",\n",
    "    source=\"gistic\",\n",
    "    level=\"gene\",\n",
    ")\n",
    "\n",
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2565294b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "has_uniq          10.000000\n",
       "not_has_uniq      31.400000\n",
       "shared            48.000000\n",
       "neither           52.600000\n",
       "n                118.200000\n",
       "has_uniq_prop      0.085736\n",
       "not_has_prop       0.264292\n",
       "shared_prop        0.403418\n",
       "neither_prop       0.468406\n",
       "dtype: float64"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res.drop(index=[\"all\", \"ov\"]).mean()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
