{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e5ca066d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import cnvutils\n",
    "\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "from cnvutils.filenames import (\n",
    "    get_ttest_results_path,\n",
    ")\n",
    "\n",
    "from cnvutils.constants import (\n",
    "    SIG_CUTOFF,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e0db8f7a",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                  coad\n",
      "group         protein                 \n",
      "has_event     ADAM28               NaN\n",
      "              ADAM9       4.970439e-01\n",
      "              ADAMDEC1   -1.422175e-06\n",
      "              ADGRA2      9.329758e-01\n",
      "              AGPAT5      3.484816e-01\n",
      "              ANGPT2     -6.729069e-04\n",
      "              ARHGEF10   -6.122925e-03\n",
      "              ASAH1      -6.403538e-03\n",
      "              ASH2L       9.589616e-06\n",
      "              ATP6V1B2   -2.435455e-03\n",
      "              BAG4       -6.719475e-01\n",
      "              BIN3       -4.259459e-09\n",
      "              BMP1        9.904116e-01\n",
      "              BNIP3L               NaN\n",
      "              BRF2       -1.852613e-01\n",
      "              CCAR2       6.694730e-02\n",
      "              CCDC25     -4.308169e-04\n",
      "              CDCA2                NaN\n",
      "              CHMP7      -1.626773e-06\n",
      "              CLDN23               NaN\n",
      "              CLN8                 NaN\n",
      "              CLU        -1.333792e-06\n",
      "              CNOT7       5.694846e-01\n",
      "              CSGALNACT1           NaN\n",
      "              CTSB       -2.084619e-01\n",
      "              DCTN6      -1.324010e-10\n",
      "              DDHD2      -1.686095e-07\n",
      "              DEFA3       1.618773e-03\n",
      "              DEFA4       1.052640e-01\n",
      "              DEFA5      -8.690498e-04\n",
      "...                                ...\n",
      "not_has_event REEP4       1.575330e-08\n",
      "              SARAF       5.330181e-02\n",
      "              SCARA3               NaN\n",
      "              SCARA5     -1.901675e-15\n",
      "              SFTPC                NaN\n",
      "              SH2D4A      1.568418e-08\n",
      "              SLC25A37             NaN\n",
      "              SLC39A14    6.985171e-06\n",
      "              SLC7A2               NaN\n",
      "              SORBS3     -2.728043e-16\n",
      "              SOX7                 NaN\n",
      "              STC1                 NaN\n",
      "              TACC1      -2.943825e-13\n",
      "              TDRP                 NaN\n",
      "              TM2D2                NaN\n",
      "              TNFRSF10A            NaN\n",
      "              TNFRSF10B  -9.307185e-01\n",
      "              TNFRSF10C            NaN\n",
      "              TNFRSF10D            NaN\n",
      "              TNKS                 NaN\n",
      "              TRIM35               NaN\n",
      "              TTI2        1.246316e-04\n",
      "              TUSC3                NaN\n",
      "              UBXN8                NaN\n",
      "              VPS37A     -2.365123e-07\n",
      "              WRN         8.654321e-01\n",
      "              XPO7       -9.595154e-01\n",
      "              ZDHHC2               NaN\n",
      "              ZNF395               NaN\n",
      "              ZNF703      4.773893e-10\n",
      "\n",
      "[276 rows x 1 columns]\n",
      "                                 hnscc\n",
      "group         protein                 \n",
      "has_event     ADAM28      6.951630e-01\n",
      "              ADAM9       3.898140e-07\n",
      "              ADAMDEC1    2.762722e-01\n",
      "              ADGRA2     -1.311133e-03\n",
      "              AGPAT5     -8.431414e-03\n",
      "              ANGPT2      8.482371e-01\n",
      "              ARHGEF10   -4.637602e-02\n",
      "              ASAH1      -2.400201e-01\n",
      "              ASH2L       2.607954e-02\n",
      "              ATP6V1B2   -1.579270e-03\n",
      "              BAG4        6.736448e-02\n",
      "              BIN3       -8.157197e-07\n",
      "              BMP1        3.395323e-06\n",
      "              BNIP3L     -9.092174e-03\n",
      "              BRF2        2.151064e-01\n",
      "              CCAR2      -8.178645e-02\n",
      "              CCDC25     -9.039093e-08\n",
      "              CDCA2      -4.477286e-01\n",
      "              CHMP7      -4.423895e-01\n",
      "              CLDN23               NaN\n",
      "              CLN8        7.589653e-01\n",
      "              CLU        -3.568783e-05\n",
      "              CNOT7      -5.426030e-01\n",
      "              CSGALNACT1  9.596938e-01\n",
      "              CTSB        8.431414e-03\n",
      "              DCTN6      -1.844253e-06\n",
      "              DDHD2      -3.288807e-02\n",
      "              DEFA3       4.532117e-03\n",
      "              DEFA4       9.607336e-04\n",
      "              DEFA5                NaN\n",
      "...                                ...\n",
      "not_has_event REEP4       2.413168e-02\n",
      "              SARAF       4.757355e-01\n",
      "              SCARA3      5.957575e-01\n",
      "              SCARA5     -5.609919e-17\n",
      "              SFTPC      -7.486748e-01\n",
      "              SH2D4A     -6.585878e-07\n",
      "              SLC25A37             NaN\n",
      "              SLC39A14    1.916725e-15\n",
      "              SLC7A2     -1.086914e-01\n",
      "              SORBS3     -2.097217e-20\n",
      "              SOX7                 NaN\n",
      "              STC1        5.575603e-01\n",
      "              TACC1      -1.092274e-05\n",
      "              TDRP        9.440693e-06\n",
      "              TM2D2                NaN\n",
      "              TNFRSF10A   4.396329e-05\n",
      "              TNFRSF10B  -8.483368e-01\n",
      "              TNFRSF10C  -9.880785e-01\n",
      "              TNFRSF10D            NaN\n",
      "              TNKS       -7.891916e-01\n",
      "              TRIM35               NaN\n",
      "              TTI2        6.806248e-05\n",
      "              TUSC3       7.891916e-01\n",
      "              UBXN8      -9.678302e-01\n",
      "              VPS37A      3.930253e-01\n",
      "              WRN         4.480677e-02\n",
      "              XPO7        9.143474e-03\n",
      "              ZDHHC2               NaN\n",
      "              ZNF395      9.937342e-02\n",
      "              ZNF703     -7.663292e-02\n",
      "\n",
      "[276 rows x 1 columns]\n",
      "                                  lscc\n",
      "group         protein                 \n",
      "has_event     ADAM28      2.631891e-02\n",
      "              ADAM9       7.687595e-08\n",
      "              ADAMDEC1    1.594819e-06\n",
      "              ADGRA2     -8.512702e-04\n",
      "              AGPAT5      6.633920e-02\n",
      "              ANGPT2     -5.954201e-05\n",
      "              ARHGEF10   -9.063330e-24\n",
      "              ASAH1      -2.082820e-21\n",
      "              ASH2L       4.961908e-11\n",
      "              ATP6V1B2   -2.646096e-14\n",
      "              BAG4        3.664086e-05\n",
      "              BIN3       -2.893082e-15\n",
      "              BMP1        8.427926e-10\n",
      "              BNIP3L     -4.312674e-07\n",
      "              BRF2        7.390081e-06\n",
      "              CCAR2       6.614400e-02\n",
      "              CCDC25      4.390366e-01\n",
      "              CDCA2       1.490772e-05\n",
      "              CHMP7                NaN\n",
      "              CLDN23     -4.414896e-01\n",
      "              CLN8       -2.586130e-01\n",
      "              CLU        -7.262143e-08\n",
      "              CNOT7       3.425201e-02\n",
      "              CSGALNACT1 -5.702447e-02\n",
      "              CTSB        4.527900e-03\n",
      "              DCTN6      -6.565939e-20\n",
      "              DDHD2       7.513667e-01\n",
      "              DEFA3      -9.283823e-01\n",
      "              DEFA4       5.961853e-01\n",
      "              DEFA5      -5.379559e-06\n",
      "...                                ...\n",
      "not_has_event REEP4       7.698334e-22\n",
      "              SARAF      -7.498878e-04\n",
      "              SCARA3      6.722680e-01\n",
      "              SCARA5     -1.204387e-24\n",
      "              SFTPC      -2.070405e-38\n",
      "              SH2D4A     -1.264362e-09\n",
      "              SLC25A37    2.174339e-01\n",
      "              SLC39A14    4.475710e-30\n",
      "              SLC7A2      7.797161e-01\n",
      "              SORBS3     -1.963948e-57\n",
      "              SOX7       -1.070342e-01\n",
      "              STC1        1.067018e-05\n",
      "              TACC1      -5.893340e-10\n",
      "              TDRP       -3.081218e-08\n",
      "              TM2D2      -1.711500e-01\n",
      "              TNFRSF10A   3.458006e-05\n",
      "              TNFRSF10B  -1.315067e-01\n",
      "              TNFRSF10C  -9.890788e-03\n",
      "              TNFRSF10D            NaN\n",
      "              TNKS        8.428639e-03\n",
      "              TRIM35     -3.878503e-02\n",
      "              TTI2        2.790725e-20\n",
      "              TUSC3       8.885207e-02\n",
      "              UBXN8      -4.476771e-08\n",
      "              VPS37A      7.013590e-01\n",
      "              WRN         6.596142e-06\n",
      "              XPO7       -9.874566e-01\n",
      "              ZDHHC2     -6.371928e-05\n",
      "              ZNF395     -1.731138e-02\n",
      "              ZNF703      2.901756e-13\n",
      "\n",
      "[276 rows x 1 columns]\n",
      "                                  luad\n",
      "group         protein                 \n",
      "has_event     ADAM28      2.581651e-01\n",
      "              ADAM9       6.552871e-09\n",
      "              ADAMDEC1    5.712288e-01\n",
      "              ADGRA2     -1.140117e-03\n",
      "              AGPAT5      2.826090e-01\n",
      "              ANGPT2     -2.351690e-11\n",
      "              ARHGEF10   -6.355713e-23\n",
      "              ASAH1      -1.044575e-07\n",
      "              ASH2L       3.198511e-03\n",
      "              ATP6V1B2   -4.400725e-06\n",
      "              BAG4        3.030358e-01\n",
      "              BIN3       -4.589114e-10\n",
      "              BMP1        5.797146e-03\n",
      "              BNIP3L      8.736301e-01\n",
      "              BRF2        3.383344e-01\n",
      "              CCAR2      -4.344174e-01\n",
      "              CCDC25      3.153626e-01\n",
      "              CDCA2       1.067110e-04\n",
      "              CHMP7      -4.687867e-03\n",
      "              CLDN23     -4.368996e-01\n",
      "              CLN8        2.581651e-01\n",
      "              CLU        -4.735686e-13\n",
      "              CNOT7       1.615348e-01\n",
      "              CSGALNACT1 -1.730556e-01\n",
      "              CTSB        3.869980e-08\n",
      "              DCTN6      -2.694726e-12\n",
      "              DDHD2      -3.223216e-06\n",
      "              DEFA3      -2.581651e-01\n",
      "              DEFA4      -1.198738e-01\n",
      "              DEFA5      -1.414881e-01\n",
      "...                                ...\n",
      "not_has_event REEP4       1.968966e-02\n",
      "              SARAF       2.045016e-01\n",
      "              SCARA3     -6.487713e-05\n",
      "              SCARA5     -1.030452e-20\n",
      "              SFTPC      -6.887337e-03\n",
      "              SH2D4A      1.946181e-05\n",
      "              SLC25A37    1.414021e-02\n",
      "              SLC39A14    5.492258e-07\n",
      "              SLC7A2      5.095295e-03\n",
      "              SORBS3     -5.062085e-50\n",
      "              SOX7                 NaN\n",
      "              STC1        2.264983e-03\n",
      "              TACC1      -5.558702e-17\n",
      "              TDRP       -4.046470e-03\n",
      "              TM2D2      -1.124508e-04\n",
      "              TNFRSF10A   6.370963e-05\n",
      "              TNFRSF10B  -1.037749e-06\n",
      "              TNFRSF10C   3.526077e-02\n",
      "              TNFRSF10D   6.416081e-01\n",
      "              TNKS        4.634458e-03\n",
      "              TRIM35     -6.188037e-04\n",
      "              TTI2        2.429804e-12\n",
      "              TUSC3      -5.164917e-01\n",
      "              UBXN8      -2.495104e-10\n",
      "              VPS37A      7.630699e-15\n",
      "              WRN         2.496005e-02\n",
      "              XPO7        1.099144e-01\n",
      "              ZDHHC2     -4.041749e-02\n",
      "              ZNF395     -6.084005e-05\n",
      "              ZNF703      1.783868e-02\n",
      "\n",
      "[276 rows x 1 columns]\n",
      "                                ov\n",
      "group         protein             \n",
      "has_event     ADAM28           NaN\n",
      "              ADAM9      -0.082996\n",
      "              ADAMDEC1   -0.062467\n",
      "              ADGRA2     -0.000016\n",
      "              AGPAT5     -0.948677\n",
      "              ANGPT2     -0.153037\n",
      "              ARHGEF10   -0.031525\n",
      "              ASAH1      -0.762484\n",
      "              ASH2L       0.082996\n",
      "              ATP6V1B2   -0.248665\n",
      "              BAG4        0.479870\n",
      "              BIN3       -0.000630\n",
      "              BMP1       -0.556295\n",
      "              BNIP3L           NaN\n",
      "              BRF2       -0.556295\n",
      "              CCAR2       0.492150\n",
      "              CCDC25     -0.218477\n",
      "              CDCA2       0.762484\n",
      "              CHMP7      -0.248665\n",
      "              CLDN23           NaN\n",
      "              CLN8             NaN\n",
      "              CLU        -0.000967\n",
      "              CNOT7       0.247453\n",
      "              CSGALNACT1       NaN\n",
      "              CTSB        0.799386\n",
      "              DCTN6      -0.058097\n",
      "              DDHD2       0.958274\n",
      "              DEFA3            NaN\n",
      "              DEFA4            NaN\n",
      "              DEFA5      -0.248665\n",
      "...                            ...\n",
      "not_has_event REEP4            NaN\n",
      "              SARAF            NaN\n",
      "              SCARA3           NaN\n",
      "              SCARA5           NaN\n",
      "              SFTPC            NaN\n",
      "              SH2D4A           NaN\n",
      "              SLC25A37         NaN\n",
      "              SLC39A14         NaN\n",
      "              SLC7A2           NaN\n",
      "              SORBS3           NaN\n",
      "              SOX7             NaN\n",
      "              STC1             NaN\n",
      "              TACC1            NaN\n",
      "              TDRP             NaN\n",
      "              TM2D2            NaN\n",
      "              TNFRSF10A        NaN\n",
      "              TNFRSF10B        NaN\n",
      "              TNFRSF10C        NaN\n",
      "              TNFRSF10D        NaN\n",
      "              TNKS             NaN\n",
      "              TRIM35           NaN\n",
      "              TTI2             NaN\n",
      "              TUSC3            NaN\n",
      "              UBXN8            NaN\n",
      "              VPS37A           NaN\n",
      "              WRN              NaN\n",
      "              XPO7             NaN\n",
      "              ZDHHC2           NaN\n",
      "              ZNF395           NaN\n",
      "              ZNF703           NaN\n",
      "\n",
      "[276 rows x 1 columns]\n",
      "                                  pdac\n",
      "group         protein                 \n",
      "has_event     ADAM28      1.438343e-01\n",
      "              ADAM9       8.170005e-03\n",
      "              ADAMDEC1   -9.317476e-01\n",
      "              ADGRA2      9.434787e-01\n",
      "              AGPAT5      8.739556e-01\n",
      "              ANGPT2      8.739556e-01\n",
      "              ARHGEF10    4.106330e-02\n",
      "              ASAH1      -7.892438e-01\n",
      "              ASH2L      -5.682814e-02\n",
      "              ATP6V1B2   -9.631442e-01\n",
      "              BAG4        9.317476e-01\n",
      "              BIN3        3.310319e-01\n",
      "              BMP1        3.829893e-02\n",
      "              BNIP3L               NaN\n",
      "              BRF2                 NaN\n",
      "              CCAR2      -8.170005e-03\n",
      "              CCDC25     -3.829893e-02\n",
      "              CDCA2                NaN\n",
      "              CHMP7      -8.026442e-03\n",
      "              CLDN23               NaN\n",
      "              CLN8                 NaN\n",
      "              CLU        -8.739556e-01\n",
      "              CNOT7      -7.671026e-01\n",
      "              CSGALNACT1           NaN\n",
      "              CTSB        1.364092e-01\n",
      "              DCTN6      -5.496833e-01\n",
      "              DDHD2       1.385188e-01\n",
      "              DEFA3       8.739556e-01\n",
      "              DEFA4       8.552587e-01\n",
      "              DEFA5      -8.739556e-01\n",
      "...                                ...\n",
      "not_has_event REEP4       5.350674e-02\n",
      "              SARAF       2.301080e-04\n",
      "              SCARA3               NaN\n",
      "              SCARA5     -7.311636e-04\n",
      "              SFTPC                NaN\n",
      "              SH2D4A     -6.439627e-01\n",
      "              SLC25A37             NaN\n",
      "              SLC39A14   -5.859271e-22\n",
      "              SLC7A2     -5.352141e-03\n",
      "              SORBS3      2.308404e-07\n",
      "              SOX7                 NaN\n",
      "              STC1        1.281139e-02\n",
      "              TACC1      -7.294756e-01\n",
      "              TDRP       -1.006793e-01\n",
      "              TM2D2                NaN\n",
      "              TNFRSF10A   1.036876e-02\n",
      "              TNFRSF10B   4.273402e-01\n",
      "              TNFRSF10C   2.242856e-02\n",
      "              TNFRSF10D            NaN\n",
      "              TNKS        1.172009e-01\n",
      "              TRIM35               NaN\n",
      "              TTI2       -1.624978e-01\n",
      "              TUSC3      -1.768311e-01\n",
      "              UBXN8                NaN\n",
      "              VPS37A      1.360058e-05\n",
      "              WRN        -3.831120e-03\n",
      "              XPO7        6.390166e-01\n",
      "              ZDHHC2      5.495313e-01\n",
      "              ZNF395     -8.041995e-04\n",
      "              ZNF703      2.561901e-07\n",
      "\n",
      "[276 rows x 1 columns]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>group</th>\n",
       "      <th>has_event</th>\n",
       "      <th>not_has_event</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>all</th>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coad</th>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hnscc</th>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lscc</th>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>luad</th>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ov</th>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pdac</th>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "group  has_event  not_has_event\n",
       "all            8              0\n",
       "coad           8              0\n",
       "hnscc          8              0\n",
       "lscc           8              0\n",
       "luad           8              0\n",
       "ov             8              0\n",
       "pdac           8              0"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def groups_overlaps(\n",
    "    chromosome,\n",
    "    arm,\n",
    "    gain_or_loss,\n",
    "    cis_or_trans,\n",
    "    proteomics_or_transcriptomics,\n",
    "    comparison,\n",
    "    source,\n",
    "    level=None,\n",
    "    data_dir=os.path.join(os.getcwd(), \"..\", \"data\"),\n",
    "):\n",
    "    \n",
    "    if comparison == \"has_event\":\n",
    "        comparison_name = \"has_vs_not_has_event\"\n",
    "        groups = [\"tumor\", \"normal\"]\n",
    "        \n",
    "    elif comparison == \"tumor\":\n",
    "        comparison_name = \"tumor_vs_normal\"\n",
    "        groups = [\"has_event\", \"not_has_event\"]\n",
    "\n",
    "    all_ttest = pd.DataFrame()\n",
    "    for group in groups:\n",
    "        \n",
    "        group_path = get_ttest_results_path(\n",
    "            data_dir=data_dir,\n",
    "            source=source,\n",
    "            level=level,\n",
    "            chromosome=chromosome,\n",
    "            arm=arm,\n",
    "            gain_or_loss=gain_or_loss,\n",
    "            cis_or_trans=cis_or_trans,\n",
    "            proteomics_or_transcriptomics=proteomics_or_transcriptomics,\n",
    "            group=group,\n",
    "            comparison_name=comparison_name,\n",
    "        )\n",
    "        \n",
    "        group_ttest = pd.\\\n",
    "        read_csv(group_path, sep=\"\\t\").\\\n",
    "        dropna(subset=\"adj_p\")\n",
    "        \n",
    "        # We're going to combine adj_p and change for isoforms, weighted by sample\n",
    "        # size (n) for each isoform\n",
    "        \n",
    "        # 1. Calculate sample sizes for each isoform\n",
    "        group_ttest = group_ttest.assign(\n",
    "            n=group_ttest[\"tumor_sample_size\"] + group_ttest[\"not_tumor_sample_size\"],\n",
    "        ).\\\n",
    "        drop(columns=[\"tumor_sample_size\", \"not_tumor_sample_size\"])\n",
    "        \n",
    "        # 2. Calculate the overall sample size for all isoforms in each protein\n",
    "        group_ns = group_ttest.\\\n",
    "        groupby([\"cancer_type\", \"protein\"]).\\\n",
    "        apply(lambda x: x[\"n\"].sum()).\\\n",
    "        rename(\"group_n\")\n",
    "        \n",
    "        group_ttest = group_ttest.merge(\n",
    "            right=group_ns,\n",
    "            on=[\"protein\", \"cancer_type\"],\n",
    "            how=\"outer\",\n",
    "        )\n",
    "        \n",
    "        # 3. Calculate weights by dividing the sample size for each isoform by the \n",
    "        # overall sample size for the protein. Note that in this way, the weights within\n",
    "        # each protein sum to 1.\n",
    "        group_ttest = group_ttest.assign(weight=group_ttest[\"n\"] / group_ttest[\"group_n\"])\n",
    "        \n",
    "        # 4. Multiply the weights by the adj_p and change columns\n",
    "        group_ttest = group_ttest.assign(\n",
    "            adj_p=group_ttest[\"adj_p\"] * group_ttest[\"weight\"],\n",
    "            change=group_ttest[\"change\"] * group_ttest[\"weight\"],\n",
    "        )\n",
    "        \n",
    "        # 5. Sum the weighted adj_p and change columns to get the mean for isoforms\n",
    "        # Proteins with only one form will be unaffected \n",
    "        # Because the weights were normalized to sum to 1 within each protein, we\n",
    "        # don't have to divide the sums by the number of isoforms per protein\n",
    "        group_ttest = group_ttest.\\\n",
    "        groupby([\"protein\", \"cancer_type\"])[[\"adj_p\", \"change\"]].\\\n",
    "        sum().\\\n",
    "        reset_index(drop=False)\n",
    "        \n",
    "        # To capture whether a protein is up- or down-regulated, multiply p\n",
    "        # values by the sign of the associated change\n",
    "        group_ttest = group_ttest.assign(\n",
    "            sign_adj_p=group_ttest[\"adj_p\"] * np.sign(group_ttest[\"change\"])\n",
    "        )\n",
    "\n",
    "        # Pivot the table to have one row per protein and one column per cancer\n",
    "        group_ttest = group_ttest.pivot(\n",
    "            index=\"protein\",\n",
    "            columns=\"cancer_type\",\n",
    "            values=\"sign_adj_p\",\n",
    "        ).\\\n",
    "        reset_index(drop=False).\\\n",
    "        assign(group=group)\n",
    "        \n",
    "        group_ttest.columns.name = None\n",
    "        \n",
    "        all_ttest = pd.concat([all_ttest, group_ttest])\n",
    "        \n",
    "    all_ttest = all_ttest.\\\n",
    "    reset_index(drop=True).\\\n",
    "    set_index([\"group\", \"protein\"])\n",
    "    \n",
    "    # Count how many proteins only show up in has_event or not_has_event, across\n",
    "    # all cancer types aggregated together\n",
    "    all_sig_count = (\n",
    "        (all_ttest <= SIG_CUTOFF) &\n",
    "        (all_ttest >= -SIG_CUTOFF)\n",
    "    ).\\\n",
    "    sum(axis=1).\\\n",
    "    rename(\"sig_count\").\\\n",
    "    reset_index(drop=False).\\\n",
    "    pivot(\n",
    "        index=\"protein\",\n",
    "        columns=\"group\",\n",
    "        values=\"sig_count\",\n",
    "    ).\\\n",
    "    isna().\\\n",
    "    sum().\\\n",
    "    rename(\"all\")\n",
    "    \n",
    "    all_sig_props = all_sig_count #/ all_ttest.shape[0]\n",
    "    all_sig_props = all_sig_props.to_frame().transpose()\n",
    "    \n",
    "    # Do that for each cancer type\n",
    "    for cancer_type in all_ttest.columns:\n",
    "        \n",
    "        cancer_ttest = all_ttest[[cancer_type]]\n",
    "        \n",
    "        cancer_sig_count = (\n",
    "            (cancer_ttest <= SIG_CUTOFF) &\n",
    "            (cancer_ttest >= -SIG_CUTOFF)\n",
    "        ).\\\n",
    "        dropna().\\\n",
    "        reset_index(drop=False).\\\n",
    "        pivot(\n",
    "            index=\"protein\",\n",
    "            columns=\"group\",\n",
    "            values=cancer_type,\n",
    "        ).\\\n",
    "        isna().\\\n",
    "        sum().\\\n",
    "        rename(cancer_type)\n",
    "        \n",
    "        cancer_sig_prop = cancer_sig_count# / cancer_ttest.shape[0]\n",
    "        \n",
    "        all_sig_props = pd.concat([all_sig_props, cancer_sig_prop.to_frame().transpose()])\n",
    "    \n",
    "    return all_sig_props\n",
    "\n",
    "res = groups_overlaps(\n",
    "    chromosome=8,\n",
    "    arm=\"p\",\n",
    "    gain_or_loss=\"loss\",\n",
    "    cis_or_trans=\"cis\",\n",
    "    proteomics_or_transcriptomics=\"proteomics\",\n",
    "    comparison=\"tumor\",\n",
    "    source=\"gistic\",\n",
    "    level=\"gene\",\n",
    ")\n",
    "\n",
    "res"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
