{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e5ca066d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import cnvutils\n",
    "\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "from cnvutils.filenames import (\n",
    "    get_ttest_results_path,\n",
    ")\n",
    "\n",
    "from cnvutils.constants import (\n",
    "    SIG_CUTOFF,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "e0db8f7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>group</th>\n",
       "      <th>has_event</th>\n",
       "      <th>not_has_event</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>all</th>\n",
       "      <td>0.028986</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coad</th>\n",
       "      <td>0.028986</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hnscc</th>\n",
       "      <td>0.028986</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lscc</th>\n",
       "      <td>0.028986</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>luad</th>\n",
       "      <td>0.028986</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ov</th>\n",
       "      <td>0.028986</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pdac</th>\n",
       "      <td>0.028986</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "group  has_event  not_has_event\n",
       "all     0.028986            0.0\n",
       "coad    0.028986            0.0\n",
       "hnscc   0.028986            0.0\n",
       "lscc    0.028986            0.0\n",
       "luad    0.028986            0.0\n",
       "ov      0.028986            0.0\n",
       "pdac    0.028986            0.0"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def groups_overlaps(\n",
    "    chromosome,\n",
    "    arm,\n",
    "    gain_or_loss,\n",
    "    cis_or_trans,\n",
    "    proteomics_or_transcriptomics,\n",
    "    comparison,\n",
    "    source,\n",
    "    level=None,\n",
    "    data_dir=os.path.join(os.getcwd(), \"..\", \"data\"),\n",
    "):\n",
    "    \n",
    "    if comparison == \"has_event\":\n",
    "        comparison_name = \"has_vs_not_has_event\"\n",
    "        groups = [\"tumor\", \"normal\"]\n",
    "        \n",
    "    elif comparison == \"tumor\":\n",
    "        comparison_name = \"tumor_vs_normal\"\n",
    "        groups = [\"has_event\", \"not_has_event\"]\n",
    "\n",
    "    all_ttest = pd.DataFrame()\n",
    "    for group in groups:\n",
    "        \n",
    "        group_path = get_ttest_results_path(\n",
    "            data_dir=data_dir,\n",
    "            source=source,\n",
    "            level=level,\n",
    "            chromosome=chromosome,\n",
    "            arm=arm,\n",
    "            gain_or_loss=gain_or_loss,\n",
    "            cis_or_trans=cis_or_trans,\n",
    "            proteomics_or_transcriptomics=proteomics_or_transcriptomics,\n",
    "            group=group,\n",
    "            comparison_name=comparison_name,\n",
    "        )\n",
    "        \n",
    "        group_ttest = pd.\\\n",
    "        read_csv(group_path, sep=\"\\t\").\\\n",
    "        dropna(subset=\"adj_p\")\n",
    "        \n",
    "        # We're going to combine adj_p and change for isoforms, weighted by sample\n",
    "        # size (n) for each isoform\n",
    "        \n",
    "        # 1. Calculate sample sizes for each isoform\n",
    "        group_ttest = group_ttest.assign(\n",
    "            n=group_ttest[\"tumor_sample_size\"] + group_ttest[\"not_tumor_sample_size\"],\n",
    "        ).\\\n",
    "        drop(columns=[\"tumor_sample_size\", \"not_tumor_sample_size\"])\n",
    "        \n",
    "        # 2. Calculate the overall sample size for all isoforms in each protein\n",
    "        group_ns = group_ttest.\\\n",
    "        groupby([\"cancer_type\", \"protein\"]).\\\n",
    "        apply(lambda x: x[\"n\"].sum()).\\\n",
    "        rename(\"group_n\")\n",
    "        \n",
    "        group_ttest = group_ttest.merge(\n",
    "            right=group_ns,\n",
    "            on=[\"protein\", \"cancer_type\"],\n",
    "            how=\"outer\",\n",
    "        )\n",
    "        \n",
    "        # 3. Calculate weights by dividing the sample size for each isoform by the \n",
    "        # overall sample size for the protein. Note that in this way, the weights within\n",
    "        # each protein sum to 1.\n",
    "        group_ttest = group_ttest.assign(weight=group_ttest[\"n\"] / group_ttest[\"group_n\"])\n",
    "        \n",
    "        # 4. Multiply the weights by the adj_p and change columns\n",
    "        group_ttest = group_ttest.assign(\n",
    "            adj_p=group_ttest[\"adj_p\"] * group_ttest[\"weight\"],\n",
    "            change=group_ttest[\"change\"] * group_ttest[\"weight\"],\n",
    "        )\n",
    "        \n",
    "        # 5. Sum the weighted adj_p and change columns to get the mean for isoforms\n",
    "        # Proteins with only one form will be unaffected \n",
    "        # Because the weights were normalized to sum to 1 within each protein, we\n",
    "        # don't have to divide the sums by the number of isoforms per protein\n",
    "        group_ttest = group_ttest.\\\n",
    "        groupby([\"protein\", \"cancer_type\"])[[\"adj_p\", \"change\"]].\\\n",
    "        sum().\\\n",
    "        reset_index(drop=False)\n",
    "        \n",
    "        # To capture whether a protein is up- or down-regulated, multiply p\n",
    "        # values by the sign of the associated change\n",
    "        group_ttest = group_ttest.assign(\n",
    "            sign_adj_p=group_ttest[\"adj_p\"] * np.sign(group_ttest[\"change\"])\n",
    "        )\n",
    "\n",
    "        # Pivot the table to have one row per protein and one column per cancer\n",
    "        group_ttest = group_ttest.pivot(\n",
    "            index=\"protein\",\n",
    "            columns=\"cancer_type\",\n",
    "            values=\"sign_adj_p\",\n",
    "        ).\\\n",
    "        reset_index(drop=False).\\\n",
    "        assign(group=group)\n",
    "        \n",
    "        group_ttest.columns.name = None\n",
    "        \n",
    "        all_ttest = pd.concat([all_ttest, group_ttest])\n",
    "        \n",
    "    all_ttest = all_ttest.\\\n",
    "    reset_index(drop=True).\\\n",
    "    set_index([\"group\", \"protein\"])\n",
    "    \n",
    "    # Count how many proteins only show up in has_event or not_has_event, across\n",
    "    # all cancer types aggregated together\n",
    "    all_sig_count = (\n",
    "        (all_ttest <= SIG_CUTOFF) &\n",
    "        (all_ttest >= -SIG_CUTOFF)\n",
    "    ).\\\n",
    "    sum(axis=1).\\\n",
    "    rename(\"sig_count\").\\\n",
    "    reset_index(drop=False).\\\n",
    "    pivot(\n",
    "        index=\"protein\",\n",
    "        columns=\"group\",\n",
    "        values=\"sig_count\",\n",
    "    ).\\\n",
    "    isna().\\\n",
    "    sum().\\\n",
    "    rename(\"all\")\n",
    "    \n",
    "    all_sig_props = (all_sig_count / all_ttest.shape[0]).to_frame().transpose()\n",
    "    \n",
    "    # Do that for each cancer type\n",
    "    for cancer_type in all_ttest.columns:\n",
    "        \n",
    "        cancer_ttest = all_ttest[[cancer_type]]\n",
    "        \n",
    "        cancer_sig_count = (\n",
    "            (cancer_ttest <= SIG_CUTOFF) &\n",
    "            (cancer_ttest >= -SIG_CUTOFF)\n",
    "        ).\\\n",
    "        dropna().\\\n",
    "        reset_index(drop=False).\\\n",
    "        pivot(\n",
    "            index=\"protein\",\n",
    "            columns=\"group\",\n",
    "            values=cancer_type,\n",
    "        ).\\\n",
    "        isna().\\\n",
    "        sum().\\\n",
    "        rename(cancer_type)\n",
    "        \n",
    "        cancer_sig_prop = cancer_sig_count / cancer_ttest.shape[0]\n",
    "        \n",
    "        all_sig_props = pd.concat([all_sig_props, cancer_sig_prop.to_frame().transpose()])\n",
    "    \n",
    "    return all_sig_props\n",
    "\n",
    "res = groups_overlaps(\n",
    "    chromosome=8,\n",
    "    arm=\"p\",\n",
    "    gain_or_loss=\"loss\",\n",
    "    cis_or_trans=\"cis\",\n",
    "    proteomics_or_transcriptomics=\"proteomics\",\n",
    "    comparison=\"tumor\",\n",
    "    source=\"gistic\",\n",
    "    level=\"gene\",\n",
    ")\n",
    "\n",
    "res"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
