{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e5adc167",
   "metadata": {},
   "source": [
    "# Number of proteins different between tumor and normal in samples with an event versus samples without it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2ff8d605",
   "metadata": {},
   "outputs": [],
   "source": [
    "import altair as alt\n",
    "import cnvutils\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3f82134d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': True, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': True, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': True, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': True, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': False, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': False, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': False, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': False, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': True, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': True, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': True, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': True, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': False, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': False, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': False, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': False, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': True, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': True, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': True, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': True, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': False, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': False, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': False, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': False, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': True, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': True, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': True, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': True, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': False, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': False, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': False, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': False, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "all_pvals = cnvutils.get_has_vs_not_has_tumor_normal_diff_props(\n",
    "    chromosomes_events={\n",
    "            8: {\n",
    "                \"p\": [\"loss\"],\n",
    "                \"q\": [\"gain\"],\n",
    "            },\n",
    "        },\n",
    "    sources=[\"cptac\", \"gistic\"],\n",
    "    levels=[\"gene\"],    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0096ec0b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>sig_props</th>\n",
       "      <th>not_sig_props</th>\n",
       "      <th>na_props</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>p_all</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.07979</td>\n",
       "      <td>0.013382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>p_cptac</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.677334</td>\n",
       "      <td>0.010166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>p_cptac_proteomics</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.25101</td>\n",
       "      <td>0.010693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>p_cptac_proteomics_cis</td>\n",
       "      <td>0.000705</td>\n",
       "      <td>0.384116</td>\n",
       "      <td>0.084441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>p_cptac_proteomics_cis_loss</td>\n",
       "      <td>0.053797</td>\n",
       "      <td>0.634489</td>\n",
       "      <td>0.261586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>p_cptac_proteomics_cis_gain</td>\n",
       "      <td>0.009951</td>\n",
       "      <td>0.520527</td>\n",
       "      <td>0.263331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>p_cptac_proteomics_trans</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>0.490871</td>\n",
       "      <td>0.081461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>p_cptac_proteomics_trans_loss</td>\n",
       "      <td>0.005651</td>\n",
       "      <td>0.580663</td>\n",
       "      <td>0.26475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>p_cptac_proteomics_trans_gain</td>\n",
       "      <td>0.002257</td>\n",
       "      <td>0.741134</td>\n",
       "      <td>0.257374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>p_cptac_transcriptomics</td>\n",
       "      <td>0.00187</td>\n",
       "      <td>0.004282</td>\n",
       "      <td>0.008568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>p_cptac_transcriptomics_cis</td>\n",
       "      <td>0.011739</td>\n",
       "      <td>0.115545</td>\n",
       "      <td>0.133975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>p_cptac_transcriptomics_cis_loss</td>\n",
       "      <td>0.166312</td>\n",
       "      <td>0.731538</td>\n",
       "      <td>0.148743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>p_cptac_transcriptomics_cis_gain</td>\n",
       "      <td>0.08552</td>\n",
       "      <td>0.089962</td>\n",
       "      <td>0.741801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>p_cptac_transcriptomics_trans</td>\n",
       "      <td>0.003641</td>\n",
       "      <td>0.0043</td>\n",
       "      <td>0.000569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>p_cptac_transcriptomics_trans_loss</td>\n",
       "      <td>0.148229</td>\n",
       "      <td>0.157463</td>\n",
       "      <td>0.041323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>p_cptac_transcriptomics_trans_gain</td>\n",
       "      <td>0.01312</td>\n",
       "      <td>0.016185</td>\n",
       "      <td>0.027913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>p_gistic</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.007482</td>\n",
       "      <td>0.338007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>p_gistic_proteomics</td>\n",
       "      <td>0.001423</td>\n",
       "      <td>0.197723</td>\n",
       "      <td>0.40486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>p_gistic_proteomics_cis</td>\n",
       "      <td>0.031795</td>\n",
       "      <td>0.44716</td>\n",
       "      <td>0.534534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>p_gistic_proteomics_cis_loss</td>\n",
       "      <td>0.202007</td>\n",
       "      <td>0.095488</td>\n",
       "      <td>0.663887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>p_gistic_proteomics_cis_gain</td>\n",
       "      <td>0.107454</td>\n",
       "      <td>0.690525</td>\n",
       "      <td>0.214804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>p_gistic_proteomics_trans</td>\n",
       "      <td>0.026682</td>\n",
       "      <td>0.313658</td>\n",
       "      <td>0.6021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>p_gistic_proteomics_trans_loss</td>\n",
       "      <td>0.283101</td>\n",
       "      <td>0.088707</td>\n",
       "      <td>0.641287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>p_gistic_proteomics_trans_gain</td>\n",
       "      <td>0.041409</td>\n",
       "      <td>0.87038</td>\n",
       "      <td>0.201411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>p_gistic_transcriptomics</td>\n",
       "      <td>0.001729</td>\n",
       "      <td>0.002273</td>\n",
       "      <td>0.021103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>p_gistic_transcriptomics_cis</td>\n",
       "      <td>0.078257</td>\n",
       "      <td>0.096074</td>\n",
       "      <td>0.15245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>p_gistic_transcriptomics_cis_loss</td>\n",
       "      <td>0.172504</td>\n",
       "      <td>0.19036</td>\n",
       "      <td>0.156424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>p_gistic_transcriptomics_cis_gain</td>\n",
       "      <td>0.318915</td>\n",
       "      <td>0.313619</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>p_gistic_transcriptomics_trans</td>\n",
       "      <td>0.013464</td>\n",
       "      <td>0.013817</td>\n",
       "      <td>0.009503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>p_gistic_transcriptomics_trans_loss</td>\n",
       "      <td>0.14649</td>\n",
       "      <td>0.147898</td>\n",
       "      <td>0.122363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>p_gistic_transcriptomics_trans_gain</td>\n",
       "      <td>0.097302</td>\n",
       "      <td>0.098836</td>\n",
       "      <td>0.079526</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   name sig_props not_sig_props  na_props\n",
       "0                                 p_all       0.0       0.07979  0.013382\n",
       "1                               p_cptac       0.0      0.677334  0.010166\n",
       "2                    p_cptac_proteomics       0.0       0.25101  0.010693\n",
       "3                p_cptac_proteomics_cis  0.000705      0.384116  0.084441\n",
       "4           p_cptac_proteomics_cis_loss  0.053797      0.634489  0.261586\n",
       "5           p_cptac_proteomics_cis_gain  0.009951      0.520527  0.263331\n",
       "6              p_cptac_proteomics_trans  0.000019      0.490871  0.081461\n",
       "7         p_cptac_proteomics_trans_loss  0.005651      0.580663   0.26475\n",
       "8         p_cptac_proteomics_trans_gain  0.002257      0.741134  0.257374\n",
       "9               p_cptac_transcriptomics   0.00187      0.004282  0.008568\n",
       "10          p_cptac_transcriptomics_cis  0.011739      0.115545  0.133975\n",
       "11     p_cptac_transcriptomics_cis_loss  0.166312      0.731538  0.148743\n",
       "12     p_cptac_transcriptomics_cis_gain   0.08552      0.089962  0.741801\n",
       "13        p_cptac_transcriptomics_trans  0.003641        0.0043  0.000569\n",
       "14   p_cptac_transcriptomics_trans_loss  0.148229      0.157463  0.041323\n",
       "15   p_cptac_transcriptomics_trans_gain   0.01312      0.016185  0.027913\n",
       "16                             p_gistic  0.000006      0.007482  0.338007\n",
       "17                  p_gistic_proteomics  0.001423      0.197723   0.40486\n",
       "18              p_gistic_proteomics_cis  0.031795       0.44716  0.534534\n",
       "19         p_gistic_proteomics_cis_loss  0.202007      0.095488  0.663887\n",
       "20         p_gistic_proteomics_cis_gain  0.107454      0.690525  0.214804\n",
       "21            p_gistic_proteomics_trans  0.026682      0.313658    0.6021\n",
       "22       p_gistic_proteomics_trans_loss  0.283101      0.088707  0.641287\n",
       "23       p_gistic_proteomics_trans_gain  0.041409       0.87038  0.201411\n",
       "24             p_gistic_transcriptomics  0.001729      0.002273  0.021103\n",
       "25         p_gistic_transcriptomics_cis  0.078257      0.096074   0.15245\n",
       "26    p_gistic_transcriptomics_cis_loss  0.172504       0.19036  0.156424\n",
       "27    p_gistic_transcriptomics_cis_gain  0.318915      0.313619       1.0\n",
       "28       p_gistic_transcriptomics_trans  0.013464      0.013817  0.009503\n",
       "29  p_gistic_transcriptomics_trans_loss   0.14649      0.147898  0.122363\n",
       "30  p_gistic_transcriptomics_trans_gain  0.097302      0.098836  0.079526"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_pvals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5de6ffee",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-ea102105ed164b9b8fbdd0ed2a86c3ed\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-ea102105ed164b9b8fbdd0ed2a86c3ed\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-ea102105ed164b9b8fbdd0ed2a86c3ed\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.17.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"4.17.0\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}}, \"layer\": [{\"mark\": \"point\", \"encoding\": {\"color\": {\"field\": \"group\", \"type\": \"nominal\"}, \"x\": {\"field\": \"name\", \"type\": \"nominal\"}, \"y\": {\"field\": \"neg_log_p\", \"type\": \"quantitative\"}}}, {\"mark\": {\"type\": \"rule\", \"color\": \"crimson\"}, \"encoding\": {\"y\": {\"field\": \"line\", \"type\": \"quantitative\"}}}], \"data\": {\"name\": \"data-c6985df01219acf5fcbb7279ffdbd68d\"}, \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.17.0.json\", \"datasets\": {\"data-c6985df01219acf5fcbb7279ffdbd68d\": [{\"name\": \"p_all\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 10.471004717467924}, {\"name\": \"p_cptac\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 9.569432407509359}, {\"name\": \"p_cptac_proteomics\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 7.469666600267349}, {\"name\": \"p_cptac_proteomics_cis\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 3.1517056660670297}, {\"name\": \"p_cptac_proteomics_cis_loss\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 1.2692442970422555}, {\"name\": \"p_cptac_proteomics_cis_gain\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 2.002133486778289}, {\"name\": \"p_cptac_proteomics_trans\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 4.7137805685342276}, {\"name\": \"p_cptac_proteomics_trans_loss\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 2.2479015323958484}, {\"name\": \"p_cptac_proteomics_trans_gain\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 2.6464763832179474}, {\"name\": \"p_cptac_transcriptomics\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 2.7281207312476}, {\"name\": \"p_cptac_transcriptomics_cis\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 1.930383673296208}, {\"name\": \"p_cptac_transcriptomics_cis_loss\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 0.7790760686666216}, {\"name\": \"p_cptac_transcriptomics_cis_gain\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 1.067930944928729}, {\"name\": \"p_cptac_transcriptomics_trans\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 2.43883536426692}, {\"name\": \"p_cptac_transcriptomics_trans_loss\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 0.8290678153215344}, {\"name\": \"p_cptac_transcriptomics_trans_gain\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 1.8820808864955598}, {\"name\": \"p_gistic\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 5.2468578469960905}, {\"name\": \"p_gistic_proteomics\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 2.846876826760932}, {\"name\": \"p_gistic_proteomics_cis\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 1.497647760686271}, {\"name\": \"p_gistic_proteomics_cis_loss\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 0.6946329932676517}, {\"name\": \"p_gistic_proteomics_cis_gain\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 0.9687778680231262}, {\"name\": \"p_gistic_proteomics_trans\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 1.573777553605981}, {\"name\": \"p_gistic_proteomics_trans_loss\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 0.5480588303406492}, {\"name\": \"p_gistic_proteomics_trans_gain\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 1.3829078796849927}, {\"name\": \"p_gistic_transcriptomics\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 2.762087077459119}, {\"name\": \"p_gistic_transcriptomics_cis\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 1.1064761435456574}, {\"name\": \"p_gistic_transcriptomics_cis_loss\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 0.7632008576447693}, {\"name\": \"p_gistic_transcriptomics_cis_gain\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 0.4963246519126221}, {\"name\": \"p_gistic_transcriptomics_trans\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 1.8708191126788378}, {\"name\": \"p_gistic_transcriptomics_trans_loss\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 0.8341916491718997}, {\"name\": \"p_gistic_transcriptomics_trans_gain\", \"line\": 1.3010299956639813, \"group\": \"neg_log_sig_props\", \"neg_log_p\": 1.0118799962828637}, {\"name\": \"p_all\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 1.098051428219888}, {\"name\": \"p_cptac\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 0.1691970416143015}, {\"name\": \"p_cptac_proteomics\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 0.6003096349400199}, {\"name\": \"p_cptac_proteomics_cis\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 0.41553789194615554}, {\"name\": \"p_cptac_proteomics_cis_loss\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 0.19757572557457487}, {\"name\": \"p_cptac_proteomics_cis_gain\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 0.28355682058073506}, {\"name\": \"p_cptac_proteomics_trans\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 0.3090329705730096}, {\"name\": \"p_cptac_proteomics_trans_loss\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 0.23607587203798885}, {\"name\": \"p_cptac_proteomics_trans_gain\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 0.13010334549936128}, {\"name\": \"p_cptac_transcriptomics\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 2.368392448033574}, {\"name\": \"p_cptac_transcriptomics_cis\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 0.937248812435206}, {\"name\": \"p_cptac_transcriptomics_cis_loss\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 0.13576335995331007}, {\"name\": \"p_cptac_transcriptomics_cis_gain\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 1.0459405072108496}, {\"name\": \"p_cptac_transcriptomics_trans\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 2.3665325278097376}, {\"name\": \"p_cptac_transcriptomics_trans_loss\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 0.8028208501971393}, {\"name\": \"p_cptac_transcriptomics_trans_gain\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 1.790895736180416}, {\"name\": \"p_gistic\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 2.1259863187859764}, {\"name\": \"p_gistic_proteomics\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 0.7039435188240308}, {\"name\": \"p_gistic_proteomics_cis\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 0.3495371867760947}, {\"name\": \"p_gistic_proteomics_cis_loss\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 1.020050450110881}, {\"name\": \"p_gistic_proteomics_cis_gain\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 0.16082090542909316}, {\"name\": \"p_gistic_proteomics_trans\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 0.5035438258765439}, {\"name\": \"p_gistic_proteomics_trans_loss\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 1.052040133171195}, {\"name\": \"p_gistic_proteomics_trans_gain\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 0.060290861616182925}, {\"name\": \"p_gistic_transcriptomics\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 2.6434616763390513}, {\"name\": \"p_gistic_transcriptomics_cis\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 1.017392177701218}, {\"name\": \"p_gistic_transcriptomics_cis_loss\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 0.7204254029076912}, {\"name\": \"p_gistic_transcriptomics_cis_gain\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 0.5035974972167871}, {\"name\": \"p_gistic_transcriptomics_trans\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 1.8595770123415416}, {\"name\": \"p_gistic_transcriptomics_trans_loss\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 0.8300365121062859}, {\"name\": \"p_gistic_transcriptomics_trans_gain\", \"line\": 1.3010299956639813, \"group\": \"neg_log_not_sig_props\", \"neg_log_p\": 1.005084898268741}, {\"name\": \"p_all\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 1.873493116825204}, {\"name\": \"p_cptac\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 1.9928306136077498}, {\"name\": \"p_cptac_proteomics\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 1.9709034890595347}, {\"name\": \"p_cptac_proteomics_cis\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 1.0734455126628346}, {\"name\": \"p_cptac_proteomics_cis_loss\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 0.5823847169099294}, {\"name\": \"p_cptac_proteomics_cis_gain\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 0.5794981491051928}, {\"name\": \"p_cptac_proteomics_trans\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 1.0890512107879873}, {\"name\": \"p_cptac_proteomics_trans_loss\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 0.577164630431442}, {\"name\": \"p_cptac_proteomics_trans_gain\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 0.5894361019803118}, {\"name\": \"p_cptac_transcriptomics\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 2.0671086708702173}, {\"name\": \"p_cptac_transcriptomics_cis\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 0.8729761512128511}, {\"name\": \"p_cptac_transcriptomics_cis_loss\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 0.8275620933960598}, {\"name\": \"p_cptac_transcriptomics_cis_gain\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 0.129712520998888}, {\"name\": \"p_cptac_transcriptomics_trans\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 3.244863781936298}, {\"name\": \"p_cptac_transcriptomics_trans_loss\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 1.38380674089758}, {\"name\": \"p_cptac_transcriptomics_trans_gain\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 1.5541973977528514}, {\"name\": \"p_gistic\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 0.4710745736238774}, {\"name\": \"p_gistic_proteomics\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 0.39269517499379347}, {\"name\": \"p_gistic_proteomics_cis\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 0.2720250005800358}, {\"name\": \"p_gistic_proteomics_cis_loss\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 0.1779059106174358}, {\"name\": \"p_gistic_proteomics_cis_gain\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 0.6679578616577735}, {\"name\": \"p_gistic_proteomics_trans\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 0.22033173000710887}, {\"name\": \"p_gistic_proteomics_trans_loss\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 0.19294773714885766}, {\"name\": \"p_gistic_proteomics_trans_gain\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 0.6959171603275001}, {\"name\": \"p_gistic_transcriptomics\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 1.675656370409353}, {\"name\": \"p_gistic_transcriptomics_cis\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 0.8168716818979999}, {\"name\": \"p_gistic_transcriptomics_cis_loss\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 0.8056961165248897}, {\"name\": \"p_gistic_transcriptomics_cis_gain\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": -0.0}, {\"name\": \"p_gistic_transcriptomics_trans\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 2.022132539506821}, {\"name\": \"p_gistic_transcriptomics_trans_loss\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 0.9123483632516947}, {\"name\": \"p_gistic_transcriptomics_trans_gain\", \"line\": 1.3010299956639813, \"group\": \"neg_log_na_props\", \"neg_log_p\": 1.0994890738252798}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.LayerChart(...)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_pvals = all_pvals.infer_objects()\n",
    "\n",
    "all_pvals = all_pvals.assign(\n",
    "    line=-np.log10(0.05),\n",
    "    neg_log_sig_props=-np.log10(all_pvals[\"sig_props\"]),\n",
    "    neg_log_not_sig_props=-np.log10(all_pvals[\"not_sig_props\"]),\n",
    "    neg_log_na_props=-np.log10(all_pvals[\"na_props\"]),\n",
    ")\n",
    "\n",
    "all_pvals = all_pvals.drop(columns=[\"sig_props\", \"not_sig_props\", \"na_props\"])\n",
    "all_pvals = all_pvals.melt(\n",
    "    id_vars=[\"name\", \"line\"],\n",
    "    value_vars=[\"neg_log_sig_props\", \"neg_log_not_sig_props\", \"neg_log_na_props\"],\n",
    "    var_name=\"group\",\n",
    "    value_name=\"neg_log_p\",\n",
    ")\n",
    "\n",
    "base = alt.Chart(all_pvals)\n",
    "\n",
    "dots = base.mark_point().encode(\n",
    "    x=\"name\",\n",
    "    y=\"neg_log_p\",\n",
    "    color=\"group\",\n",
    ")\n",
    "\n",
    "line = base.mark_rule(color=\"crimson\").encode(\n",
    "    y=\"line\"\n",
    ")\n",
    "\n",
    "dots + line"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
