{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e5adc167",
   "metadata": {},
   "source": [
    "# Number of proteins different between tumor and normal in samples with an event versus samples without it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2ff8d605",
   "metadata": {},
   "outputs": [],
   "source": [
    "import altair as alt\n",
    "import cnvutils\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3f82134d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': True, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': True, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': False, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': True, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': False, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': False, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': False, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': True, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': True, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': True, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': True, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': True, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': False, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': False, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': False, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov'], 'source': 'cptac', 'has_event': False, 'level': None, 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': True, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': True, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': True, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': False, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': True, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': False, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': False, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'p', 'gain_or_loss': 'loss', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': False, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': True, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': True, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': True, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': False, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': True, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'proteomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': False, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'cis', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': False, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n",
      "Running _get_ttest_sig_counts with {'chromosome': 8, 'arm': 'q', 'gain_or_loss': 'gain', 'cis_or_trans': 'trans', 'proteomics_or_transcriptomics': 'transcriptomics', 'cancer_types': ['brca', 'coad', 'hnscc', 'lscc', 'luad', 'ov', 'pdac'], 'source': 'gistic', 'has_event': False, 'level': 'gene', 'data_dir': '/home/caleb/github/PayneLab/pancancerCNV/00_functions/chr08/../data'}...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "all_pvals = cnvutils.get_has_vs_not_has_tumor_normal_diff_props(\n",
    "    chromosomes_events={\n",
    "            8: {\n",
    "                \"p\": [\"loss\"],\n",
    "                \"q\": [\"gain\"],\n",
    "            },\n",
    "        },\n",
    "    sources=[\"cptac\", \"gistic\"],\n",
    "    levels=[\"gene\"],    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5de6ffee",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>line</th>\n",
       "      <th>group</th>\n",
       "      <th>neg_log_p</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>p_all</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>10.471005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>p_cptac</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>9.569432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>p_cptac_proteomics</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>7.469667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>p_cptac_proteomics_cis</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>3.151706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>p_cptac_proteomics_cis_loss</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>1.269244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>p_cptac_proteomics_cis_gain</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>2.002133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>p_cptac_proteomics_trans</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>4.713781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>p_cptac_proteomics_trans_loss</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>2.247902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>p_cptac_proteomics_trans_gain</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>2.646476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>p_cptac_transcriptomics</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>2.728121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>p_cptac_transcriptomics_cis</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>1.930384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>p_cptac_transcriptomics_cis_loss</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>0.779076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>p_cptac_transcriptomics_cis_gain</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>1.067931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>p_cptac_transcriptomics_trans</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>2.438835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>p_cptac_transcriptomics_trans_loss</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>0.829068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>p_cptac_transcriptomics_trans_gain</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>1.882081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>p_gistic</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>5.246858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>p_gistic_proteomics</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>2.846877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>p_gistic_proteomics_cis</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>1.497648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>p_gistic_proteomics_cis_loss</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>0.694633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>p_gistic_proteomics_cis_gain</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>0.968778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>p_gistic_proteomics_trans</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>1.573778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>p_gistic_proteomics_trans_loss</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>0.548059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>p_gistic_proteomics_trans_gain</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>1.382908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>p_gistic_transcriptomics</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>2.762087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>p_gistic_transcriptomics_cis</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>1.106476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>p_gistic_transcriptomics_cis_loss</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>0.763201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>p_gistic_transcriptomics_cis_gain</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>0.496325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>p_gistic_transcriptomics_trans</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>1.870819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>p_gistic_transcriptomics_trans_loss</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_sig_props</td>\n",
       "      <td>0.834192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>p_cptac</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>1.992831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>p_cptac_proteomics</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>1.970903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>p_cptac_proteomics_cis</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>1.073446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>p_cptac_proteomics_cis_loss</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>0.582385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>p_cptac_proteomics_cis_gain</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>0.579498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>p_cptac_proteomics_trans</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>1.089051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>p_cptac_proteomics_trans_loss</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>0.577165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>p_cptac_proteomics_trans_gain</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>0.589436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>p_cptac_transcriptomics</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>2.067109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>p_cptac_transcriptomics_cis</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>0.872976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>p_cptac_transcriptomics_cis_loss</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>0.827562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>p_cptac_transcriptomics_cis_gain</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>0.129713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>p_cptac_transcriptomics_trans</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>3.244864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>p_cptac_transcriptomics_trans_loss</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>1.383807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>p_cptac_transcriptomics_trans_gain</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>1.554197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>p_gistic</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>0.471075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>p_gistic_proteomics</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>0.392695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>p_gistic_proteomics_cis</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>0.272025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>p_gistic_proteomics_cis_loss</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>0.177906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>p_gistic_proteomics_cis_gain</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>0.667958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>p_gistic_proteomics_trans</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>0.220332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>p_gistic_proteomics_trans_loss</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>0.192948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>p_gistic_proteomics_trans_gain</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>0.695917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>p_gistic_transcriptomics</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>1.675656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>p_gistic_transcriptomics_cis</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>0.816872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>p_gistic_transcriptomics_cis_loss</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>0.805696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>p_gistic_transcriptomics_cis_gain</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>-0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>p_gistic_transcriptomics_trans</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>2.022133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>p_gistic_transcriptomics_trans_loss</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>0.912348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>p_gistic_transcriptomics_trans_gain</td>\n",
       "      <td>1.30103</td>\n",
       "      <td>neg_log_na_props</td>\n",
       "      <td>1.099489</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>93 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   name     line              group  neg_log_p\n",
       "0                                 p_all  1.30103  neg_log_sig_props  10.471005\n",
       "1                               p_cptac  1.30103  neg_log_sig_props   9.569432\n",
       "2                    p_cptac_proteomics  1.30103  neg_log_sig_props   7.469667\n",
       "3                p_cptac_proteomics_cis  1.30103  neg_log_sig_props   3.151706\n",
       "4           p_cptac_proteomics_cis_loss  1.30103  neg_log_sig_props   1.269244\n",
       "5           p_cptac_proteomics_cis_gain  1.30103  neg_log_sig_props   2.002133\n",
       "6              p_cptac_proteomics_trans  1.30103  neg_log_sig_props   4.713781\n",
       "7         p_cptac_proteomics_trans_loss  1.30103  neg_log_sig_props   2.247902\n",
       "8         p_cptac_proteomics_trans_gain  1.30103  neg_log_sig_props   2.646476\n",
       "9               p_cptac_transcriptomics  1.30103  neg_log_sig_props   2.728121\n",
       "10          p_cptac_transcriptomics_cis  1.30103  neg_log_sig_props   1.930384\n",
       "11     p_cptac_transcriptomics_cis_loss  1.30103  neg_log_sig_props   0.779076\n",
       "12     p_cptac_transcriptomics_cis_gain  1.30103  neg_log_sig_props   1.067931\n",
       "13        p_cptac_transcriptomics_trans  1.30103  neg_log_sig_props   2.438835\n",
       "14   p_cptac_transcriptomics_trans_loss  1.30103  neg_log_sig_props   0.829068\n",
       "15   p_cptac_transcriptomics_trans_gain  1.30103  neg_log_sig_props   1.882081\n",
       "16                             p_gistic  1.30103  neg_log_sig_props   5.246858\n",
       "17                  p_gistic_proteomics  1.30103  neg_log_sig_props   2.846877\n",
       "18              p_gistic_proteomics_cis  1.30103  neg_log_sig_props   1.497648\n",
       "19         p_gistic_proteomics_cis_loss  1.30103  neg_log_sig_props   0.694633\n",
       "20         p_gistic_proteomics_cis_gain  1.30103  neg_log_sig_props   0.968778\n",
       "21            p_gistic_proteomics_trans  1.30103  neg_log_sig_props   1.573778\n",
       "22       p_gistic_proteomics_trans_loss  1.30103  neg_log_sig_props   0.548059\n",
       "23       p_gistic_proteomics_trans_gain  1.30103  neg_log_sig_props   1.382908\n",
       "24             p_gistic_transcriptomics  1.30103  neg_log_sig_props   2.762087\n",
       "25         p_gistic_transcriptomics_cis  1.30103  neg_log_sig_props   1.106476\n",
       "26    p_gistic_transcriptomics_cis_loss  1.30103  neg_log_sig_props   0.763201\n",
       "27    p_gistic_transcriptomics_cis_gain  1.30103  neg_log_sig_props   0.496325\n",
       "28       p_gistic_transcriptomics_trans  1.30103  neg_log_sig_props   1.870819\n",
       "29  p_gistic_transcriptomics_trans_loss  1.30103  neg_log_sig_props   0.834192\n",
       "..                                  ...      ...                ...        ...\n",
       "63                              p_cptac  1.30103   neg_log_na_props   1.992831\n",
       "64                   p_cptac_proteomics  1.30103   neg_log_na_props   1.970903\n",
       "65               p_cptac_proteomics_cis  1.30103   neg_log_na_props   1.073446\n",
       "66          p_cptac_proteomics_cis_loss  1.30103   neg_log_na_props   0.582385\n",
       "67          p_cptac_proteomics_cis_gain  1.30103   neg_log_na_props   0.579498\n",
       "68             p_cptac_proteomics_trans  1.30103   neg_log_na_props   1.089051\n",
       "69        p_cptac_proteomics_trans_loss  1.30103   neg_log_na_props   0.577165\n",
       "70        p_cptac_proteomics_trans_gain  1.30103   neg_log_na_props   0.589436\n",
       "71              p_cptac_transcriptomics  1.30103   neg_log_na_props   2.067109\n",
       "72          p_cptac_transcriptomics_cis  1.30103   neg_log_na_props   0.872976\n",
       "73     p_cptac_transcriptomics_cis_loss  1.30103   neg_log_na_props   0.827562\n",
       "74     p_cptac_transcriptomics_cis_gain  1.30103   neg_log_na_props   0.129713\n",
       "75        p_cptac_transcriptomics_trans  1.30103   neg_log_na_props   3.244864\n",
       "76   p_cptac_transcriptomics_trans_loss  1.30103   neg_log_na_props   1.383807\n",
       "77   p_cptac_transcriptomics_trans_gain  1.30103   neg_log_na_props   1.554197\n",
       "78                             p_gistic  1.30103   neg_log_na_props   0.471075\n",
       "79                  p_gistic_proteomics  1.30103   neg_log_na_props   0.392695\n",
       "80              p_gistic_proteomics_cis  1.30103   neg_log_na_props   0.272025\n",
       "81         p_gistic_proteomics_cis_loss  1.30103   neg_log_na_props   0.177906\n",
       "82         p_gistic_proteomics_cis_gain  1.30103   neg_log_na_props   0.667958\n",
       "83            p_gistic_proteomics_trans  1.30103   neg_log_na_props   0.220332\n",
       "84       p_gistic_proteomics_trans_loss  1.30103   neg_log_na_props   0.192948\n",
       "85       p_gistic_proteomics_trans_gain  1.30103   neg_log_na_props   0.695917\n",
       "86             p_gistic_transcriptomics  1.30103   neg_log_na_props   1.675656\n",
       "87         p_gistic_transcriptomics_cis  1.30103   neg_log_na_props   0.816872\n",
       "88    p_gistic_transcriptomics_cis_loss  1.30103   neg_log_na_props   0.805696\n",
       "89    p_gistic_transcriptomics_cis_gain  1.30103   neg_log_na_props  -0.000000\n",
       "90       p_gistic_transcriptomics_trans  1.30103   neg_log_na_props   2.022133\n",
       "91  p_gistic_transcriptomics_trans_loss  1.30103   neg_log_na_props   0.912348\n",
       "92  p_gistic_transcriptomics_trans_gain  1.30103   neg_log_na_props   1.099489\n",
       "\n",
       "[93 rows x 4 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_pvals = all_pvals.infer_objects()\n",
    "\n",
    "all_pvals = all_pvals.assign(\n",
    "    line=-np.log10(0.05),\n",
    "    neg_log_sig_props=-np.log10(all_pvals[\"sig_props\"]),\n",
    "    neg_log_not_sig_props=-np.log10(all_pvals[\"not_sig_props\"]),\n",
    "    neg_log_na_props=-np.log10(all_pvals[\"na_props\"]),\n",
    ")\n",
    "\n",
    "all_pvals = all_pvals.drop(columns=[\"sig_props\", \"not_sig_props\", \"na_props\"])\n",
    "all_pvals = all_pvals.melt(\n",
    "    id_vars=[\"name\", \"line\"],\n",
    "    value_vars=[\"neg_log_sig_props\", \"neg_log_not_sig_props\", \"neg_log_na_props\"],\n",
    "    var_name=\"group\",\n",
    "    value_name=\"neg_log_p\",\n",
    ")\n",
    "\n",
    "base = alt.Chart(all_pvals)\n",
    "\n",
    "dots = base.mark_point().encode(\n",
    "    x=\"name\",\n",
    "    y=\"neg_log_p\",\n",
    "    color=\"group\",\n",
    ")\n",
    "\n",
    "line = base.mark_rule(color=\"crimson\").encode(\n",
    "    y=\"line\"\n",
    ")\n",
    "\n",
    "dots + line"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
