{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c073ac0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import cnvutils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fc49b8c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import altair as alt\n",
    "import numpy as np\n",
    "import os\n",
    "import statsmodels\n",
    "\n",
    "from cnvutils.filenames import (\n",
    "    get_proportions_perm_p_values_path,\n",
    ")\n",
    "\n",
    "from cnvutils.constants import SIG_CUTOFF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e6e769e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-25dfad4e16c44d8883aaf0a04dc5a88c\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-25dfad4e16c44d8883aaf0a04dc5a88c\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-25dfad4e16c44d8883aaf0a04dc5a88c\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.17.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"4.17.0\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}}, \"layer\": [{\"mark\": \"point\", \"encoding\": {\"x\": {\"field\": \"name\", \"type\": \"nominal\"}, \"y\": {\"field\": \"neg_log_adj_p\", \"type\": \"quantitative\"}}}, {\"mark\": {\"type\": \"rule\", \"color\": \"crimson\"}, \"encoding\": {\"y\": {\"field\": \"line\", \"type\": \"quantitative\"}}}], \"data\": {\"name\": \"data-759bb282b00126f0dfaa6b29c30fecdd\"}, \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.17.0.json\", \"datasets\": {\"data-759bb282b00126f0dfaa6b29c30fecdd\": [{\"name\": \"p_gistic\", \"adj_p\": 0.027, \"line\": 1.3010299956639813, \"neg_log_adj_p\": 1.5686362358410126}, {\"name\": \"p_gistic_proteomics\", \"adj_p\": 0.064125, \"line\": 1.3010299956639813, \"neg_log_adj_p\": 1.1929726218801273}, {\"name\": \"p_gistic_proteomics_cis\", \"adj_p\": 0.03685714285714286, \"line\": 1.3010299956639813, \"neg_log_adj_p\": 1.4334783340510266}, {\"name\": \"p_gistic_proteomics_cis_gain\", \"adj_p\": 0.19724999999999998, \"line\": 1.3010299956639813, \"neg_log_adj_p\": 0.7049829881185421}, {\"name\": \"p_gistic_proteomics_cis_loss\", \"adj_p\": 0.027, \"line\": 1.3010299956639813, \"neg_log_adj_p\": 1.5686362358410126}, {\"name\": \"p_gistic_proteomics_trans\", \"adj_p\": 0.11220000000000001, \"line\": 1.3010299956639813, \"neg_log_adj_p\": 0.9500071430798573}, {\"name\": \"p_gistic_proteomics_trans_gain\", \"adj_p\": 0.4129285714285714, \"line\": 1.3010299956639813, \"neg_log_adj_p\": 0.38412506630312265}, {\"name\": \"p_gistic_proteomics_trans_loss\", \"adj_p\": 0.024999999999999998, \"line\": 1.3010299956639813, \"neg_log_adj_p\": 1.6020599913279625}, {\"name\": \"p_gistic_transcriptomics\", \"adj_p\": 0.03685714285714286, \"line\": 1.3010299956639813, \"neg_log_adj_p\": 1.4334783340510266}, {\"name\": \"p_gistic_transcriptomics_cis\", \"adj_p\": 0.006, \"line\": 1.3010299956639813, \"neg_log_adj_p\": 2.221848749616356}, {\"name\": \"p_gistic_transcriptomics_cis_gain\", \"adj_p\": 0.01875, \"line\": 1.3010299956639813, \"neg_log_adj_p\": 1.7269987279362624}, {\"name\": \"p_gistic_transcriptomics_cis_loss\", \"adj_p\": 0.11220000000000001, \"line\": 1.3010299956639813, \"neg_log_adj_p\": 0.9500071430798573}, {\"name\": \"p_gistic_transcriptomics_trans\", \"adj_p\": 0.4039615384615385, \"line\": 1.3010299956639813, \"neg_log_adj_p\": 0.3936599824861229}, {\"name\": \"p_gistic_transcriptomics_trans_gain\", \"adj_p\": 0.4676, \"line\": 1.3010299956639813, \"neg_log_adj_p\": 0.3301254975101975}, {\"name\": \"p_gistic_transcriptomics_trans_loss\", \"adj_p\": 0.1864090909090909, \"line\": 1.3010299956639813, \"neg_log_adj_p\": 0.7295329115347216}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.LayerChart(...)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def prop_perm_p_vals_manhattan_plot(\n",
    "    chromosome,\n",
    "    data_dir=os.path.join(os.getcwd(), \"..\", \"data\"),\n",
    "):\n",
    "    \n",
    "    p_path = get_proportions_perm_p_values_path(\n",
    "        data_dir=data_dir,\n",
    "        chromosome=chromosome,\n",
    "    )\n",
    "    \n",
    "    p_vals = pd.read_csv(p_path, sep=\"\\t\")\n",
    "    \n",
    "    # Filter out just the p values we want\n",
    "    p_vals = p_vals[p_vals[\"group\"] == \"sig_props\"][[\"name\", \"p\"]]\n",
    "    p_vals = p_vals[p_vals[\"name\"].str.contains(\"gistic\")]\n",
    "    \n",
    "    # Benjamini-Hochberg FDR correction\n",
    "    reject, adj_p, alphac_sidak, alphac_bonf = statsmodels.stats.multitest.multipletests(\n",
    "        pvals=p_vals[\"p\"],\n",
    "        alpha=SIG_CUTOFF,\n",
    "        method=\"fdr_bh\",\n",
    "    )\n",
    "    \n",
    "    p_vals = p_vals.assign(adj_p=adj_p)\n",
    "    \n",
    "#     return p_vals\n",
    "\n",
    "    # Make Manhattan plot\n",
    "    p_vals = p_vals.\\\n",
    "    assign(\n",
    "        line=-np.log10(SIG_CUTOFF),\n",
    "        neg_log_adj_p=-np.log10(p_vals[\"adj_p\"]),\n",
    "    ).\\\n",
    "    drop(columns=\"p\")\n",
    "\n",
    "    base = alt.Chart(p_vals)\n",
    "\n",
    "    dots = base.mark_point().encode(\n",
    "        x=\"name\",\n",
    "        y=\"neg_log_adj_p\",\n",
    "    )\n",
    "\n",
    "    line = base.mark_rule(color=\"crimson\").encode(\n",
    "        y=\"line\"\n",
    "    )\n",
    "\n",
    "    return dots + line\n",
    "\n",
    "prop_perm_p_vals_manhattan_plot(chromosome=8)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
