{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SCNA analysis step 2: Get locations of genes\n",
    "\n",
    "- For each cancer type:\n",
    "    1. Assign locations to each gene\n",
    "    2. Based on location, assign which chromosome arm it's on"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cptac\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pyensembl\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Prepare Ensembl API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make sure we have the Ensembl data downloaded\n",
    "# The most recent release is 100, but 99 is still\n",
    "# recent (Jan 2020), and PyEnsembl only supports\n",
    "# up to 99 right now.\n",
    "ensembl = pyensembl.EnsemblRelease(99)\n",
    "\n",
    "try:\n",
    "    ensembl.genes() # If this fails, we need to download the data again.\n",
    "except ValueError as e:\n",
    "    if str(e).startswith(\"Missing genome data file from \"):\n",
    "        ensembl.download()\n",
    "        ensembl.index()\n",
    "    else:\n",
    "        raise e from None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get centromere locations from cytoband table\n",
    "Table downloaded from https://genome.ucsc.edu/cgi-bin/hgTables?hgsid=856734045_VsAQvZHdBDGH2PuUBBiicYCEoYFm&clade=mammal&org=Human&db=hg38&hgta_group=allTables&hgta_track=hg38&hgta_table=cytoBand&hgta_regionType=genome&position=chrX%3A15%2C560%2C138-15%2C602%2C945&hgta_outputType=primaryTable&hgta_outFileName="
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>123400000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>39800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>53400000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr12</td>\n",
       "      <td>35500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr13</td>\n",
       "      <td>17700000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr14</td>\n",
       "      <td>17200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr15</td>\n",
       "      <td>19000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr16</td>\n",
       "      <td>36800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr17</td>\n",
       "      <td>25100000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr18</td>\n",
       "      <td>18500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr19</td>\n",
       "      <td>26200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr2</td>\n",
       "      <td>93900000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr20</td>\n",
       "      <td>28100000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr21</td>\n",
       "      <td>12000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr22</td>\n",
       "      <td>15000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr3</td>\n",
       "      <td>90900000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr4</td>\n",
       "      <td>50000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr5</td>\n",
       "      <td>48800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr6</td>\n",
       "      <td>59800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr7</td>\n",
       "      <td>60100000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr8</td>\n",
       "      <td>45200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr9</td>\n",
       "      <td>43000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chrX</td>\n",
       "      <td>61000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chrY</td>\n",
       "      <td>10400000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom   position\n",
       "0    chr1  123400000\n",
       "1   chr10   39800000\n",
       "2   chr11   53400000\n",
       "3   chr12   35500000\n",
       "4   chr13   17700000\n",
       "5   chr14   17200000\n",
       "6   chr15   19000000\n",
       "7   chr16   36800000\n",
       "8   chr17   25100000\n",
       "9   chr18   18500000\n",
       "10  chr19   26200000\n",
       "11   chr2   93900000\n",
       "12  chr20   28100000\n",
       "13  chr21   12000000\n",
       "14  chr22   15000000\n",
       "15   chr3   90900000\n",
       "16   chr4   50000000\n",
       "17   chr5   48800000\n",
       "18   chr6   59800000\n",
       "19   chr7   60100000\n",
       "20   chr8   45200000\n",
       "21   chr9   43000000\n",
       "22   chrX   61000000\n",
       "23   chrY   10400000"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytobands = pd.read_csv(\"ucsc_cytoband.tsv\", sep=\"\\t\").\\\n",
    "    rename(columns={\"#chrom\": \"chrom\"})\n",
    "\n",
    "centromeres = cytobands.\\\n",
    "    loc[cytobands.gieStain == \"acen\", [\"chrom\", \"chromStart\", \"chromEnd\"]].\\\n",
    "    melt(id_vars=\"chrom\", value_name=\"position\").\\\n",
    "    drop(columns=\"variable\").\\\n",
    "    sort_values(by=[\"chrom\", \"position\"])\n",
    "\n",
    "centromeres = centromeres[centromeres.duplicatedk(keep=False)].\\\n",
    "    drop_duplicates(keep=\"first\").\\\n",
    "    reset_index(drop=True)\n",
    "\n",
    "centromeres\n",
    "\n",
    "# TODO: Is the p arm always before the centromere?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get gene locations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_genes_locations(cancer_type, data_dir):\n",
    "    \"\"\"For a particular cancer type, load the CNA table, add location and \n",
    "    chromosome information, and then save it back to the same location.\n",
    "    \n",
    "    Parameters:\n",
    "    cancer_type (str): The name of the cancer type. Must match its name in the file name to read from.\n",
    "    data_dir (str): The path to the directory with the CNA files from step 1.\n",
    "    \"\"\"\n",
    "    \n",
    "    file_path = os.path.join(data_dir, f\"{cancer_type}_cna_long.tsv.gz\")\n",
    "    cna = pd.read_csv(file_path, sep='\\t')\n",
    "\n",
    "    chrs = []\n",
    "    starts = []\n",
    "    ends = []\n",
    "\n",
    "    for row in cna.itertuples(index=False):\n",
    "        db_id = row.Database_ID\n",
    "\n",
    "        if pd.notnull(db_id):\n",
    "            try:\n",
    "                info = ensembl.gene_by_id(db_id)\n",
    "            except ValueError as e:\n",
    "                if str(e).startswith(\"Gene not found: \"):\n",
    "                    pass # This will go down to the next try/catch and attempt lookup by name instead of ID\n",
    "                else:\n",
    "                    raise e from None\n",
    "            else:\n",
    "                chrs.append(info.contig)\n",
    "                starts.append(info.start)\n",
    "                ends.append(info.end)\n",
    "                continue\n",
    "\n",
    "        # We get to the following try/catch either if Database_ID is null, or \n",
    "        # if nothing was found by querying by Database_ID. That way we if don't\n",
    "        # find anything with the Database_ID, we try again with the gene name.\n",
    "        \n",
    "        # It appears that some genes have old names that are out of date,\n",
    "        # such as LSMD1. If we want to get even better coverage, we could\n",
    "        # try querying HGNC with old gene names, if the below returns nothing.\n",
    "        # But we're getting fairly good coverage right now, and we don't\n",
    "        # want to waste time going down an unnecessary rabbit hole.\n",
    "        try:\n",
    "            info = ensembl.genes_by_name(row.gene)\n",
    "        except ValueError as e:\n",
    "            if str(e).startswith(\"No results found for query\"):\n",
    "                chrs.append(np.nan)\n",
    "                starts.append(np.nan)\n",
    "                ends.append(np.nan)\n",
    "            else:\n",
    "                raise e from None\n",
    "        else:\n",
    "            chrs.append(info[0].contig)\n",
    "            starts.append(info[0].start)\n",
    "            ends.append(info[0].end)\n",
    "\n",
    "    # Add the columns we created\n",
    "    cna = cna.assign(\n",
    "        chromosome=chrs,\n",
    "        start=starts,\n",
    "        end=ends\n",
    "    )\n",
    "\n",
    "    # Check what proportion of genes we didn't find info for\n",
    "    not_found_prop = pd.isnull(cna[\"start\"]).sum() / cna.shape[0]\n",
    "\n",
    "    # Select only the genes we found info for\n",
    "    cna = cna[pd.notnull(cna[\"start\"])]\n",
    "\n",
    "    # Save it back to the same file path\n",
    "    cna.to_csv(file_path, index=False, compression=\"gzip\", sep=\"\\t\")\n",
    "    \n",
    "    return not_found_prop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_types = [\n",
    "    \"brca\",\n",
    "    \"ccrcc\",\n",
    "    \"colon\",\n",
    "    \"endometrial\",\n",
    "    \"gbm\",\n",
    "    \"hnscc\",\n",
    "    \"lscc\",\n",
    "    \"luad\",\n",
    "    \"ovarian\"\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for cancer_type in cancer_types:\n",
    "    print(f\"{cancer_type}: {add_genes_locations * 100}% not found.\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
