{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SCNA analysis step 2: Summarize amplifications and deletions on each arm\n",
    "\n",
    "We will create a table for each cancer type that has, for each chromosome arm in each patient, the proportion of the arm that was amplified and the proportion that was deleted."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get chromosome arm lengths from cytoband table\n",
    "Table downloaded from https://genome.ucsc.edu/cgi-bin/hgTables?hgsid=856734045_VsAQvZHdBDGH2PuUBBiicYCEoYFm&clade=mammal&org=Human&db=hg38&hgta_group=allTables&hgta_track=hg38&hgta_table=cytoBand&hgta_regionType=genome&position=chrX%3A15%2C560%2C138-15%2C602%2C945&hgta_outputType=primaryTable&hgta_outFileName=ucsc_cytoband.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>arm</th>\n",
       "      <th>arm_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>p</td>\n",
       "      <td>123400000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>q</td>\n",
       "      <td>125556422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10</td>\n",
       "      <td>p</td>\n",
       "      <td>39800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10</td>\n",
       "      <td>q</td>\n",
       "      <td>93997422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11</td>\n",
       "      <td>p</td>\n",
       "      <td>53400000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>11</td>\n",
       "      <td>q</td>\n",
       "      <td>81686622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>12</td>\n",
       "      <td>p</td>\n",
       "      <td>35500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>12</td>\n",
       "      <td>q</td>\n",
       "      <td>97775309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>13</td>\n",
       "      <td>p</td>\n",
       "      <td>17700000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>13</td>\n",
       "      <td>q</td>\n",
       "      <td>96664328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>14</td>\n",
       "      <td>p</td>\n",
       "      <td>17200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>14</td>\n",
       "      <td>q</td>\n",
       "      <td>89843718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>15</td>\n",
       "      <td>p</td>\n",
       "      <td>19000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>15</td>\n",
       "      <td>q</td>\n",
       "      <td>82991189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>16</td>\n",
       "      <td>p</td>\n",
       "      <td>36800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>q</td>\n",
       "      <td>53538345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17</td>\n",
       "      <td>p</td>\n",
       "      <td>25100000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>q</td>\n",
       "      <td>58157441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>18</td>\n",
       "      <td>p</td>\n",
       "      <td>18500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>18</td>\n",
       "      <td>q</td>\n",
       "      <td>61873285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>19</td>\n",
       "      <td>p</td>\n",
       "      <td>26200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>19</td>\n",
       "      <td>q</td>\n",
       "      <td>32417616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>2</td>\n",
       "      <td>p</td>\n",
       "      <td>93900000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>2</td>\n",
       "      <td>q</td>\n",
       "      <td>148293529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>20</td>\n",
       "      <td>p</td>\n",
       "      <td>28100000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>20</td>\n",
       "      <td>q</td>\n",
       "      <td>36344167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>21</td>\n",
       "      <td>p</td>\n",
       "      <td>12000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>21</td>\n",
       "      <td>q</td>\n",
       "      <td>34709983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>22</td>\n",
       "      <td>p</td>\n",
       "      <td>15000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>22</td>\n",
       "      <td>q</td>\n",
       "      <td>35818468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>3</td>\n",
       "      <td>p</td>\n",
       "      <td>90900000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>3</td>\n",
       "      <td>q</td>\n",
       "      <td>107395559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>4</td>\n",
       "      <td>p</td>\n",
       "      <td>50000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>4</td>\n",
       "      <td>q</td>\n",
       "      <td>140214555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>5</td>\n",
       "      <td>p</td>\n",
       "      <td>48800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>5</td>\n",
       "      <td>q</td>\n",
       "      <td>132738259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>6</td>\n",
       "      <td>p</td>\n",
       "      <td>59800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>6</td>\n",
       "      <td>q</td>\n",
       "      <td>111005979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>7</td>\n",
       "      <td>p</td>\n",
       "      <td>60100000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>7</td>\n",
       "      <td>q</td>\n",
       "      <td>99245973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>8</td>\n",
       "      <td>p</td>\n",
       "      <td>45200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>8</td>\n",
       "      <td>q</td>\n",
       "      <td>99938636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>9</td>\n",
       "      <td>p</td>\n",
       "      <td>43000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>9</td>\n",
       "      <td>q</td>\n",
       "      <td>95394717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>X</td>\n",
       "      <td>p</td>\n",
       "      <td>61000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>X</td>\n",
       "      <td>q</td>\n",
       "      <td>95040895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Y</td>\n",
       "      <td>p</td>\n",
       "      <td>10400000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Y</td>\n",
       "      <td>q</td>\n",
       "      <td>46827415</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome arm  arm_length\n",
       "0           1   p   123400000\n",
       "1           1   q   125556422\n",
       "2          10   p    39800000\n",
       "3          10   q    93997422\n",
       "4          11   p    53400000\n",
       "5          11   q    81686622\n",
       "6          12   p    35500000\n",
       "7          12   q    97775309\n",
       "8          13   p    17700000\n",
       "9          13   q    96664328\n",
       "10         14   p    17200000\n",
       "11         14   q    89843718\n",
       "12         15   p    19000000\n",
       "13         15   q    82991189\n",
       "14         16   p    36800000\n",
       "15         16   q    53538345\n",
       "16         17   p    25100000\n",
       "17         17   q    58157441\n",
       "18         18   p    18500000\n",
       "19         18   q    61873285\n",
       "20         19   p    26200000\n",
       "21         19   q    32417616\n",
       "22          2   p    93900000\n",
       "23          2   q   148293529\n",
       "24         20   p    28100000\n",
       "25         20   q    36344167\n",
       "26         21   p    12000000\n",
       "27         21   q    34709983\n",
       "28         22   p    15000000\n",
       "29         22   q    35818468\n",
       "30          3   p    90900000\n",
       "31          3   q   107395559\n",
       "32          4   p    50000000\n",
       "33          4   q   140214555\n",
       "34          5   p    48800000\n",
       "35          5   q   132738259\n",
       "36          6   p    59800000\n",
       "37          6   q   111005979\n",
       "38          7   p    60100000\n",
       "39          7   q    99245973\n",
       "40          8   p    45200000\n",
       "41          8   q    99938636\n",
       "42          9   p    43000000\n",
       "43          9   q    95394717\n",
       "44          X   p    61000000\n",
       "45          X   q    95040895\n",
       "46          Y   p    10400000\n",
       "47          Y   q    46827415"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in the cytoband table\n",
    "cytobands = pd.read_csv(\"ucsc_cytoband.tsv\", sep=\"\\t\").\\\n",
    "    rename(columns={\"#chrom\": \"chromosome\"})\n",
    "\n",
    "\n",
    "## Find the length of each chromosome by taking the maximum of all locations given for cytobands\n",
    "\n",
    "chr_lengths = cytobands[\n",
    "        ~cytobands[\"chromosome\"].str.contains(\"_\") & \n",
    "        ~cytobands[\"chromosome\"].str.endswith(\"M\")\n",
    "    ].\\\n",
    "    rename(columns={\"chromEnd\": \"chromosome_len\"}).\\\n",
    "    groupby(\"chromosome\")[[\"chromosome_len\"]].\\\n",
    "    max().\\\n",
    "    reset_index()\n",
    "\n",
    "# Trim the chromosome names to match other tables\n",
    "chr_lengths = chr_lengths.assign(chromosome=chr_lengths[\"chromosome\"].str[3:])\n",
    "\n",
    "## Find the lengths of the p arms by getting the locations of the centromeres\n",
    "\n",
    "# Select just the \"acen\" stained bands, which are the centromeres\n",
    "centromeres = cytobands.\\\n",
    "    loc[cytobands.gieStain == \"acen\", [\"chromosome\", \"chromStart\", \"chromEnd\"]].\\\n",
    "    melt(id_vars=\"chromosome\", value_name=\"centromere_location\").\\\n",
    "    drop(columns=\"variable\").\\\n",
    "    sort_values(by=[\"chromosome\", \"centromere_location\"])\n",
    "\n",
    "# Get the central location for each centromere\n",
    "# Also trim the chromosome names to not start with \"chr\"\n",
    "centromeres = centromeres[centromeres.duplicated(keep=False)].\\\n",
    "    drop_duplicates(keep=\"first\").\\\n",
    "    assign(chromosome=centromeres[\"chromosome\"].str[3:]).\\\n",
    "    sort_values(by=[\"chromosome\", \"centromere_location\"]).\\\n",
    "    reset_index(drop=True)\n",
    "\n",
    "p_arms = centromeres.\\\n",
    "    assign(arm=\"p\").\\\n",
    "    rename(columns={\"centromere_location\": \"arm_length\"})[[\"chromosome\", \"arm\", \"arm_length\"]]\n",
    "\n",
    "## Find the lengths of the q arms by subtracting the position of the centromere\n",
    "## from the overall length of the chromosome\n",
    "\n",
    "q_arms = chr_lengths.merge(\n",
    "    right=centromeres,\n",
    "    how=\"outer\",\n",
    "    on=\"chromosome\",\n",
    "    validate=\"one_to_one\"\n",
    ")\n",
    "\n",
    "q_arms = q_arms.assign(\n",
    "    arm=\"q\",\n",
    "    arm_length=q_arms[\"chromosome_len\"] - q_arms[\"centromere_location\"]\n",
    ")[[\"chromosome\", \"arm\", \"arm_length\"]]\n",
    "\n",
    "## Combine it all into one table\n",
    "chr_arm_lengths = p_arms.\\\n",
    "    append(q_arms).\\\n",
    "    sort_values(by=[\"chromosome\", \"arm\"]).\\\n",
    "    reset_index(drop=True)\n",
    "\n",
    "chr_arm_lengths"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate summaries for each arm in each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def summarize_arms(\n",
    "    cancer_type, \n",
    "    arm_lengths, \n",
    "    cutoff, \n",
    "    input_dir, \n",
    "    output_dir\n",
    "):\n",
    "    \n",
    "    input_file = os.path.join(input_dir, f\"{cancer_type}_cna_long.tsv.gz\")\n",
    "    \n",
    "    # It's important that we specify to read the chromosome column as \n",
    "    # objects, so that the numbers will be strings, not ints, and thus\n",
    "    # merge correctly with the arm lengths table\n",
    "    cna = pd.read_csv(input_file, sep='\\t', dtype={\"chromosome\": \"O\"})\n",
    "    \n",
    "    # Join in the chromosome lengths\n",
    "    cna = cna.merge(\n",
    "        right=arm_lengths,\n",
    "        how=\"outer\",\n",
    "        on=[\"chromosome\", \"arm\"],\n",
    "        validate=\"many_to_one\"\n",
    "    )\n",
    "    \n",
    "    # Slice out only the genes that pass the cutoff\n",
    "    cna = cna[cna[\"cna_val\"].abs() >= cutoff]\n",
    "    \n",
    "    # For each gene, calculate the proportion of its chromosome arm that it covers.\n",
    "    cna = cna.assign(prop_arm=(cna[\"end\"] - cna[\"start\"]).abs() / cna[\"arm_length\"])\n",
    "    \n",
    "    # Slice out just the columns we need now\n",
    "    cna = cna[[\"Patient_ID\", \"gene\", \"Database_ID\", \"cna_val\", \"chromosome\", \"arm\", \"prop_arm\"]]\n",
    "    \n",
    "    # Separate the amplifications and the deletions\n",
    "    amps = cna[cna[\"cna_val\"] > 0]\n",
    "    dels = cna[cna[\"cna_val\"] < 0]\n",
    "    \n",
    "    # Calculate the proportion of each arm of each chromosome that is\n",
    "    # amplified and deleted in each sample\n",
    "    amp_summary = amps.\\\n",
    "        groupby([\"Patient_ID\", \"chromosome\", \"arm\"])[[\"prop_arm\"]].\\\n",
    "        sum().\\\n",
    "        rename(columns={\"prop_arm\": \"prop_arm_amplified\"}).\\\n",
    "        reset_index(drop=False)\n",
    "    \n",
    "    del_summary = dels.\\\n",
    "        groupby([\"Patient_ID\", \"chromosome\", \"arm\"])[[\"prop_arm\"]].\\\n",
    "        sum().\\\n",
    "        rename(columns={\"prop_arm\": \"prop_arm_deleted\"}).\\\n",
    "        reset_index(drop=False)\n",
    "    \n",
    "    # Merge them into one summary table\n",
    "    summary = amp_summary.\\\n",
    "        merge(\n",
    "            right=del_summary,\n",
    "            how=\"outer\",\n",
    "            on=[\"Patient_ID\", \"chromosome\", \"arm\"],\n",
    "            validate=\"one_to_one\"\n",
    "        ).\\\n",
    "        fillna(0)\n",
    "    \n",
    "    # Save the output\n",
    "    if not os.path.isdir(output_dir):\n",
    "        os.mkdir(output_dir)\n",
    "\n",
    "    file_path = os.path.join(output_dir, f\"{cancer_type}_cna_long.tsv.gz\")\n",
    "    cna.to_csv(file_path, index=False, compression=\"gzip\", sep=\"\\t\")\n",
    "\n",
    "    return summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_types = [\n",
    "    \"brca\",\n",
    "    \"ccrcc\",\n",
    "    \"colon\",\n",
    "    \"endometrial\",\n",
    "    \"gbm\",\n",
    "    \"hnscc\",\n",
    "    \"lscc\",\n",
    "    \"luad\",\n",
    "    \"ovarian\"\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for cancer_type in cancer_types:\n",
    "    summarize_arms(\n",
    "        cancer_type=cancer_type,\n",
    "        arm_lengths=chr_arm_lengths,\n",
    "        cutoff=0.2,\n",
    "        input_dir=\"long_cna_tables\",\n",
    "        output_dir=\"summary_tables\"\n",
    "    )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
