{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# \\*\\*\\*WARNING\\*\\*\\*: Bad analysis technique\n",
    "\n",
    "I realized that this folder of analyses falls prey to the fallacy of indirect comparison. We compare with_event_tumor to normal, and without_event_tumor to normal, and then compare the with_tumor and without_tumor results from that to each other. It's better to just directly compare with_tumor to without_tumor, which we do with a t-test in the finished analyses."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Enrichment NB 03: Analyzing rSEA results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import altair as alt\n",
    "from toolz.curried import pipe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "ALPHA = 0.05\n",
    "MSV = 0.5\n",
    "\n",
    "plots_dir = \"altair_data\"\n",
    "\n",
    "go_results_file = os.path.join(\"results_02_rSEA\", \"enrichment_rsea_thresh_0.99_lib_GO_Biological_Process_2018.tsv\")\n",
    "reactome_results_file = os.path.join(\"results_02_rSEA\", \"enrichment_rsea_thresh_0.99_lib_ReactomePathways.tsv\")\n",
    "\n",
    "diff_expr_file = os.path.join(\"results_01_grouped_expr\", \"all_results.tsv.gz\")\n",
    "\n",
    "gene_set_dir = \"gene_set_libraries\"\n",
    "go_gmt = os.path.join(gene_set_dir, \"GO_Biological_Process_2018.gmt\")\n",
    "reactome_gmt = os.path.join(gene_set_dir, \"ReactomePathways.gmt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataTransformerRegistry.enable('json_dir')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Altair options\n",
    "alt.data_transformers.disable_max_rows()\n",
    "\n",
    "def json_dir(data, data_dir):\n",
    "    os.makedirs(data_dir, exist_ok=True)\n",
    "    return pipe(data, alt.to_json(filename=os.path.join(data_dir, \"{prefix}-{hash}.{extension}\")) )\n",
    "\n",
    "alt.data_transformers.register(\"json_dir\", json_dir)\n",
    "alt.data_transformers.enable(\"json_dir\", data_dir=plots_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_files = {\n",
    "    \"go\": go_results_file,\n",
    "    \"reactome\": reactome_results_file\n",
    "}\n",
    "\n",
    "results = {}\n",
    "\n",
    "for gene_set in results_files.keys():\n",
    "    \n",
    "    result = pd.read_csv(results_files[gene_set], sep=\"\\t\", index_col=0)\n",
    "    result = result[\n",
    "        (result[\"Coverage\"] > 0) & \n",
    "        (\n",
    "            (result[\"SC.adjP\"] <= ALPHA) | \n",
    "            (result[\"Comp.0.99.adjP\"] <= ALPHA)\n",
    "        )]\n",
    "    \n",
    "    grp_split = result[\"cancer_type_group\"].str.split(\"_\", n=1, expand=True)\n",
    "    \n",
    "    result = result.assign(\n",
    "        cancer_type=grp_split[0],\n",
    "        group=grp_split[1]\n",
    "    ).\\\n",
    "    drop(columns=\"cancer_type_group\").\\\n",
    "    sort_values(\n",
    "        by=[\"cancer_type\", \"group\", \"SC.adjP\", \"Coverage\", \"Size\"],\n",
    "        ascending=[True, True, True, False, False]\n",
    "    )\n",
    "    \n",
    "    results[gene_set] = result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Distribution of proportion of gene sets covered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-7c653b69dd294827accec5437b773b0c\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-7c653b69dd294827accec5437b773b0c\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-7c653b69dd294827accec5437b773b0c\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.8.1?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function loadScript(lib) {\n",
       "      return new Promise(function(resolve, reject) {\n",
       "        var s = document.createElement('script');\n",
       "        s.src = paths[lib];\n",
       "        s.async = true;\n",
       "        s.onload = () => resolve(paths[lib]);\n",
       "        s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "        document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "      });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else if (typeof vegaEmbed === \"function\") {\n",
       "      displayChart(vegaEmbed);\n",
       "    } else {\n",
       "      loadScript(\"vega\")\n",
       "        .then(() => loadScript(\"vega-lite\"))\n",
       "        .then(() => loadScript(\"vega-embed\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}}, \"data\": {\"url\": \"altair_data/altair-data-e2fac5cd0e03e3fafdc590709e29216d.json\", \"format\": {\"type\": \"json\"}}, \"mark\": \"bar\", \"encoding\": {\"x\": {\"type\": \"quantitative\", \"bin\": {\"step\": 0.01}, \"field\": \"Coverage\", \"scale\": {\"domain\": [0, 1]}}, \"y\": {\"type\": \"quantitative\", \"aggregate\": \"count\"}}, \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.8.1.json\"}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alt.Chart(results[\"go\"]).mark_bar().encode(\n",
    "    x=alt.X(\n",
    "        \"Coverage\",\n",
    "        bin=alt.Bin(step=0.01),\n",
    "        scale=alt.Scale(domain=[0, 1])\n",
    "    ),\n",
    "    y=alt.Y(\n",
    "        \"count()\"\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-123e2c1c7c4c4a8ea79e63d7caa90c5e\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-123e2c1c7c4c4a8ea79e63d7caa90c5e\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-123e2c1c7c4c4a8ea79e63d7caa90c5e\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.8.1?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function loadScript(lib) {\n",
       "      return new Promise(function(resolve, reject) {\n",
       "        var s = document.createElement('script');\n",
       "        s.src = paths[lib];\n",
       "        s.async = true;\n",
       "        s.onload = () => resolve(paths[lib]);\n",
       "        s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "        document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "      });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else if (typeof vegaEmbed === \"function\") {\n",
       "      displayChart(vegaEmbed);\n",
       "    } else {\n",
       "      loadScript(\"vega\")\n",
       "        .then(() => loadScript(\"vega-lite\"))\n",
       "        .then(() => loadScript(\"vega-embed\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}}, \"data\": {\"url\": \"altair_data/altair-data-38a3dd7b8e52660a72f6c558b790270e.json\", \"format\": {\"type\": \"json\"}}, \"mark\": {\"type\": \"bar\", \"clip\": true}, \"encoding\": {\"x\": {\"type\": \"quantitative\", \"bin\": {\"step\": 0.01}, \"field\": \"Coverage\", \"scale\": {\"domain\": [0, 1]}}, \"y\": {\"type\": \"quantitative\", \"aggregate\": \"count\"}}, \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.8.1.json\"}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Due to there being a duplicated gene name, and a gene set comprised only of that gene,\n",
    "# there is one entry that's marked as having a coverage proportion of 2. So that doesn't\n",
    "# mess up our chart, we'll just clip it out.\n",
    "\n",
    "alt.Chart(results[\"reactome\"]).mark_bar(clip=True).encode(\n",
    "    x=alt.X(\n",
    "        \"Coverage\",\n",
    "        bin=alt.Bin(step=0.01),\n",
    "        scale=alt.Scale(domain=[0, 1])\n",
    "    ),\n",
    "    y=alt.Y(\n",
    "        \"count()\"\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Look at individual groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Size</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>TDP.bound</th>\n",
       "      <th>TDP.estimate</th>\n",
       "      <th>SC.adjP</th>\n",
       "      <th>Comp.0.99.adjP</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cancer_type</th>\n",
       "      <th>group</th>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"30\" valign=\"top\">colon</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">with_both</th>\n",
       "      <th>protein export from nucleus (GO:0006611)</th>\n",
       "      <td>29</td>\n",
       "      <td>0.03</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.644517e-07</td>\n",
       "      <td>6.644517e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intracellular protein transport (GO:0006886)</th>\n",
       "      <td>348</td>\n",
       "      <td>0.02</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.644517e-07</td>\n",
       "      <td>4.999214e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclear export (GO:0051168)</th>\n",
       "      <td>93</td>\n",
       "      <td>0.02</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.644517e-07</td>\n",
       "      <td>4.999214e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mitotic nuclear envelope reassembly (GO:0007084)</th>\n",
       "      <td>11</td>\n",
       "      <td>0.18</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.536441e-06</td>\n",
       "      <td>4.999214e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>peptidyl-serine dephosphorylation (GO:0070262)</th>\n",
       "      <td>11</td>\n",
       "      <td>0.18</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.536441e-06</td>\n",
       "      <td>4.999214e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">with_gain</th>\n",
       "      <th>Golgi vesicle budding (GO:0048194)</th>\n",
       "      <td>10</td>\n",
       "      <td>0.20</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.866931e-09</td>\n",
       "      <td>4.996401e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>positive regulation of viral genome replication (GO:0045070)</th>\n",
       "      <td>32</td>\n",
       "      <td>0.16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.866931e-09</td>\n",
       "      <td>4.996401e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>regulation of viral genome replication (GO:0045069)</th>\n",
       "      <td>64</td>\n",
       "      <td>0.14</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.866931e-09</td>\n",
       "      <td>4.996401e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>positive regulation of viral life cycle (GO:1903902)</th>\n",
       "      <td>45</td>\n",
       "      <td>0.13</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.866931e-09</td>\n",
       "      <td>4.996401e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>modification by symbiont of host morphology or physiology (GO:0044003)</th>\n",
       "      <td>8</td>\n",
       "      <td>0.12</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.866931e-09</td>\n",
       "      <td>4.866931e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">with_loss</th>\n",
       "      <th>intracellular protein transport (GO:0006886)</th>\n",
       "      <td>348</td>\n",
       "      <td>0.03</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.042659e-09</td>\n",
       "      <td>4.979156e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein export from nucleus (GO:0006611)</th>\n",
       "      <td>29</td>\n",
       "      <td>0.03</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.042659e-09</td>\n",
       "      <td>3.042659e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclear export (GO:0051168)</th>\n",
       "      <td>93</td>\n",
       "      <td>0.02</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.042659e-09</td>\n",
       "      <td>4.979156e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>positive regulation of exocytosis (GO:0045921)</th>\n",
       "      <td>35</td>\n",
       "      <td>0.14</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.263430e-07</td>\n",
       "      <td>4.979156e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>positive regulation of secretion by cell (GO:1903532)</th>\n",
       "      <td>53</td>\n",
       "      <td>0.11</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.263430e-07</td>\n",
       "      <td>4.979156e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">without_both</th>\n",
       "      <th>snRNA transcription (GO:0009301)</th>\n",
       "      <td>71</td>\n",
       "      <td>0.17</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.052051e-12</td>\n",
       "      <td>4.989211e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snRNA transcription from RNA polymerase II promoter (GO:0042795)</th>\n",
       "      <td>71</td>\n",
       "      <td>0.17</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.052051e-12</td>\n",
       "      <td>4.989211e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcription elongation from RNA polymerase II promoter (GO:0006368)</th>\n",
       "      <td>96</td>\n",
       "      <td>0.16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.052051e-12</td>\n",
       "      <td>4.989211e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcription initiation from RNA polymerase II promoter (GO:0006367)</th>\n",
       "      <td>161</td>\n",
       "      <td>0.14</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.052051e-12</td>\n",
       "      <td>4.989211e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNA-templated transcription, elongation (GO:0006354)</th>\n",
       "      <td>109</td>\n",
       "      <td>0.14</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.052051e-12</td>\n",
       "      <td>4.989211e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">without_gain</th>\n",
       "      <th>establishment of mitochondrion localization, microtubule-mediated (GO:0034643)</th>\n",
       "      <td>7</td>\n",
       "      <td>0.14</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.865762e-07</td>\n",
       "      <td>4.865762e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>positive regulation of mitochondrial membrane permeability involved in apoptotic process (GO:1902110)</th>\n",
       "      <td>9</td>\n",
       "      <td>0.11</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.865762e-07</td>\n",
       "      <td>4.865762e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>regulation of mitochondrion organization (GO:0010821)</th>\n",
       "      <td>49</td>\n",
       "      <td>0.10</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.865762e-07</td>\n",
       "      <td>4.976360e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mitochondrion organization (GO:0007005)</th>\n",
       "      <td>168</td>\n",
       "      <td>0.09</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.865762e-07</td>\n",
       "      <td>4.976360e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>organelle transport along microtubule (GO:0072384)</th>\n",
       "      <td>32</td>\n",
       "      <td>0.09</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.865762e-07</td>\n",
       "      <td>4.976360e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">without_loss</th>\n",
       "      <th>intracellular lipid transport (GO:0032365)</th>\n",
       "      <td>16</td>\n",
       "      <td>0.19</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.857542e-10</td>\n",
       "      <td>4.986906e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fatty acid transmembrane transport (GO:1902001)</th>\n",
       "      <td>12</td>\n",
       "      <td>0.17</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.857542e-10</td>\n",
       "      <td>4.982104e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>carnitine shuttle (GO:0006853)</th>\n",
       "      <td>12</td>\n",
       "      <td>0.17</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.857542e-10</td>\n",
       "      <td>4.982104e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>long-chain fatty acid transport (GO:0015909)</th>\n",
       "      <td>26</td>\n",
       "      <td>0.15</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.857542e-10</td>\n",
       "      <td>4.986906e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fatty acid metabolic process (GO:0006631)</th>\n",
       "      <td>107</td>\n",
       "      <td>0.12</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.857542e-10</td>\n",
       "      <td>4.986906e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"30\" valign=\"top\">ovarian</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">with_both</th>\n",
       "      <th>purine nucleoside monophosphate biosynthetic process (GO:0009127)</th>\n",
       "      <td>13</td>\n",
       "      <td>0.15</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.146675e-02</td>\n",
       "      <td>4.966365e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ribonucleoside monophosphate biosynthetic process (GO:0009156)</th>\n",
       "      <td>15</td>\n",
       "      <td>0.13</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.146675e-02</td>\n",
       "      <td>4.966365e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>purine ribonucleoside monophosphate biosynthetic process (GO:0009168)</th>\n",
       "      <td>40</td>\n",
       "      <td>0.10</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.146675e-02</td>\n",
       "      <td>4.966365e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>purine ribonucleoside monophosphate metabolic process (GO:0009167)</th>\n",
       "      <td>40</td>\n",
       "      <td>0.05</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.146675e-02</td>\n",
       "      <td>4.966365e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hemidesmosome assembly (GO:0031581)</th>\n",
       "      <td>12</td>\n",
       "      <td>0.25</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.045690e-02</td>\n",
       "      <td>4.966365e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">with_gain</th>\n",
       "      <th>coenzyme A metabolic process (GO:0015936)</th>\n",
       "      <td>11</td>\n",
       "      <td>0.18</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.965082e-04</td>\n",
       "      <td>4.993389e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ribonucleoside bisphosphate biosynthetic process (GO:0034030)</th>\n",
       "      <td>8</td>\n",
       "      <td>0.12</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.965082e-04</td>\n",
       "      <td>7.965082e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>purine nucleoside bisphosphate biosynthetic process (GO:0034033)</th>\n",
       "      <td>8</td>\n",
       "      <td>0.12</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.965082e-04</td>\n",
       "      <td>7.965082e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coenzyme A biosynthetic process (GO:0015937)</th>\n",
       "      <td>8</td>\n",
       "      <td>0.12</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.965082e-04</td>\n",
       "      <td>7.965082e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coenzyme biosynthetic process (GO:0009108)</th>\n",
       "      <td>30</td>\n",
       "      <td>0.10</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.965082e-04</td>\n",
       "      <td>4.993389e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">with_loss</th>\n",
       "      <th>regulation of response to cytokine stimulus (GO:0060759)</th>\n",
       "      <td>21</td>\n",
       "      <td>0.33</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.194165e-04</td>\n",
       "      <td>4.979121e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>positive regulation of response to cytokine stimulus (GO:0060760)</th>\n",
       "      <td>15</td>\n",
       "      <td>0.33</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.194165e-04</td>\n",
       "      <td>4.979121e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>regulation of interferon-alpha secretion (GO:1902739)</th>\n",
       "      <td>7</td>\n",
       "      <td>0.29</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.194165e-04</td>\n",
       "      <td>4.979121e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>positive regulation of interferon-alpha secretion (GO:1902741)</th>\n",
       "      <td>7</td>\n",
       "      <td>0.29</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.194165e-04</td>\n",
       "      <td>4.979121e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>positive regulation of tumor necrosis factor secretion (GO:1904469)</th>\n",
       "      <td>12</td>\n",
       "      <td>0.25</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.194165e-04</td>\n",
       "      <td>4.979121e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">without_both</th>\n",
       "      <th>retrograde vesicle-mediated transport, Golgi to ER (GO:0006890)</th>\n",
       "      <td>81</td>\n",
       "      <td>0.20</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.682340e-04</td>\n",
       "      <td>4.991475e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COPII-coated vesicle budding (GO:0090114)</th>\n",
       "      <td>69</td>\n",
       "      <td>0.17</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.682340e-04</td>\n",
       "      <td>4.991475e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Golgi vesicle transport (GO:0048193)</th>\n",
       "      <td>272</td>\n",
       "      <td>0.16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.682340e-04</td>\n",
       "      <td>4.991475e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ER to Golgi vesicle-mediated transport (GO:0006888)</th>\n",
       "      <td>181</td>\n",
       "      <td>0.16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.682340e-04</td>\n",
       "      <td>4.991475e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>antigen processing and presentation of exogenous peptide antigen via MHC class I (GO:0042590)</th>\n",
       "      <td>79</td>\n",
       "      <td>0.14</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.682340e-04</td>\n",
       "      <td>4.991475e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">without_gain</th>\n",
       "      <th>misfolded or incompletely synthesized protein catabolic process (GO:0006515)</th>\n",
       "      <td>26</td>\n",
       "      <td>0.15</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.511367e-03</td>\n",
       "      <td>4.999259e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cellular response to misfolded protein (GO:0071218)</th>\n",
       "      <td>22</td>\n",
       "      <td>0.14</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.511367e-03</td>\n",
       "      <td>4.999259e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ER-associated misfolded protein catabolic process (GO:0071712)</th>\n",
       "      <td>14</td>\n",
       "      <td>0.14</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.511367e-03</td>\n",
       "      <td>4.999259e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>endoplasmic reticulum mannose trimming (GO:1904380)</th>\n",
       "      <td>20</td>\n",
       "      <td>0.10</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.511367e-03</td>\n",
       "      <td>4.999259e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein alpha-1,2-demannosylation (GO:0036508)</th>\n",
       "      <td>25</td>\n",
       "      <td>0.08</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.511367e-03</td>\n",
       "      <td>4.999259e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">without_loss</th>\n",
       "      <th>regulation of DNA-templated transcription, elongation (GO:0032784)</th>\n",
       "      <td>21</td>\n",
       "      <td>0.14</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.936456e-06</td>\n",
       "      <td>4.996409e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>negative regulation of proteasomal ubiquitin-dependent protein catabolic process (GO:0032435)</th>\n",
       "      <td>25</td>\n",
       "      <td>0.12</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.936456e-06</td>\n",
       "      <td>4.996409e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>negative regulation of intrinsic apoptotic signaling pathway in response to DNA damage (GO:1902230)</th>\n",
       "      <td>25</td>\n",
       "      <td>0.12</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.936456e-06</td>\n",
       "      <td>4.996409e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>regulation of intrinsic apoptotic signaling pathway in response to DNA damage (GO:1902229)</th>\n",
       "      <td>17</td>\n",
       "      <td>0.12</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.936456e-06</td>\n",
       "      <td>4.996409e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>negative regulation of catalytic activity (GO:0043086)</th>\n",
       "      <td>27</td>\n",
       "      <td>0.11</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.936456e-06</td>\n",
       "      <td>4.996409e-02</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>150 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                             Size  \\\n",
       "cancer_type group        Name                                                       \n",
       "colon       with_both    protein export from nucleus (GO:0006611)              29   \n",
       "                         intracellular protein transport (GO:0006886)         348   \n",
       "                         nuclear export (GO:0051168)                           93   \n",
       "                         mitotic nuclear envelope reassembly (GO:0007084)      11   \n",
       "                         peptidyl-serine dephosphorylation (GO:0070262)        11   \n",
       "            with_gain    Golgi vesicle budding (GO:0048194)                    10   \n",
       "                         positive regulation of viral genome replication...    32   \n",
       "                         regulation of viral genome replication (GO:0045...    64   \n",
       "                         positive regulation of viral life cycle (GO:190...    45   \n",
       "                         modification by symbiont of host morphology or ...     8   \n",
       "            with_loss    intracellular protein transport (GO:0006886)         348   \n",
       "                         protein export from nucleus (GO:0006611)              29   \n",
       "                         nuclear export (GO:0051168)                           93   \n",
       "                         positive regulation of exocytosis (GO:0045921)        35   \n",
       "                         positive regulation of secretion by cell (GO:19...    53   \n",
       "            without_both snRNA transcription (GO:0009301)                      71   \n",
       "                         snRNA transcription from RNA polymerase II prom...    71   \n",
       "                         transcription elongation from RNA polymerase II...    96   \n",
       "                         transcription initiation from RNA polymerase II...   161   \n",
       "                         DNA-templated transcription, elongation (GO:000...   109   \n",
       "            without_gain establishment of mitochondrion localization, mi...     7   \n",
       "                         positive regulation of mitochondrial membrane p...     9   \n",
       "                         regulation of mitochondrion organization (GO:00...    49   \n",
       "                         mitochondrion organization (GO:0007005)              168   \n",
       "                         organelle transport along microtubule (GO:0072384)    32   \n",
       "            without_loss intracellular lipid transport (GO:0032365)            16   \n",
       "                         fatty acid transmembrane transport (GO:1902001)       12   \n",
       "                         carnitine shuttle (GO:0006853)                        12   \n",
       "                         long-chain fatty acid transport (GO:0015909)          26   \n",
       "                         fatty acid metabolic process (GO:0006631)            107   \n",
       "...                                                                           ...   \n",
       "ovarian     with_both    purine nucleoside monophosphate biosynthetic pr...    13   \n",
       "                         ribonucleoside monophosphate biosynthetic proce...    15   \n",
       "                         purine ribonucleoside monophosphate biosyntheti...    40   \n",
       "                         purine ribonucleoside monophosphate metabolic p...    40   \n",
       "                         hemidesmosome assembly (GO:0031581)                   12   \n",
       "            with_gain    coenzyme A metabolic process (GO:0015936)             11   \n",
       "                         ribonucleoside bisphosphate biosynthetic proces...     8   \n",
       "                         purine nucleoside bisphosphate biosynthetic pro...     8   \n",
       "                         coenzyme A biosynthetic process (GO:0015937)           8   \n",
       "                         coenzyme biosynthetic process (GO:0009108)            30   \n",
       "            with_loss    regulation of response to cytokine stimulus (GO...    21   \n",
       "                         positive regulation of response to cytokine sti...    15   \n",
       "                         regulation of interferon-alpha secretion (GO:19...     7   \n",
       "                         positive regulation of interferon-alpha secreti...     7   \n",
       "                         positive regulation of tumor necrosis factor se...    12   \n",
       "            without_both retrograde vesicle-mediated transport, Golgi to...    81   \n",
       "                         COPII-coated vesicle budding (GO:0090114)             69   \n",
       "                         Golgi vesicle transport (GO:0048193)                 272   \n",
       "                         ER to Golgi vesicle-mediated transport (GO:0006...   181   \n",
       "                         antigen processing and presentation of exogenou...    79   \n",
       "            without_gain misfolded or incompletely synthesized protein c...    26   \n",
       "                         cellular response to misfolded protein (GO:0071...    22   \n",
       "                         ER-associated misfolded protein catabolic proce...    14   \n",
       "                         endoplasmic reticulum mannose trimming (GO:1904...    20   \n",
       "                         protein alpha-1,2-demannosylation (GO:0036508)        25   \n",
       "            without_loss regulation of DNA-templated transcription, elon...    21   \n",
       "                         negative regulation of proteasomal ubiquitin-de...    25   \n",
       "                         negative regulation of intrinsic apoptotic sign...    25   \n",
       "                         regulation of intrinsic apoptotic signaling pat...    17   \n",
       "                         negative regulation of catalytic activity (GO:0...    27   \n",
       "\n",
       "                                                                             Coverage  \\\n",
       "cancer_type group        Name                                                           \n",
       "colon       with_both    protein export from nucleus (GO:0006611)                0.03   \n",
       "                         intracellular protein transport (GO:0006886)            0.02   \n",
       "                         nuclear export (GO:0051168)                             0.02   \n",
       "                         mitotic nuclear envelope reassembly (GO:0007084)        0.18   \n",
       "                         peptidyl-serine dephosphorylation (GO:0070262)          0.18   \n",
       "            with_gain    Golgi vesicle budding (GO:0048194)                      0.20   \n",
       "                         positive regulation of viral genome replication...      0.16   \n",
       "                         regulation of viral genome replication (GO:0045...      0.14   \n",
       "                         positive regulation of viral life cycle (GO:190...      0.13   \n",
       "                         modification by symbiont of host morphology or ...      0.12   \n",
       "            with_loss    intracellular protein transport (GO:0006886)            0.03   \n",
       "                         protein export from nucleus (GO:0006611)                0.03   \n",
       "                         nuclear export (GO:0051168)                             0.02   \n",
       "                         positive regulation of exocytosis (GO:0045921)          0.14   \n",
       "                         positive regulation of secretion by cell (GO:19...      0.11   \n",
       "            without_both snRNA transcription (GO:0009301)                        0.17   \n",
       "                         snRNA transcription from RNA polymerase II prom...      0.17   \n",
       "                         transcription elongation from RNA polymerase II...      0.16   \n",
       "                         transcription initiation from RNA polymerase II...      0.14   \n",
       "                         DNA-templated transcription, elongation (GO:000...      0.14   \n",
       "            without_gain establishment of mitochondrion localization, mi...      0.14   \n",
       "                         positive regulation of mitochondrial membrane p...      0.11   \n",
       "                         regulation of mitochondrion organization (GO:00...      0.10   \n",
       "                         mitochondrion organization (GO:0007005)                 0.09   \n",
       "                         organelle transport along microtubule (GO:0072384)      0.09   \n",
       "            without_loss intracellular lipid transport (GO:0032365)              0.19   \n",
       "                         fatty acid transmembrane transport (GO:1902001)         0.17   \n",
       "                         carnitine shuttle (GO:0006853)                          0.17   \n",
       "                         long-chain fatty acid transport (GO:0015909)            0.15   \n",
       "                         fatty acid metabolic process (GO:0006631)               0.12   \n",
       "...                                                                               ...   \n",
       "ovarian     with_both    purine nucleoside monophosphate biosynthetic pr...      0.15   \n",
       "                         ribonucleoside monophosphate biosynthetic proce...      0.13   \n",
       "                         purine ribonucleoside monophosphate biosyntheti...      0.10   \n",
       "                         purine ribonucleoside monophosphate metabolic p...      0.05   \n",
       "                         hemidesmosome assembly (GO:0031581)                     0.25   \n",
       "            with_gain    coenzyme A metabolic process (GO:0015936)               0.18   \n",
       "                         ribonucleoside bisphosphate biosynthetic proces...      0.12   \n",
       "                         purine nucleoside bisphosphate biosynthetic pro...      0.12   \n",
       "                         coenzyme A biosynthetic process (GO:0015937)            0.12   \n",
       "                         coenzyme biosynthetic process (GO:0009108)              0.10   \n",
       "            with_loss    regulation of response to cytokine stimulus (GO...      0.33   \n",
       "                         positive regulation of response to cytokine sti...      0.33   \n",
       "                         regulation of interferon-alpha secretion (GO:19...      0.29   \n",
       "                         positive regulation of interferon-alpha secreti...      0.29   \n",
       "                         positive regulation of tumor necrosis factor se...      0.25   \n",
       "            without_both retrograde vesicle-mediated transport, Golgi to...      0.20   \n",
       "                         COPII-coated vesicle budding (GO:0090114)               0.17   \n",
       "                         Golgi vesicle transport (GO:0048193)                    0.16   \n",
       "                         ER to Golgi vesicle-mediated transport (GO:0006...      0.16   \n",
       "                         antigen processing and presentation of exogenou...      0.14   \n",
       "            without_gain misfolded or incompletely synthesized protein c...      0.15   \n",
       "                         cellular response to misfolded protein (GO:0071...      0.14   \n",
       "                         ER-associated misfolded protein catabolic proce...      0.14   \n",
       "                         endoplasmic reticulum mannose trimming (GO:1904...      0.10   \n",
       "                         protein alpha-1,2-demannosylation (GO:0036508)          0.08   \n",
       "            without_loss regulation of DNA-templated transcription, elon...      0.14   \n",
       "                         negative regulation of proteasomal ubiquitin-de...      0.12   \n",
       "                         negative regulation of intrinsic apoptotic sign...      0.12   \n",
       "                         regulation of intrinsic apoptotic signaling pat...      0.12   \n",
       "                         negative regulation of catalytic activity (GO:0...      0.11   \n",
       "\n",
       "                                                                             TDP.bound  \\\n",
       "cancer_type group        Name                                                            \n",
       "colon       with_both    protein export from nucleus (GO:0006611)                  1.0   \n",
       "                         intracellular protein transport (GO:0006886)              1.0   \n",
       "                         nuclear export (GO:0051168)                               1.0   \n",
       "                         mitotic nuclear envelope reassembly (GO:0007084)          1.0   \n",
       "                         peptidyl-serine dephosphorylation (GO:0070262)            1.0   \n",
       "            with_gain    Golgi vesicle budding (GO:0048194)                        1.0   \n",
       "                         positive regulation of viral genome replication...        1.0   \n",
       "                         regulation of viral genome replication (GO:0045...        1.0   \n",
       "                         positive regulation of viral life cycle (GO:190...        1.0   \n",
       "                         modification by symbiont of host morphology or ...        1.0   \n",
       "            with_loss    intracellular protein transport (GO:0006886)              1.0   \n",
       "                         protein export from nucleus (GO:0006611)                  1.0   \n",
       "                         nuclear export (GO:0051168)                               1.0   \n",
       "                         positive regulation of exocytosis (GO:0045921)            1.0   \n",
       "                         positive regulation of secretion by cell (GO:19...        1.0   \n",
       "            without_both snRNA transcription (GO:0009301)                          1.0   \n",
       "                         snRNA transcription from RNA polymerase II prom...        1.0   \n",
       "                         transcription elongation from RNA polymerase II...        1.0   \n",
       "                         transcription initiation from RNA polymerase II...        1.0   \n",
       "                         DNA-templated transcription, elongation (GO:000...        1.0   \n",
       "            without_gain establishment of mitochondrion localization, mi...        1.0   \n",
       "                         positive regulation of mitochondrial membrane p...        1.0   \n",
       "                         regulation of mitochondrion organization (GO:00...        1.0   \n",
       "                         mitochondrion organization (GO:0007005)                   1.0   \n",
       "                         organelle transport along microtubule (GO:0072384)        1.0   \n",
       "            without_loss intracellular lipid transport (GO:0032365)                1.0   \n",
       "                         fatty acid transmembrane transport (GO:1902001)           1.0   \n",
       "                         carnitine shuttle (GO:0006853)                            1.0   \n",
       "                         long-chain fatty acid transport (GO:0015909)              1.0   \n",
       "                         fatty acid metabolic process (GO:0006631)                 1.0   \n",
       "...                                                                                ...   \n",
       "ovarian     with_both    purine nucleoside monophosphate biosynthetic pr...        1.0   \n",
       "                         ribonucleoside monophosphate biosynthetic proce...        1.0   \n",
       "                         purine ribonucleoside monophosphate biosyntheti...        1.0   \n",
       "                         purine ribonucleoside monophosphate metabolic p...        1.0   \n",
       "                         hemidesmosome assembly (GO:0031581)                       1.0   \n",
       "            with_gain    coenzyme A metabolic process (GO:0015936)                 1.0   \n",
       "                         ribonucleoside bisphosphate biosynthetic proces...        1.0   \n",
       "                         purine nucleoside bisphosphate biosynthetic pro...        1.0   \n",
       "                         coenzyme A biosynthetic process (GO:0015937)              1.0   \n",
       "                         coenzyme biosynthetic process (GO:0009108)                1.0   \n",
       "            with_loss    regulation of response to cytokine stimulus (GO...        1.0   \n",
       "                         positive regulation of response to cytokine sti...        1.0   \n",
       "                         regulation of interferon-alpha secretion (GO:19...        1.0   \n",
       "                         positive regulation of interferon-alpha secreti...        1.0   \n",
       "                         positive regulation of tumor necrosis factor se...        1.0   \n",
       "            without_both retrograde vesicle-mediated transport, Golgi to...        1.0   \n",
       "                         COPII-coated vesicle budding (GO:0090114)                 1.0   \n",
       "                         Golgi vesicle transport (GO:0048193)                      1.0   \n",
       "                         ER to Golgi vesicle-mediated transport (GO:0006...        1.0   \n",
       "                         antigen processing and presentation of exogenou...        1.0   \n",
       "            without_gain misfolded or incompletely synthesized protein c...        1.0   \n",
       "                         cellular response to misfolded protein (GO:0071...        1.0   \n",
       "                         ER-associated misfolded protein catabolic proce...        1.0   \n",
       "                         endoplasmic reticulum mannose trimming (GO:1904...        1.0   \n",
       "                         protein alpha-1,2-demannosylation (GO:0036508)            1.0   \n",
       "            without_loss regulation of DNA-templated transcription, elon...        1.0   \n",
       "                         negative regulation of proteasomal ubiquitin-de...        1.0   \n",
       "                         negative regulation of intrinsic apoptotic sign...        1.0   \n",
       "                         regulation of intrinsic apoptotic signaling pat...        1.0   \n",
       "                         negative regulation of catalytic activity (GO:0...        1.0   \n",
       "\n",
       "                                                                             TDP.estimate  \\\n",
       "cancer_type group        Name                                                               \n",
       "colon       with_both    protein export from nucleus (GO:0006611)                     1.0   \n",
       "                         intracellular protein transport (GO:0006886)                 1.0   \n",
       "                         nuclear export (GO:0051168)                                  1.0   \n",
       "                         mitotic nuclear envelope reassembly (GO:0007084)             1.0   \n",
       "                         peptidyl-serine dephosphorylation (GO:0070262)               1.0   \n",
       "            with_gain    Golgi vesicle budding (GO:0048194)                           1.0   \n",
       "                         positive regulation of viral genome replication...           1.0   \n",
       "                         regulation of viral genome replication (GO:0045...           1.0   \n",
       "                         positive regulation of viral life cycle (GO:190...           1.0   \n",
       "                         modification by symbiont of host morphology or ...           1.0   \n",
       "            with_loss    intracellular protein transport (GO:0006886)                 1.0   \n",
       "                         protein export from nucleus (GO:0006611)                     1.0   \n",
       "                         nuclear export (GO:0051168)                                  1.0   \n",
       "                         positive regulation of exocytosis (GO:0045921)               1.0   \n",
       "                         positive regulation of secretion by cell (GO:19...           1.0   \n",
       "            without_both snRNA transcription (GO:0009301)                             1.0   \n",
       "                         snRNA transcription from RNA polymerase II prom...           1.0   \n",
       "                         transcription elongation from RNA polymerase II...           1.0   \n",
       "                         transcription initiation from RNA polymerase II...           1.0   \n",
       "                         DNA-templated transcription, elongation (GO:000...           1.0   \n",
       "            without_gain establishment of mitochondrion localization, mi...           1.0   \n",
       "                         positive regulation of mitochondrial membrane p...           1.0   \n",
       "                         regulation of mitochondrion organization (GO:00...           1.0   \n",
       "                         mitochondrion organization (GO:0007005)                      1.0   \n",
       "                         organelle transport along microtubule (GO:0072384)           1.0   \n",
       "            without_loss intracellular lipid transport (GO:0032365)                   1.0   \n",
       "                         fatty acid transmembrane transport (GO:1902001)              1.0   \n",
       "                         carnitine shuttle (GO:0006853)                               1.0   \n",
       "                         long-chain fatty acid transport (GO:0015909)                 1.0   \n",
       "                         fatty acid metabolic process (GO:0006631)                    1.0   \n",
       "...                                                                                   ...   \n",
       "ovarian     with_both    purine nucleoside monophosphate biosynthetic pr...           1.0   \n",
       "                         ribonucleoside monophosphate biosynthetic proce...           1.0   \n",
       "                         purine ribonucleoside monophosphate biosyntheti...           1.0   \n",
       "                         purine ribonucleoside monophosphate metabolic p...           1.0   \n",
       "                         hemidesmosome assembly (GO:0031581)                          1.0   \n",
       "            with_gain    coenzyme A metabolic process (GO:0015936)                    1.0   \n",
       "                         ribonucleoside bisphosphate biosynthetic proces...           1.0   \n",
       "                         purine nucleoside bisphosphate biosynthetic pro...           1.0   \n",
       "                         coenzyme A biosynthetic process (GO:0015937)                 1.0   \n",
       "                         coenzyme biosynthetic process (GO:0009108)                   1.0   \n",
       "            with_loss    regulation of response to cytokine stimulus (GO...           1.0   \n",
       "                         positive regulation of response to cytokine sti...           1.0   \n",
       "                         regulation of interferon-alpha secretion (GO:19...           1.0   \n",
       "                         positive regulation of interferon-alpha secreti...           1.0   \n",
       "                         positive regulation of tumor necrosis factor se...           1.0   \n",
       "            without_both retrograde vesicle-mediated transport, Golgi to...           1.0   \n",
       "                         COPII-coated vesicle budding (GO:0090114)                    1.0   \n",
       "                         Golgi vesicle transport (GO:0048193)                         1.0   \n",
       "                         ER to Golgi vesicle-mediated transport (GO:0006...           1.0   \n",
       "                         antigen processing and presentation of exogenou...           1.0   \n",
       "            without_gain misfolded or incompletely synthesized protein c...           1.0   \n",
       "                         cellular response to misfolded protein (GO:0071...           1.0   \n",
       "                         ER-associated misfolded protein catabolic proce...           1.0   \n",
       "                         endoplasmic reticulum mannose trimming (GO:1904...           1.0   \n",
       "                         protein alpha-1,2-demannosylation (GO:0036508)               1.0   \n",
       "            without_loss regulation of DNA-templated transcription, elon...           1.0   \n",
       "                         negative regulation of proteasomal ubiquitin-de...           1.0   \n",
       "                         negative regulation of intrinsic apoptotic sign...           1.0   \n",
       "                         regulation of intrinsic apoptotic signaling pat...           1.0   \n",
       "                         negative regulation of catalytic activity (GO:0...           1.0   \n",
       "\n",
       "                                                                                  SC.adjP  \\\n",
       "cancer_type group        Name                                                               \n",
       "colon       with_both    protein export from nucleus (GO:0006611)            6.644517e-07   \n",
       "                         intracellular protein transport (GO:0006886)        6.644517e-07   \n",
       "                         nuclear export (GO:0051168)                         6.644517e-07   \n",
       "                         mitotic nuclear envelope reassembly (GO:0007084)    2.536441e-06   \n",
       "                         peptidyl-serine dephosphorylation (GO:0070262)      2.536441e-06   \n",
       "            with_gain    Golgi vesicle budding (GO:0048194)                  4.866931e-09   \n",
       "                         positive regulation of viral genome replication...  4.866931e-09   \n",
       "                         regulation of viral genome replication (GO:0045...  4.866931e-09   \n",
       "                         positive regulation of viral life cycle (GO:190...  4.866931e-09   \n",
       "                         modification by symbiont of host morphology or ...  4.866931e-09   \n",
       "            with_loss    intracellular protein transport (GO:0006886)        3.042659e-09   \n",
       "                         protein export from nucleus (GO:0006611)            3.042659e-09   \n",
       "                         nuclear export (GO:0051168)                         3.042659e-09   \n",
       "                         positive regulation of exocytosis (GO:0045921)      1.263430e-07   \n",
       "                         positive regulation of secretion by cell (GO:19...  1.263430e-07   \n",
       "            without_both snRNA transcription (GO:0009301)                    2.052051e-12   \n",
       "                         snRNA transcription from RNA polymerase II prom...  2.052051e-12   \n",
       "                         transcription elongation from RNA polymerase II...  2.052051e-12   \n",
       "                         transcription initiation from RNA polymerase II...  2.052051e-12   \n",
       "                         DNA-templated transcription, elongation (GO:000...  2.052051e-12   \n",
       "            without_gain establishment of mitochondrion localization, mi...  4.865762e-07   \n",
       "                         positive regulation of mitochondrial membrane p...  4.865762e-07   \n",
       "                         regulation of mitochondrion organization (GO:00...  4.865762e-07   \n",
       "                         mitochondrion organization (GO:0007005)             4.865762e-07   \n",
       "                         organelle transport along microtubule (GO:0072384)  4.865762e-07   \n",
       "            without_loss intracellular lipid transport (GO:0032365)          3.857542e-10   \n",
       "                         fatty acid transmembrane transport (GO:1902001)     3.857542e-10   \n",
       "                         carnitine shuttle (GO:0006853)                      3.857542e-10   \n",
       "                         long-chain fatty acid transport (GO:0015909)        3.857542e-10   \n",
       "                         fatty acid metabolic process (GO:0006631)           3.857542e-10   \n",
       "...                                                                                   ...   \n",
       "ovarian     with_both    purine nucleoside monophosphate biosynthetic pr...  2.146675e-02   \n",
       "                         ribonucleoside monophosphate biosynthetic proce...  2.146675e-02   \n",
       "                         purine ribonucleoside monophosphate biosyntheti...  2.146675e-02   \n",
       "                         purine ribonucleoside monophosphate metabolic p...  2.146675e-02   \n",
       "                         hemidesmosome assembly (GO:0031581)                 3.045690e-02   \n",
       "            with_gain    coenzyme A metabolic process (GO:0015936)           7.965082e-04   \n",
       "                         ribonucleoside bisphosphate biosynthetic proces...  7.965082e-04   \n",
       "                         purine nucleoside bisphosphate biosynthetic pro...  7.965082e-04   \n",
       "                         coenzyme A biosynthetic process (GO:0015937)        7.965082e-04   \n",
       "                         coenzyme biosynthetic process (GO:0009108)          7.965082e-04   \n",
       "            with_loss    regulation of response to cytokine stimulus (GO...  4.194165e-04   \n",
       "                         positive regulation of response to cytokine sti...  4.194165e-04   \n",
       "                         regulation of interferon-alpha secretion (GO:19...  4.194165e-04   \n",
       "                         positive regulation of interferon-alpha secreti...  4.194165e-04   \n",
       "                         positive regulation of tumor necrosis factor se...  4.194165e-04   \n",
       "            without_both retrograde vesicle-mediated transport, Golgi to...  1.682340e-04   \n",
       "                         COPII-coated vesicle budding (GO:0090114)           1.682340e-04   \n",
       "                         Golgi vesicle transport (GO:0048193)                1.682340e-04   \n",
       "                         ER to Golgi vesicle-mediated transport (GO:0006...  1.682340e-04   \n",
       "                         antigen processing and presentation of exogenou...  1.682340e-04   \n",
       "            without_gain misfolded or incompletely synthesized protein c...  4.511367e-03   \n",
       "                         cellular response to misfolded protein (GO:0071...  4.511367e-03   \n",
       "                         ER-associated misfolded protein catabolic proce...  4.511367e-03   \n",
       "                         endoplasmic reticulum mannose trimming (GO:1904...  4.511367e-03   \n",
       "                         protein alpha-1,2-demannosylation (GO:0036508)      4.511367e-03   \n",
       "            without_loss regulation of DNA-templated transcription, elon...  7.936456e-06   \n",
       "                         negative regulation of proteasomal ubiquitin-de...  7.936456e-06   \n",
       "                         negative regulation of intrinsic apoptotic sign...  7.936456e-06   \n",
       "                         regulation of intrinsic apoptotic signaling pat...  7.936456e-06   \n",
       "                         negative regulation of catalytic activity (GO:0...  7.936456e-06   \n",
       "\n",
       "                                                                             Comp.0.99.adjP  \n",
       "cancer_type group        Name                                                                \n",
       "colon       with_both    protein export from nucleus (GO:0006611)              6.644517e-07  \n",
       "                         intracellular protein transport (GO:0006886)          4.999214e-02  \n",
       "                         nuclear export (GO:0051168)                           4.999214e-02  \n",
       "                         mitotic nuclear envelope reassembly (GO:0007084)      4.999214e-02  \n",
       "                         peptidyl-serine dephosphorylation (GO:0070262)        4.999214e-02  \n",
       "            with_gain    Golgi vesicle budding (GO:0048194)                    4.996401e-02  \n",
       "                         positive regulation of viral genome replication...    4.996401e-02  \n",
       "                         regulation of viral genome replication (GO:0045...    4.996401e-02  \n",
       "                         positive regulation of viral life cycle (GO:190...    4.996401e-02  \n",
       "                         modification by symbiont of host morphology or ...    4.866931e-09  \n",
       "            with_loss    intracellular protein transport (GO:0006886)          4.979156e-02  \n",
       "                         protein export from nucleus (GO:0006611)              3.042659e-09  \n",
       "                         nuclear export (GO:0051168)                           4.979156e-02  \n",
       "                         positive regulation of exocytosis (GO:0045921)        4.979156e-02  \n",
       "                         positive regulation of secretion by cell (GO:19...    4.979156e-02  \n",
       "            without_both snRNA transcription (GO:0009301)                      4.989211e-02  \n",
       "                         snRNA transcription from RNA polymerase II prom...    4.989211e-02  \n",
       "                         transcription elongation from RNA polymerase II...    4.989211e-02  \n",
       "                         transcription initiation from RNA polymerase II...    4.989211e-02  \n",
       "                         DNA-templated transcription, elongation (GO:000...    4.989211e-02  \n",
       "            without_gain establishment of mitochondrion localization, mi...    4.865762e-07  \n",
       "                         positive regulation of mitochondrial membrane p...    4.865762e-07  \n",
       "                         regulation of mitochondrion organization (GO:00...    4.976360e-02  \n",
       "                         mitochondrion organization (GO:0007005)               4.976360e-02  \n",
       "                         organelle transport along microtubule (GO:0072384)    4.976360e-02  \n",
       "            without_loss intracellular lipid transport (GO:0032365)            4.986906e-02  \n",
       "                         fatty acid transmembrane transport (GO:1902001)       4.982104e-02  \n",
       "                         carnitine shuttle (GO:0006853)                        4.982104e-02  \n",
       "                         long-chain fatty acid transport (GO:0015909)          4.986906e-02  \n",
       "                         fatty acid metabolic process (GO:0006631)             4.986906e-02  \n",
       "...                                                                                     ...  \n",
       "ovarian     with_both    purine nucleoside monophosphate biosynthetic pr...    4.966365e-02  \n",
       "                         ribonucleoside monophosphate biosynthetic proce...    4.966365e-02  \n",
       "                         purine ribonucleoside monophosphate biosyntheti...    4.966365e-02  \n",
       "                         purine ribonucleoside monophosphate metabolic p...    4.966365e-02  \n",
       "                         hemidesmosome assembly (GO:0031581)                   4.966365e-02  \n",
       "            with_gain    coenzyme A metabolic process (GO:0015936)             4.993389e-02  \n",
       "                         ribonucleoside bisphosphate biosynthetic proces...    7.965082e-04  \n",
       "                         purine nucleoside bisphosphate biosynthetic pro...    7.965082e-04  \n",
       "                         coenzyme A biosynthetic process (GO:0015937)          7.965082e-04  \n",
       "                         coenzyme biosynthetic process (GO:0009108)            4.993389e-02  \n",
       "            with_loss    regulation of response to cytokine stimulus (GO...    4.979121e-02  \n",
       "                         positive regulation of response to cytokine sti...    4.979121e-02  \n",
       "                         regulation of interferon-alpha secretion (GO:19...    4.979121e-02  \n",
       "                         positive regulation of interferon-alpha secreti...    4.979121e-02  \n",
       "                         positive regulation of tumor necrosis factor se...    4.979121e-02  \n",
       "            without_both retrograde vesicle-mediated transport, Golgi to...    4.991475e-02  \n",
       "                         COPII-coated vesicle budding (GO:0090114)             4.991475e-02  \n",
       "                         Golgi vesicle transport (GO:0048193)                  4.991475e-02  \n",
       "                         ER to Golgi vesicle-mediated transport (GO:0006...    4.991475e-02  \n",
       "                         antigen processing and presentation of exogenou...    4.991475e-02  \n",
       "            without_gain misfolded or incompletely synthesized protein c...    4.999259e-02  \n",
       "                         cellular response to misfolded protein (GO:0071...    4.999259e-02  \n",
       "                         ER-associated misfolded protein catabolic proce...    4.999259e-02  \n",
       "                         endoplasmic reticulum mannose trimming (GO:1904...    4.999259e-02  \n",
       "                         protein alpha-1,2-demannosylation (GO:0036508)        4.999259e-02  \n",
       "            without_loss regulation of DNA-templated transcription, elon...    4.996409e-02  \n",
       "                         negative regulation of proteasomal ubiquitin-de...    4.996409e-02  \n",
       "                         negative regulation of intrinsic apoptotic sign...    4.996409e-02  \n",
       "                         regulation of intrinsic apoptotic signaling pat...    4.996409e-02  \n",
       "                         negative regulation of catalytic activity (GO:0...    4.996409e-02  \n",
       "\n",
       "[150 rows x 6 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results[\"go\"].groupby([\"cancer_type\", \"group\"]).head().set_index([\"cancer_type\", \"group\", \"Name\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Look at which gene sets are most enriched across cancer types, for each group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_top_ten(\n",
    "    enrich_file_path, \n",
    "    expr_file_path, \n",
    "    gmt_file_path, \n",
    "    enrichment_group,\n",
    "    xtitle, \n",
    "    pathway_min_size, \n",
    "    min_cancers_enriched,\n",
    "    sort_col, \n",
    "    sort_asc, \n",
    "    cutoff_col, \n",
    "    cutoff=0.05\n",
    "):\n",
    "    \"\"\"Make a bubble chart where the x axis is pathway mean rank across all cancer types, the \n",
    "    y axis is cancer type, circle size is enrichment rank in cancer type, and circle color is\n",
    "    gene set mean change in expression in tumor compared to normal.\n",
    "    \n",
    "    Parameters:\n",
    "    enrich_file_path (str): The path to the file with rSEA enrichment analysis results.\n",
    "    expr_file_path (str): The path to the file with differential expression analysis results.\n",
    "    gmt_file_path (str): The path to the .gmt file that was used for the rSEA analysis.\n",
    "    enrichment_group (str): Which CNV event group to look at. Either 'with_gain', 'without_gain', \n",
    "        'with_loss', 'without_loss', 'with_both', or 'without_both'. \n",
    "    xtitle (str): The title for the plot.\n",
    "    pathway_min_size (int): The minimum number of genes a pathway/gene set must have to be included\n",
    "        in the chart.\n",
    "    min_cancers_enriched (int): The minimum number of cancers a pathway/gene set must be enriched\n",
    "        in, in order to be included in the chart.\n",
    "    sort_col (str): Which column in the rSEA results dataframe to use for ranking pathways.\n",
    "    sort_asc (bool): Whether to rank the sort column in ascending order.\n",
    "    cutoff_col (str): Which column in the rSEA results to use for a significance cutoff.\n",
    "    cutoff (float, optional): What cutoff to use when deciding whether to count a pathway as enriched.\n",
    "    \n",
    "    Returns:\n",
    "    altair.Chart: The chart.\n",
    "    pandas.DataFrame: The enrichment data used for the chart\n",
    "    pandas.DataFrame: The expression data used for the chart\n",
    "    pandas.DataFrame: The summary of the enrichment data, created in the process of making the chart\n",
    "    \"\"\"\n",
    "    \n",
    "    # Read in the expression data, selecting only the data for the specified group\n",
    "    all_expression_data = pd.read_csv(expr_file_path, sep=\"\\t\")\n",
    "    \n",
    "    all_expression_data = all_expression_data[all_expression_data[\"group\"] == enrichment_group]\n",
    "\n",
    "    # Make a column where all increases are +1 and all decreases \n",
    "    # are -1, because these are ratioed abundances, so we can't \n",
    "    # compare magnitudes between genes--we can only compare whether \n",
    "    # there was a change, and whether it was positive or negative\n",
    "    all_expression_data = all_expression_data.assign(simplified_change=np.nan)\n",
    "\n",
    "    # adj p < 0.05 and change > 1 => +1\n",
    "    all_expression_data.loc[\n",
    "        (all_expression_data[\"change\"] > 0) & (all_expression_data[\"adj_p\"] < 0.05), \n",
    "        \"simplified_change\"\n",
    "    ] = 1\n",
    "\n",
    "    # adj p >= 0.05 and change > 1 => +0.5\n",
    "    all_expression_data.loc[(all_expression_data[\"change\"] > 0) & (all_expression_data[\"adj_p\"] >= 0.05),\n",
    "        \"simplified_change\"\n",
    "    ] = MSV\n",
    "\n",
    "    # change == 0 => 0\n",
    "    all_expression_data.loc[\n",
    "        all_expression_data[\"change\"] == 0,\n",
    "        \"simplified_change\"\n",
    "    ] = 0\n",
    "\n",
    "    # adj p >= 0.05 and change < 1 => -0.5\n",
    "    all_expression_data.loc[(all_expression_data[\"change\"] < 0) & (all_expression_data[\"adj_p\"] >= 0.05), \n",
    "        \"simplified_change\"\n",
    "    ] = -MSV\n",
    "\n",
    "    # adj p < 0.05 and change < 1 => -1\n",
    "    all_expression_data.loc[\n",
    "        (all_expression_data[\"change\"] < 0) & (all_expression_data[\"adj_p\"] < 0.05),\n",
    "        \"simplified_change\"\n",
    "    ] = -1\n",
    "\n",
    "    # Select just the proteins where we chose to reject the null hypothesis of no change\n",
    "    # We'll use this later to calculate average change in expression for each pathway\n",
    "    expression_data = all_expression_data[all_expression_data[\"adj_p\"] <= 0.05]\n",
    "    \n",
    "    # Read in the GMT file so we can have a list of which genes are in each pathway, in\n",
    "    # order to calculate the average change in expression for the pathways\n",
    "    with open(gmt_file_path, \"r\") as fp:\n",
    "        gene_lists = fp.readlines()\n",
    "\n",
    "    # Take the newline off the end, and split on the tab character to create a list of lists\n",
    "    gene_lists = [l.strip().split(\"\\t\") for l in gene_lists]\n",
    "\n",
    "    # Create a dataframe mapping pathway name and ID to contained genes\n",
    "    pathway_names = [l[0] for l in gene_lists]\n",
    "    pathway_genes = [l[2:] for l in gene_lists] # We skip index 1 -- it's a blank unused field.\n",
    "    pathway_data = pd.DataFrame({\n",
    "        \"pathway_name\": pathway_names,\n",
    "        \"pathway_genes\": pathway_genes\n",
    "    })\n",
    "\n",
    "    # Read in the enrichment data\n",
    "    enrichment_data = pd.\\\n",
    "    read_csv(enrich_file_path, sep=\"\\t\", index_col=0).\\\n",
    "    rename(columns={\"Name\": \"pathway_name\"})\n",
    "    \n",
    "    # Split the cancer_type_group column, then select the specified group\n",
    "    # This assumes that there are no underscores in the cancer type names,\n",
    "    # and that the cancer_type_group column has the format [CANCER]_[GROUP]\n",
    "    grp_split = enrichment_data[\"cancer_type_group\"].str.split(\"_\", n=1, expand=True)\n",
    "    \n",
    "    enrichment_data = enrichment_data.assign(\n",
    "        cancer_type=grp_split[0],\n",
    "        group=grp_split[1]\n",
    "    ).\\\n",
    "    drop(columns=\"cancer_type_group\")\n",
    "    \n",
    "    enrichment_data = enrichment_data[enrichment_data[\"group\"] == enrichment_group]\n",
    "    \n",
    "    # Merge pathway data into the enrichment data, so we know which pathways were enriched\n",
    "    enrichment_data = enrichment_data.merge(\n",
    "        pathway_data,\n",
    "        how=\"left\",\n",
    "        left_on=\"pathway_name\",\n",
    "        right_on=\"pathway_name\",\n",
    "        validate=\"many_to_one\"\n",
    "    )\n",
    "    \n",
    "    # Select enrichment data where pathways meet the minimum size and the\n",
    "    # p values for them being enriched pass the cutoff\n",
    "    enrichment_data = enrichment_data[\n",
    "        (enrichment_data[\"pathway_genes\"].apply(len) >= pathway_min_size) &\n",
    "        (enrichment_data[cutoff_col] <= cutoff)\n",
    "    ]\n",
    "    \n",
    "    # Assign pathway ranks within each cancer type based on the sort_col.\n",
    "    enrichment_data = enrichment_data.\\\n",
    "    assign(\n",
    "        cancer_rank=enrichment_data.\\\n",
    "        groupby(\"cancer_type\")[sort_col].\\\n",
    "        rank(ascending=sort_asc)\n",
    "    ).\\\n",
    "    sort_values(by=[\"cancer_type\", \"cancer_rank\"]).\\\n",
    "    reset_index(drop=True)\n",
    "\n",
    "    # Make a table with summary info for all pathways\n",
    "    enrichment_summary = enrichment_data[[\"pathway_name\"]].drop_duplicates(keep=\"first\")\n",
    "\n",
    "    pathway_times_enriched = enrichment_summary[\"pathway_name\"].apply(\n",
    "        lambda x: enrichment_data[enrichment_data[\"pathway_name\"] == x].shape[0])\n",
    "\n",
    "    avg_rank = enrichment_summary[\"pathway_name\"].apply(\n",
    "        lambda x: enrichment_data.loc[enrichment_data[\"pathway_name\"] == x, \"cancer_rank\"].mean())\n",
    "\n",
    "    enrichment_summary = enrichment_summary.\\\n",
    "    assign(\n",
    "        pathway_times_enriched=pathway_times_enriched,\n",
    "        pathway_avg_rank=avg_rank).\\\n",
    "    sort_values(\n",
    "        by=[\"pathway_times_enriched\", \"pathway_avg_rank\"],\n",
    "        ascending=[False, True]).\\\n",
    "    reset_index(drop=True)\n",
    "\n",
    "    # Merge the summary into the original enrichment data\n",
    "    enrichment_data = enrichment_data.\\\n",
    "    merge(\n",
    "        enrichment_summary,\n",
    "        how=\"outer\",\n",
    "        left_on=\"pathway_name\",\n",
    "        right_on=\"pathway_name\",\n",
    "        validate=\"many_to_one\"\n",
    "    ).\\\n",
    "    sort_values(\n",
    "        by=[\"pathway_times_enriched\", \"pathway_avg_rank\", \"cancer_type\"],\n",
    "        ascending=[False, True, True]\n",
    "    )\n",
    "\n",
    "    # Select top 10 for our plot\n",
    "    in_all = enrichment_summary.loc[\n",
    "        enrichment_summary[\"pathway_times_enriched\"] >= min_cancers_enriched,\n",
    "        \"pathway_name\"\n",
    "    ]\n",
    "    \n",
    "    if in_all.size <= 10:\n",
    "        top_ten = in_all\n",
    "    else:\n",
    "        top_ten = in_all[:10]\n",
    "    \n",
    "    sel_enrich = enrichment_data[enrichment_data[\"pathway_name\"].isin(top_ten)]\n",
    "\n",
    "    # Calculate the mean expression for each pathway in each cancer type\n",
    "    mean_exprs = []\n",
    "\n",
    "    for idx in sel_enrich.index:\n",
    "        genes = sel_enrich.loc[idx, \"pathway_genes\"]\n",
    "        cancer_type = sel_enrich.loc[idx, \"cancer_type\"]\n",
    "\n",
    "        genes_expr = expression_data.\\\n",
    "        loc[\n",
    "            expression_data[\"protein_str\"].isin(genes) &\n",
    "            (expression_data[\"cancer_type\"] == cancer_type),\n",
    "            \"simplified_change\"\n",
    "        ].\\\n",
    "        mean()\n",
    "\n",
    "        mean_exprs.append(genes_expr)\n",
    "\n",
    "    sel_enrich = sel_enrich.assign(mean_expr=mean_exprs)\n",
    "\n",
    "    sel_enrich = sel_enrich.assign(\n",
    "        rank_size=1 / sel_enrich[\"cancer_rank\"],\n",
    "        avg_rank_size=1 / sel_enrich[\"pathway_avg_rank\"],\n",
    "        avg_rank_label=sel_enrich[\"pathway_avg_rank\"].apply(lambda x: round(x, 2)).astype(str))\n",
    "    \n",
    "    # Take care of duplicates for the upper plot\n",
    "    for i in range(10):\n",
    "        sel_enrich[\"avg_rank_label\"] = sel_enrich[\"avg_rank_label\"].where(\n",
    "            cond=~(sel_enrich.duplicated(subset=[\"cancer_type\", \"avg_rank_label\"], keep=\"first\")),\n",
    "            other=\" \" + sel_enrich[\"avg_rank_label\"])\n",
    "\n",
    "    individual = alt.Chart(sel_enrich).mark_circle().encode(\n",
    "        x=alt.X(\n",
    "            \"pathway_name:N\",\n",
    "            sort=sel_enrich[\"pathway_name\"].values,\n",
    "            axis=alt.Axis(\n",
    "                labelAngle=-30,\n",
    "                labelFontSize=12,\n",
    "                labelLimit=500,\n",
    "                title=\"\",\n",
    "                titleFontSize=16\n",
    "            )\n",
    "        ),\n",
    "        y=alt.Y(\n",
    "            \"cancer_type:N\",\n",
    "            axis=alt.Axis(\n",
    "                title=\"Cancer type\",\n",
    "                titleFontSize=12\n",
    "            ),\n",
    "        ),\n",
    "        size=alt.Size(\n",
    "            \"rank_size:Q\",\n",
    "            legend=None\n",
    "        ),\n",
    "        color=alt.Color(\n",
    "            \"mean_expr:Q\",\n",
    "            scale=alt.Scale(\n",
    "                scheme=\"blueorange\",\n",
    "                domain=[-1, 1]\n",
    "            ),\n",
    "            legend=alt.Legend(\n",
    "                title=\"Pathway tumor expression\"\n",
    "            )\n",
    "        )\n",
    "    ).properties(\n",
    "        width=400,\n",
    "        height=300\n",
    "    )\n",
    "\n",
    "    aggregate = alt.Chart(sel_enrich).mark_circle().encode(\n",
    "        x=alt.X(\n",
    "            \"avg_rank_label:N\",\n",
    "            sort=sel_enrich[\"avg_rank_label\"].values,\n",
    "            axis=alt.Axis(\n",
    "                labelAngle=-30,\n",
    "                labelFontSize=12,\n",
    "                labelLimit=500,\n",
    "                title=\"Overall rank of pathway\",\n",
    "                titleFontSize=12\n",
    "            )\n",
    "        ),\n",
    "        size=alt.Size(\n",
    "            \"avg_rank_size:Q\",\n",
    "            legend=None\n",
    "        ),\n",
    "    ).properties(\n",
    "        width=400\n",
    "    )\n",
    "\n",
    "    full_plot = alt.vconcat(\n",
    "        aggregate, individual\n",
    "    ).properties(\n",
    "        title=xtitle\n",
    "    )\n",
    "    \n",
    "    return full_plot, enrichment_data, all_expression_data, enrichment_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_groups = [\n",
    "    'with_gain', \n",
    "    'without_gain', \n",
    "    'with_loss', \n",
    "    'without_loss', \n",
    "    'with_both', \n",
    "    'without_both'\n",
    "]\n",
    "\n",
    "reactome_plots = [\n",
    "    plot_top_ten(\n",
    "        enrich_file_path=reactome_results_file, \n",
    "        expr_file_path=diff_expr_file, \n",
    "        gmt_file_path=reactome_gmt,\n",
    "        enrichment_group=eg,\n",
    "        xtitle=f\"Reactome {eg} data - rSEA, threshold = 0.99, sort by Comp_adjP, cutoff Comp_adjP\",\n",
    "        pathway_min_size=5,\n",
    "        min_cancers_enriched=5,\n",
    "        sort_col=\"Comp.0.99.adjP\",\n",
    "        sort_asc=True,\n",
    "        cutoff_col=\"Comp.0.99.adjP\",\n",
    "        cutoff=0.05\n",
    "    )[0]\n",
    "    for eg in all_groups\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-a8fec3e6138a4f0690853c1c160756ab\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-a8fec3e6138a4f0690853c1c160756ab\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-a8fec3e6138a4f0690853c1c160756ab\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.8.1?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function loadScript(lib) {\n",
       "      return new Promise(function(resolve, reject) {\n",
       "        var s = document.createElement('script');\n",
       "        s.src = paths[lib];\n",
       "        s.async = true;\n",
       "        s.onload = () => resolve(paths[lib]);\n",
       "        s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "        document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "      });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else if (typeof vegaEmbed === \"function\") {\n",
       "      displayChart(vegaEmbed);\n",
       "    } else {\n",
       "      loadScript(\"vega\")\n",
       "        .then(() => loadScript(\"vega-lite\"))\n",
       "        .then(() => loadScript(\"vega-embed\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}, \"axis\": {\"grid\": true}}, \"vconcat\": [{\"vconcat\": [{\"mark\": \"circle\", \"encoding\": {\"size\": {\"type\": \"quantitative\", \"field\": \"avg_rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"Overall rank of pathway\", \"titleFontSize\": 12}, \"field\": \"avg_rank_label\", \"sort\": [\"304.4\", \"304.4\", \"304.4\", \"304.4\", \"304.4\", \"320.4\", \"320.4\", \"320.4\", \"320.4\", \"320.4\", \"340.7\", \"340.7\", \"340.7\", \"340.7\", \"340.7\", \"342.9\", \"342.9\", \"342.9\", \"342.9\", \"342.9\", \"351.4\", \"351.4\", \"351.4\", \"351.4\", \"351.4\", \"355.0\", \"355.0\", \"355.0\", \"355.0\", \"355.0\", \"392.9\", \"392.9\", \"392.9\", \"392.9\", \"392.9\", \"402.7\", \"402.7\", \"402.7\", \"402.7\", \"402.7\", \"460.5\", \" 460.5\", \"460.5\", \" 460.5\", \"460.5\", \" 460.5\", \"460.5\", \" 460.5\", \"460.5\", \" 460.5\"]}}, \"width\": 400}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Pathway tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"GRB2:SOS provides linkage to MAPK signaling for Integrins (R-HSA-354194)\", \"GRB2:SOS provides linkage to MAPK signaling for Integrins (R-HSA-354194)\", \"GRB2:SOS provides linkage to MAPK signaling for Integrins (R-HSA-354194)\", \"GRB2:SOS provides linkage to MAPK signaling for Integrins (R-HSA-354194)\", \"GRB2:SOS provides linkage to MAPK signaling for Integrins (R-HSA-354194)\", \"p130Cas linkage to MAPK signaling for integrins (R-HSA-372708)\", \"p130Cas linkage to MAPK signaling for integrins (R-HSA-372708)\", \"p130Cas linkage to MAPK signaling for integrins (R-HSA-372708)\", \"p130Cas linkage to MAPK signaling for integrins (R-HSA-372708)\", \"p130Cas linkage to MAPK signaling for integrins (R-HSA-372708)\", \"Negative regulation of NOTCH4 signaling (R-HSA-9604323)\", \"Negative regulation of NOTCH4 signaling (R-HSA-9604323)\", \"Negative regulation of NOTCH4 signaling (R-HSA-9604323)\", \"Negative regulation of NOTCH4 signaling (R-HSA-9604323)\", \"Negative regulation of NOTCH4 signaling (R-HSA-9604323)\", \"Oxygen-dependent proline hydroxylation of Hypoxia-inducible Factor Alpha (R-HSA-1234176)\", \"Oxygen-dependent proline hydroxylation of Hypoxia-inducible Factor Alpha (R-HSA-1234176)\", \"Oxygen-dependent proline hydroxylation of Hypoxia-inducible Factor Alpha (R-HSA-1234176)\", \"Oxygen-dependent proline hydroxylation of Hypoxia-inducible Factor Alpha (R-HSA-1234176)\", \"Oxygen-dependent proline hydroxylation of Hypoxia-inducible Factor Alpha (R-HSA-1234176)\", \"Metabolism of polyamines (R-HSA-351202)\", \"Metabolism of polyamines (R-HSA-351202)\", \"Metabolism of polyamines (R-HSA-351202)\", \"Metabolism of polyamines (R-HSA-351202)\", \"Metabolism of polyamines (R-HSA-351202)\", \"RHO GTPases activate KTN1 (R-HSA-5625970)\", \"RHO GTPases activate KTN1 (R-HSA-5625970)\", \"RHO GTPases activate KTN1 (R-HSA-5625970)\", \"RHO GTPases activate KTN1 (R-HSA-5625970)\", \"RHO GTPases activate KTN1 (R-HSA-5625970)\", \"Signaling by FGFR2 in disease (R-HSA-5655253)\", \"Signaling by FGFR2 in disease (R-HSA-5655253)\", \"Signaling by FGFR2 in disease (R-HSA-5655253)\", \"Signaling by FGFR2 in disease (R-HSA-5655253)\", \"Signaling by FGFR2 in disease (R-HSA-5655253)\", \"Potential therapeutics for SARS (R-HSA-9679191)\", \"Potential therapeutics for SARS (R-HSA-9679191)\", \"Potential therapeutics for SARS (R-HSA-9679191)\", \"Potential therapeutics for SARS (R-HSA-9679191)\", \"Potential therapeutics for SARS (R-HSA-9679191)\", \"BMAL1:CLOCK,NPAS2 activates circadian gene expression (R-HSA-1368108)\", \"Mitochondrial biogenesis (R-HSA-1592230)\", \"BMAL1:CLOCK,NPAS2 activates circadian gene expression (R-HSA-1368108)\", \"Mitochondrial biogenesis (R-HSA-1592230)\", \"BMAL1:CLOCK,NPAS2 activates circadian gene expression (R-HSA-1368108)\", \"Mitochondrial biogenesis (R-HSA-1592230)\", \"BMAL1:CLOCK,NPAS2 activates circadian gene expression (R-HSA-1368108)\", \"Mitochondrial biogenesis (R-HSA-1592230)\", \"BMAL1:CLOCK,NPAS2 activates circadian gene expression (R-HSA-1368108)\", \"Mitochondrial biogenesis (R-HSA-1592230)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": \"Cancer type\", \"titleFontSize\": 12}, \"field\": \"cancer_type\"}}, \"height\": 300, \"width\": 400}], \"data\": {\"url\": \"altair_data/altair-data-bb0b8f7a61753fb35239a0b7ed003a85.json\", \"format\": {\"type\": \"json\"}}, \"title\": \"Reactome with_gain data - rSEA, threshold = 0.99, sort by Comp_adjP, cutoff Comp_adjP\"}, {\"vconcat\": [{\"mark\": \"circle\", \"encoding\": {\"size\": {\"type\": \"quantitative\", \"field\": \"avg_rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"Overall rank of pathway\", \"titleFontSize\": 12}, \"field\": \"avg_rank_label\", \"sort\": [\"301.3\", \"301.3\", \"301.3\", \"301.3\", \"301.3\", \"392.8\", \"392.8\", \"392.8\", \"392.8\", \"392.8\", \"395.3\", \" 395.3\", \"  395.3\", \"395.3\", \" 395.3\", \"  395.3\", \"395.3\", \" 395.3\", \"  395.3\", \"395.3\", \" 395.3\", \"  395.3\", \"395.3\", \" 395.3\", \"  395.3\", \"407.8\", \"407.8\", \"407.8\", \"407.8\", \"407.8\", \"431.3\", \"431.3\", \"431.3\", \"431.3\", \"431.3\", \"441.9\", \"441.9\", \"441.9\", \"441.9\", \"441.9\", \"517.2\", \"517.2\", \"517.2\", \"517.2\", \"517.2\", \"518.6\", \"518.6\", \"518.6\", \"518.6\", \"518.6\"]}}, \"width\": 400}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Pathway tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Oncogene Induced Senescence (R-HSA-2559585)\", \"Oncogene Induced Senescence (R-HSA-2559585)\", \"Oncogene Induced Senescence (R-HSA-2559585)\", \"Oncogene Induced Senescence (R-HSA-2559585)\", \"Oncogene Induced Senescence (R-HSA-2559585)\", \"Signaling by ERBB2 TMD/JMD mutants (R-HSA-9665686)\", \"Signaling by ERBB2 TMD/JMD mutants (R-HSA-9665686)\", \"Signaling by ERBB2 TMD/JMD mutants (R-HSA-9665686)\", \"Signaling by ERBB2 TMD/JMD mutants (R-HSA-9665686)\", \"Signaling by ERBB2 TMD/JMD mutants (R-HSA-9665686)\", \"Signaling by ERBB2 ECD mutants (R-HSA-9665348)\", \"Signaling by Erythropoietin (R-HSA-9006335)\", \"Signaling by NTRK3 (TRKC) (R-HSA-9034015)\", \"Signaling by ERBB2 ECD mutants (R-HSA-9665348)\", \"Signaling by Erythropoietin (R-HSA-9006335)\", \"Signaling by NTRK3 (TRKC) (R-HSA-9034015)\", \"Signaling by ERBB2 ECD mutants (R-HSA-9665348)\", \"Signaling by Erythropoietin (R-HSA-9006335)\", \"Signaling by NTRK3 (TRKC) (R-HSA-9034015)\", \"Signaling by ERBB2 ECD mutants (R-HSA-9665348)\", \"Signaling by Erythropoietin (R-HSA-9006335)\", \"Signaling by NTRK3 (TRKC) (R-HSA-9034015)\", \"Signaling by ERBB2 ECD mutants (R-HSA-9665348)\", \"Signaling by Erythropoietin (R-HSA-9006335)\", \"Signaling by NTRK3 (TRKC) (R-HSA-9034015)\", \"SUMOylation of chromatin organization proteins (R-HSA-4551638)\", \"SUMOylation of chromatin organization proteins (R-HSA-4551638)\", \"SUMOylation of chromatin organization proteins (R-HSA-4551638)\", \"SUMOylation of chromatin organization proteins (R-HSA-4551638)\", \"SUMOylation of chromatin organization proteins (R-HSA-4551638)\", \"Ephrin signaling (R-HSA-3928664)\", \"Ephrin signaling (R-HSA-3928664)\", \"Ephrin signaling (R-HSA-3928664)\", \"Ephrin signaling (R-HSA-3928664)\", \"Ephrin signaling (R-HSA-3928664)\", \"Deactivation of the beta-catenin transactivating complex (R-HSA-3769402)\", \"Deactivation of the beta-catenin transactivating complex (R-HSA-3769402)\", \"Deactivation of the beta-catenin transactivating complex (R-HSA-3769402)\", \"Deactivation of the beta-catenin transactivating complex (R-HSA-3769402)\", \"Deactivation of the beta-catenin transactivating complex (R-HSA-3769402)\", \"Cytosolic iron-sulfur cluster assembly (R-HSA-2564830)\", \"Cytosolic iron-sulfur cluster assembly (R-HSA-2564830)\", \"Cytosolic iron-sulfur cluster assembly (R-HSA-2564830)\", \"Cytosolic iron-sulfur cluster assembly (R-HSA-2564830)\", \"Cytosolic iron-sulfur cluster assembly (R-HSA-2564830)\", \"Signaling by MRAS-complex mutants (R-HSA-9660537)\", \"Signaling by MRAS-complex mutants (R-HSA-9660537)\", \"Signaling by MRAS-complex mutants (R-HSA-9660537)\", \"Signaling by MRAS-complex mutants (R-HSA-9660537)\", \"Signaling by MRAS-complex mutants (R-HSA-9660537)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": \"Cancer type\", \"titleFontSize\": 12}, \"field\": \"cancer_type\"}}, \"height\": 300, \"width\": 400}], \"data\": {\"url\": \"altair_data/altair-data-24e95848d05ac270500a36042ef690c8.json\", \"format\": {\"type\": \"json\"}}, \"title\": \"Reactome without_gain data - rSEA, threshold = 0.99, sort by Comp_adjP, cutoff Comp_adjP\"}, {\"vconcat\": [{\"mark\": \"circle\", \"encoding\": {\"size\": {\"type\": \"quantitative\", \"field\": \"avg_rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"Overall rank of pathway\", \"titleFontSize\": 12}, \"field\": \"avg_rank_label\", \"sort\": [\"362.2\", \"362.2\", \"362.2\", \"362.2\", \"362.2\", \"379.8\", \"379.8\", \"379.8\", \"379.8\", \"379.8\", \"381.8\", \" 381.8\", \"381.8\", \" 381.8\", \"381.8\", \" 381.8\", \"381.8\", \" 381.8\", \"381.8\", \" 381.8\", \"385.6\", \"385.6\", \"385.6\", \"385.6\", \"385.6\", \"389.3\", \"389.3\", \"389.3\", \"389.3\", \"389.3\", \"389.5\", \"389.5\", \"389.5\", \"389.5\", \"389.5\", \"392.0\", \"392.0\", \"392.0\", \"392.0\", \"392.0\", \"396.7\", \" 396.7\", \"396.7\", \" 396.7\", \"396.7\", \" 396.7\", \"396.7\", \" 396.7\", \"396.7\", \" 396.7\"]}}, \"width\": 400}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Pathway tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Response of Mtb to phagocytosis (R-HSA-9637690)\", \"Response of Mtb to phagocytosis (R-HSA-9637690)\", \"Response of Mtb to phagocytosis (R-HSA-9637690)\", \"Response of Mtb to phagocytosis (R-HSA-9637690)\", \"Response of Mtb to phagocytosis (R-HSA-9637690)\", \"Transcriptional regulation by the AP-2 (TFAP2) family of transcription factors (R-HSA-8864260)\", \"Transcriptional regulation by the AP-2 (TFAP2) family of transcription factors (R-HSA-8864260)\", \"Transcriptional regulation by the AP-2 (TFAP2) family of transcription factors (R-HSA-8864260)\", \"Transcriptional regulation by the AP-2 (TFAP2) family of transcription factors (R-HSA-8864260)\", \"Transcriptional regulation by the AP-2 (TFAP2) family of transcription factors (R-HSA-8864260)\", \"Non-integrin membrane-ECM interactions (R-HSA-3000171)\", \"Potential therapeutics for SARS (R-HSA-9679191)\", \"Non-integrin membrane-ECM interactions (R-HSA-3000171)\", \"Potential therapeutics for SARS (R-HSA-9679191)\", \"Non-integrin membrane-ECM interactions (R-HSA-3000171)\", \"Potential therapeutics for SARS (R-HSA-9679191)\", \"Non-integrin membrane-ECM interactions (R-HSA-3000171)\", \"Potential therapeutics for SARS (R-HSA-9679191)\", \"Non-integrin membrane-ECM interactions (R-HSA-3000171)\", \"Potential therapeutics for SARS (R-HSA-9679191)\", \"Translation of structural proteins (R-HSA-9683701)\", \"Translation of structural proteins (R-HSA-9683701)\", \"Translation of structural proteins (R-HSA-9683701)\", \"Translation of structural proteins (R-HSA-9683701)\", \"Translation of structural proteins (R-HSA-9683701)\", \"Molecules associated with elastic fibres (R-HSA-2129379)\", \"Molecules associated with elastic fibres (R-HSA-2129379)\", \"Molecules associated with elastic fibres (R-HSA-2129379)\", \"Molecules associated with elastic fibres (R-HSA-2129379)\", \"Molecules associated with elastic fibres (R-HSA-2129379)\", \"Infection with Mycobacterium tuberculosis (R-HSA-9635486)\", \"Infection with Mycobacterium tuberculosis (R-HSA-9635486)\", \"Infection with Mycobacterium tuberculosis (R-HSA-9635486)\", \"Infection with Mycobacterium tuberculosis (R-HSA-9635486)\", \"Infection with Mycobacterium tuberculosis (R-HSA-9635486)\", \"Iron uptake and transport (R-HSA-917937)\", \"Iron uptake and transport (R-HSA-917937)\", \"Iron uptake and transport (R-HSA-917937)\", \"Iron uptake and transport (R-HSA-917937)\", \"Iron uptake and transport (R-HSA-917937)\", \"DNA Double-Strand Break Repair (R-HSA-5693532)\", \"HDR through Homologous Recombination (HRR) or Single Strand Annealing (SSA) (R-HSA-5693567)\", \"DNA Double-Strand Break Repair (R-HSA-5693532)\", \"HDR through Homologous Recombination (HRR) or Single Strand Annealing (SSA) (R-HSA-5693567)\", \"DNA Double-Strand Break Repair (R-HSA-5693532)\", \"HDR through Homologous Recombination (HRR) or Single Strand Annealing (SSA) (R-HSA-5693567)\", \"DNA Double-Strand Break Repair (R-HSA-5693532)\", \"HDR through Homologous Recombination (HRR) or Single Strand Annealing (SSA) (R-HSA-5693567)\", \"DNA Double-Strand Break Repair (R-HSA-5693532)\", \"HDR through Homologous Recombination (HRR) or Single Strand Annealing (SSA) (R-HSA-5693567)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": \"Cancer type\", \"titleFontSize\": 12}, \"field\": \"cancer_type\"}}, \"height\": 300, \"width\": 400}], \"data\": {\"url\": \"altair_data/altair-data-8ee625e750bc4756a56a8eb2ce7ba845.json\", \"format\": {\"type\": \"json\"}}, \"title\": \"Reactome with_loss data - rSEA, threshold = 0.99, sort by Comp_adjP, cutoff Comp_adjP\"}, {\"vconcat\": [{\"mark\": \"circle\", \"encoding\": {\"size\": {\"type\": \"quantitative\", \"field\": \"avg_rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"Overall rank of pathway\", \"titleFontSize\": 12}, \"field\": \"avg_rank_label\", \"sort\": [\"385.8\", \"385.8\", \"385.8\", \"385.8\", \"385.8\", \"400.3\", \" 400.3\", \"400.3\", \" 400.3\", \"400.3\", \" 400.3\", \"400.3\", \" 400.3\", \"400.3\", \" 400.3\", \"464.9\", \"464.9\", \"464.9\", \"464.9\", \"464.9\", \"492.6\", \" 492.6\", \"492.6\", \" 492.6\", \"492.6\", \" 492.6\", \"492.6\", \" 492.6\", \"492.6\", \" 492.6\", \"492.8\", \"492.8\", \"492.8\", \"492.8\", \"492.8\", \"494.0\", \"494.0\", \"494.0\", \"494.0\", \"494.0\", \"494.7\", \"494.7\", \"494.7\", \"494.7\", \"494.7\", \"496.3\", \"496.3\", \"496.3\", \"496.3\", \"496.3\"]}}, \"width\": 400}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Pathway tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"TRAF6 mediated NF-kB activation (R-HSA-933542)\", \"TRAF6 mediated NF-kB activation (R-HSA-933542)\", \"TRAF6 mediated NF-kB activation (R-HSA-933542)\", \"TRAF6 mediated NF-kB activation (R-HSA-933542)\", \"TRAF6 mediated NF-kB activation (R-HSA-933542)\", \"Deadenylation of mRNA (R-HSA-429947)\", \"TP53 regulates transcription of additional cell cycle genes whose exact role in the p53 pathway remain uncertain (R-HSA-6804115)\", \"Deadenylation of mRNA (R-HSA-429947)\", \"TP53 regulates transcription of additional cell cycle genes whose exact role in the p53 pathway remain uncertain (R-HSA-6804115)\", \"Deadenylation of mRNA (R-HSA-429947)\", \"TP53 regulates transcription of additional cell cycle genes whose exact role in the p53 pathway remain uncertain (R-HSA-6804115)\", \"Deadenylation of mRNA (R-HSA-429947)\", \"TP53 regulates transcription of additional cell cycle genes whose exact role in the p53 pathway remain uncertain (R-HSA-6804115)\", \"Deadenylation of mRNA (R-HSA-429947)\", \"TP53 regulates transcription of additional cell cycle genes whose exact role in the p53 pathway remain uncertain (R-HSA-6804115)\", \"Recycling pathway of L1 (R-HSA-437239)\", \"Recycling pathway of L1 (R-HSA-437239)\", \"Recycling pathway of L1 (R-HSA-437239)\", \"Recycling pathway of L1 (R-HSA-437239)\", \"Recycling pathway of L1 (R-HSA-437239)\", \"Diseases associated with the TLR signaling cascade (R-HSA-5602358)\", \"Diseases of Immune System (R-HSA-5260271)\", \"Diseases associated with the TLR signaling cascade (R-HSA-5602358)\", \"Diseases of Immune System (R-HSA-5260271)\", \"Diseases associated with the TLR signaling cascade (R-HSA-5602358)\", \"Diseases of Immune System (R-HSA-5260271)\", \"Diseases associated with the TLR signaling cascade (R-HSA-5602358)\", \"Diseases of Immune System (R-HSA-5260271)\", \"Diseases associated with the TLR signaling cascade (R-HSA-5602358)\", \"Diseases of Immune System (R-HSA-5260271)\", \"Downregulation of ERBB2 signaling (R-HSA-8863795)\", \"Downregulation of ERBB2 signaling (R-HSA-8863795)\", \"Downregulation of ERBB2 signaling (R-HSA-8863795)\", \"Downregulation of ERBB2 signaling (R-HSA-8863795)\", \"Downregulation of ERBB2 signaling (R-HSA-8863795)\", \"Ca2+ pathway (R-HSA-4086398)\", \"Ca2+ pathway (R-HSA-4086398)\", \"Ca2+ pathway (R-HSA-4086398)\", \"Ca2+ pathway (R-HSA-4086398)\", \"Ca2+ pathway (R-HSA-4086398)\", \"RUNX2 regulates bone development (R-HSA-8941326)\", \"RUNX2 regulates bone development (R-HSA-8941326)\", \"RUNX2 regulates bone development (R-HSA-8941326)\", \"RUNX2 regulates bone development (R-HSA-8941326)\", \"RUNX2 regulates bone development (R-HSA-8941326)\", \"Peptide hormone metabolism (R-HSA-2980736)\", \"Peptide hormone metabolism (R-HSA-2980736)\", \"Peptide hormone metabolism (R-HSA-2980736)\", \"Peptide hormone metabolism (R-HSA-2980736)\", \"Peptide hormone metabolism (R-HSA-2980736)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": \"Cancer type\", \"titleFontSize\": 12}, \"field\": \"cancer_type\"}}, \"height\": 300, \"width\": 400}], \"data\": {\"url\": \"altair_data/altair-data-24ec514e573681f819d0ffa730df4fc9.json\", \"format\": {\"type\": \"json\"}}, \"title\": \"Reactome without_loss data - rSEA, threshold = 0.99, sort by Comp_adjP, cutoff Comp_adjP\"}, {\"vconcat\": [{\"mark\": \"circle\", \"encoding\": {\"size\": {\"type\": \"quantitative\", \"field\": \"avg_rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"Overall rank of pathway\", \"titleFontSize\": 12}, \"field\": \"avg_rank_label\", \"sort\": [\"271.7\", \"271.7\", \"271.7\", \"271.7\", \"271.7\", \"305.7\", \"305.7\", \"305.7\", \"305.7\", \"305.7\", \"352.1\", \"352.1\", \"352.1\", \"352.1\", \"352.1\", \"361.5\", \"361.5\", \"361.5\", \"361.5\", \"361.5\", \"363.4\", \"363.4\", \"363.4\", \"363.4\", \"363.4\", \"371.7\", \" 371.7\", \"  371.7\", \"371.7\", \" 371.7\", \"  371.7\", \"371.7\", \" 371.7\", \"  371.7\", \"371.7\", \" 371.7\", \"  371.7\", \"371.7\", \" 371.7\", \"  371.7\", \"374.1\", \"374.1\", \"374.1\", \"374.1\", \"374.1\", \"376.2\", \"376.2\", \"376.2\", \"376.2\", \"376.2\"]}}, \"width\": 400}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Pathway tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"G alpha (q) signalling events (R-HSA-416476)\", \"G alpha (q) signalling events (R-HSA-416476)\", \"G alpha (q) signalling events (R-HSA-416476)\", \"G alpha (q) signalling events (R-HSA-416476)\", \"G alpha (q) signalling events (R-HSA-416476)\", \"Interconversion of nucleotide di- and triphosphates (R-HSA-499943)\", \"Interconversion of nucleotide di- and triphosphates (R-HSA-499943)\", \"Interconversion of nucleotide di- and triphosphates (R-HSA-499943)\", \"Interconversion of nucleotide di- and triphosphates (R-HSA-499943)\", \"Interconversion of nucleotide di- and triphosphates (R-HSA-499943)\", \"Extra-nuclear estrogen signaling (R-HSA-9009391)\", \"Extra-nuclear estrogen signaling (R-HSA-9009391)\", \"Extra-nuclear estrogen signaling (R-HSA-9009391)\", \"Extra-nuclear estrogen signaling (R-HSA-9009391)\", \"Extra-nuclear estrogen signaling (R-HSA-9009391)\", \"Signaling by Retinoic Acid (R-HSA-5362517)\", \"Signaling by Retinoic Acid (R-HSA-5362517)\", \"Signaling by Retinoic Acid (R-HSA-5362517)\", \"Signaling by Retinoic Acid (R-HSA-5362517)\", \"Signaling by Retinoic Acid (R-HSA-5362517)\", \"DNA Damage/Telomere Stress Induced Senescence (R-HSA-2559586)\", \"DNA Damage/Telomere Stress Induced Senescence (R-HSA-2559586)\", \"DNA Damage/Telomere Stress Induced Senescence (R-HSA-2559586)\", \"DNA Damage/Telomere Stress Induced Senescence (R-HSA-2559586)\", \"DNA Damage/Telomere Stress Induced Senescence (R-HSA-2559586)\", \"Mitotic G1 phase and G1/S transition (R-HSA-453279)\", \"Nonsense Mediated Decay (NMD) enhanced by the Exon Junction Complex (EJC) (R-HSA-975957)\", \"Nonsense-Mediated Decay (NMD) (R-HSA-927802)\", \"Mitotic G1 phase and G1/S transition (R-HSA-453279)\", \"Nonsense Mediated Decay (NMD) enhanced by the Exon Junction Complex (EJC) (R-HSA-975957)\", \"Nonsense-Mediated Decay (NMD) (R-HSA-927802)\", \"Mitotic G1 phase and G1/S transition (R-HSA-453279)\", \"Nonsense Mediated Decay (NMD) enhanced by the Exon Junction Complex (EJC) (R-HSA-975957)\", \"Nonsense-Mediated Decay (NMD) (R-HSA-927802)\", \"Mitotic G1 phase and G1/S transition (R-HSA-453279)\", \"Nonsense Mediated Decay (NMD) enhanced by the Exon Junction Complex (EJC) (R-HSA-975957)\", \"Nonsense-Mediated Decay (NMD) (R-HSA-927802)\", \"Mitotic G1 phase and G1/S transition (R-HSA-453279)\", \"Nonsense Mediated Decay (NMD) enhanced by the Exon Junction Complex (EJC) (R-HSA-975957)\", \"Nonsense-Mediated Decay (NMD) (R-HSA-927802)\", \"Transcriptional activation of mitochondrial biogenesis (R-HSA-2151201)\", \"Transcriptional activation of mitochondrial biogenesis (R-HSA-2151201)\", \"Transcriptional activation of mitochondrial biogenesis (R-HSA-2151201)\", \"Transcriptional activation of mitochondrial biogenesis (R-HSA-2151201)\", \"Transcriptional activation of mitochondrial biogenesis (R-HSA-2151201)\", \"rRNA processing in the nucleus and cytosol (R-HSA-8868773)\", \"rRNA processing in the nucleus and cytosol (R-HSA-8868773)\", \"rRNA processing in the nucleus and cytosol (R-HSA-8868773)\", \"rRNA processing in the nucleus and cytosol (R-HSA-8868773)\", \"rRNA processing in the nucleus and cytosol (R-HSA-8868773)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": \"Cancer type\", \"titleFontSize\": 12}, \"field\": \"cancer_type\"}}, \"height\": 300, \"width\": 400}], \"data\": {\"url\": \"altair_data/altair-data-4abbaecb88bbcc8faa35603ab8e9d3d7.json\", \"format\": {\"type\": \"json\"}}, \"title\": \"Reactome with_both data - rSEA, threshold = 0.99, sort by Comp_adjP, cutoff Comp_adjP\"}, {\"vconcat\": [{\"mark\": \"circle\", \"encoding\": {\"size\": {\"type\": \"quantitative\", \"field\": \"avg_rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"Overall rank of pathway\", \"titleFontSize\": 12}, \"field\": \"avg_rank_label\", \"sort\": [\"323.7\", \"323.7\", \"323.7\", \"323.7\", \"323.7\", \"359.3\", \"359.3\", \"359.3\", \"359.3\", \"359.3\", \"468.2\", \"468.2\", \"468.2\", \"468.2\", \"468.2\", \"473.9\", \"473.9\", \"473.9\", \"473.9\", \"473.9\", \"483.9\", \"483.9\", \"483.9\", \"483.9\", \"483.9\", \"484.2\", \"484.2\", \"484.2\", \"484.2\", \"484.2\", \"488.9\", \"488.9\", \"488.9\", \"488.9\", \"488.9\", \"489.1\", \"489.1\", \"489.1\", \"489.1\", \"489.1\", \"489.7\", \"489.7\", \"489.7\", \"489.7\", \"489.7\", \"492.0\", \"492.0\", \"492.0\", \"492.0\", \"492.0\"]}}, \"width\": 400}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Pathway tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Dopamine Neurotransmitter Release Cycle (R-HSA-212676)\", \"Dopamine Neurotransmitter Release Cycle (R-HSA-212676)\", \"Dopamine Neurotransmitter Release Cycle (R-HSA-212676)\", \"Dopamine Neurotransmitter Release Cycle (R-HSA-212676)\", \"Dopamine Neurotransmitter Release Cycle (R-HSA-212676)\", \"E2F-enabled inhibition of pre-replication complex formation (R-HSA-113507)\", \"E2F-enabled inhibition of pre-replication complex formation (R-HSA-113507)\", \"E2F-enabled inhibition of pre-replication complex formation (R-HSA-113507)\", \"E2F-enabled inhibition of pre-replication complex formation (R-HSA-113507)\", \"E2F-enabled inhibition of pre-replication complex formation (R-HSA-113507)\", \"TP53 regulates transcription of additional cell cycle genes whose exact role in the p53 pathway remain uncertain (R-HSA-6804115)\", \"TP53 regulates transcription of additional cell cycle genes whose exact role in the p53 pathway remain uncertain (R-HSA-6804115)\", \"TP53 regulates transcription of additional cell cycle genes whose exact role in the p53 pathway remain uncertain (R-HSA-6804115)\", \"TP53 regulates transcription of additional cell cycle genes whose exact role in the p53 pathway remain uncertain (R-HSA-6804115)\", \"TP53 regulates transcription of additional cell cycle genes whose exact role in the p53 pathway remain uncertain (R-HSA-6804115)\", \"DARPP-32 events (R-HSA-180024)\", \"DARPP-32 events (R-HSA-180024)\", \"DARPP-32 events (R-HSA-180024)\", \"DARPP-32 events (R-HSA-180024)\", \"DARPP-32 events (R-HSA-180024)\", \"Interleukin-27 signaling (R-HSA-9020956)\", \"Interleukin-27 signaling (R-HSA-9020956)\", \"Interleukin-27 signaling (R-HSA-9020956)\", \"Interleukin-27 signaling (R-HSA-9020956)\", \"Interleukin-27 signaling (R-HSA-9020956)\", \"Defensins (R-HSA-1461973)\", \"Defensins (R-HSA-1461973)\", \"Defensins (R-HSA-1461973)\", \"Defensins (R-HSA-1461973)\", \"Defensins (R-HSA-1461973)\", \"PI3K/AKT activation (R-HSA-198203)\", \"PI3K/AKT activation (R-HSA-198203)\", \"PI3K/AKT activation (R-HSA-198203)\", \"PI3K/AKT activation (R-HSA-198203)\", \"PI3K/AKT activation (R-HSA-198203)\", \"DAG and IP3 signaling (R-HSA-1489509)\", \"DAG and IP3 signaling (R-HSA-1489509)\", \"DAG and IP3 signaling (R-HSA-1489509)\", \"DAG and IP3 signaling (R-HSA-1489509)\", \"DAG and IP3 signaling (R-HSA-1489509)\", \"Neurotransmitter release cycle (R-HSA-112310)\", \"Neurotransmitter release cycle (R-HSA-112310)\", \"Neurotransmitter release cycle (R-HSA-112310)\", \"Neurotransmitter release cycle (R-HSA-112310)\", \"Neurotransmitter release cycle (R-HSA-112310)\", \"Interleukin-35 Signalling (R-HSA-8984722)\", \"Interleukin-35 Signalling (R-HSA-8984722)\", \"Interleukin-35 Signalling (R-HSA-8984722)\", \"Interleukin-35 Signalling (R-HSA-8984722)\", \"Interleukin-35 Signalling (R-HSA-8984722)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": \"Cancer type\", \"titleFontSize\": 12}, \"field\": \"cancer_type\"}}, \"height\": 300, \"width\": 400}], \"data\": {\"url\": \"altair_data/altair-data-5d58df7c964e2b97e792d15a7d6d165e.json\", \"format\": {\"type\": \"json\"}}, \"title\": \"Reactome without_both data - rSEA, threshold = 0.99, sort by Comp_adjP, cutoff Comp_adjP\"}], \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.8.1.json\"}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.VConcatChart(...)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alt.vconcat(*reactome_plots).configure_axis(grid=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_groups = [\n",
    "    'with_gain', \n",
    "    'without_gain', \n",
    "    'with_loss', \n",
    "    'without_loss', \n",
    "    'with_both', \n",
    "    'without_both'\n",
    "]\n",
    "\n",
    "go_plots = [\n",
    "    plot_top_ten(\n",
    "        enrich_file_path=go_results_file, \n",
    "        expr_file_path=diff_expr_file, \n",
    "        gmt_file_path=go_gmt,\n",
    "        enrichment_group=eg,\n",
    "        xtitle=f\"GO {eg} data - rSEA, threshold = 0.99, sort by Comp_adjP, cutoff Comp_adjP\",\n",
    "        pathway_min_size=5,\n",
    "        min_cancers_enriched=5,\n",
    "        sort_col=\"Comp.0.99.adjP\",\n",
    "        sort_asc=True,\n",
    "        cutoff_col=\"Comp.0.99.adjP\",\n",
    "        cutoff=0.05\n",
    "    )[0]\n",
    "    for eg in all_groups\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-1143be2c4e5f4141abd37095ea8c3629\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-1143be2c4e5f4141abd37095ea8c3629\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-1143be2c4e5f4141abd37095ea8c3629\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.8.1?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function loadScript(lib) {\n",
       "      return new Promise(function(resolve, reject) {\n",
       "        var s = document.createElement('script');\n",
       "        s.src = paths[lib];\n",
       "        s.async = true;\n",
       "        s.onload = () => resolve(paths[lib]);\n",
       "        s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "        document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "      });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else if (typeof vegaEmbed === \"function\") {\n",
       "      displayChart(vegaEmbed);\n",
       "    } else {\n",
       "      loadScript(\"vega\")\n",
       "        .then(() => loadScript(\"vega-lite\"))\n",
       "        .then(() => loadScript(\"vega-embed\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}, \"axis\": {\"grid\": true}}, \"vconcat\": [{\"vconcat\": [{\"mark\": \"circle\", \"encoding\": {\"size\": {\"type\": \"quantitative\", \"field\": \"avg_rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"Overall rank of pathway\", \"titleFontSize\": 12}, \"field\": \"avg_rank_label\", \"sort\": [\"703.2\", \"703.2\", \"703.2\", \"703.2\", \"703.2\", \"720.0\", \"720.0\", \"720.0\", \"720.0\", \"720.0\", \"848.4\", \" 848.4\", \"848.4\", \" 848.4\", \"848.4\", \" 848.4\", \"848.4\", \" 848.4\", \"848.4\", \" 848.4\", \"864.6\", \"864.6\", \"864.6\", \"864.6\", \"864.6\", \"873.4\", \"873.4\", \"873.4\", \"873.4\", \"873.4\", \"880.0\", \" 880.0\", \"880.0\", \" 880.0\", \"880.0\", \" 880.0\", \"880.0\", \" 880.0\", \"880.0\", \" 880.0\", \"886.7\", \"886.7\", \"886.7\", \"886.7\", \"886.7\", \"891.4\", \"891.4\", \"891.4\", \"891.4\", \"891.4\"]}}, \"width\": 400}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Pathway tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"positive regulation of interleukin-8 secretion (GO:2000484)\", \"positive regulation of interleukin-8 secretion (GO:2000484)\", \"positive regulation of interleukin-8 secretion (GO:2000484)\", \"positive regulation of interleukin-8 secretion (GO:2000484)\", \"positive regulation of interleukin-8 secretion (GO:2000484)\", \"regulation of receptor internalization (GO:0002090)\", \"regulation of receptor internalization (GO:0002090)\", \"regulation of receptor internalization (GO:0002090)\", \"regulation of receptor internalization (GO:0002090)\", \"regulation of receptor internalization (GO:0002090)\", \"ESCRT complex disassembly (GO:1904896)\", \"ESCRT III complex disassembly (GO:1904903)\", \"ESCRT complex disassembly (GO:1904896)\", \"ESCRT III complex disassembly (GO:1904903)\", \"ESCRT complex disassembly (GO:1904896)\", \"ESCRT III complex disassembly (GO:1904903)\", \"ESCRT complex disassembly (GO:1904896)\", \"ESCRT III complex disassembly (GO:1904903)\", \"ESCRT complex disassembly (GO:1904896)\", \"ESCRT III complex disassembly (GO:1904903)\", \"regulation of B cell receptor signaling pathway (GO:0050855)\", \"regulation of B cell receptor signaling pathway (GO:0050855)\", \"regulation of B cell receptor signaling pathway (GO:0050855)\", \"regulation of B cell receptor signaling pathway (GO:0050855)\", \"regulation of B cell receptor signaling pathway (GO:0050855)\", \"establishment of apical/basal cell polarity (GO:0035089)\", \"establishment of apical/basal cell polarity (GO:0035089)\", \"establishment of apical/basal cell polarity (GO:0035089)\", \"establishment of apical/basal cell polarity (GO:0035089)\", \"establishment of apical/basal cell polarity (GO:0035089)\", \"positive regulation of tumor necrosis factor production (GO:0032760)\", \"positive regulation of tumor necrosis factor superfamily cytokine production (GO:1903557)\", \"positive regulation of tumor necrosis factor production (GO:0032760)\", \"positive regulation of tumor necrosis factor superfamily cytokine production (GO:1903557)\", \"positive regulation of tumor necrosis factor production (GO:0032760)\", \"positive regulation of tumor necrosis factor superfamily cytokine production (GO:1903557)\", \"positive regulation of tumor necrosis factor production (GO:0032760)\", \"positive regulation of tumor necrosis factor superfamily cytokine production (GO:1903557)\", \"positive regulation of tumor necrosis factor production (GO:0032760)\", \"positive regulation of tumor necrosis factor superfamily cytokine production (GO:1903557)\", \"regulation of centrosome cycle (GO:0046605)\", \"regulation of centrosome cycle (GO:0046605)\", \"regulation of centrosome cycle (GO:0046605)\", \"regulation of centrosome cycle (GO:0046605)\", \"regulation of centrosome cycle (GO:0046605)\", \"positive regulation of antigen receptor-mediated signaling pathway (GO:0050857)\", \"positive regulation of antigen receptor-mediated signaling pathway (GO:0050857)\", \"positive regulation of antigen receptor-mediated signaling pathway (GO:0050857)\", \"positive regulation of antigen receptor-mediated signaling pathway (GO:0050857)\", \"positive regulation of antigen receptor-mediated signaling pathway (GO:0050857)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": \"Cancer type\", \"titleFontSize\": 12}, \"field\": \"cancer_type\"}}, \"height\": 300, \"width\": 400}], \"data\": {\"url\": \"altair_data/altair-data-4d17e8a36f5618c74acd5a3fb65514a2.json\", \"format\": {\"type\": \"json\"}}, \"title\": \"GO with_gain data - rSEA, threshold = 0.99, sort by Comp_adjP, cutoff Comp_adjP\"}, {\"vconcat\": [{\"mark\": \"circle\", \"encoding\": {\"size\": {\"type\": \"quantitative\", \"field\": \"avg_rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"Overall rank of pathway\", \"titleFontSize\": 12}, \"field\": \"avg_rank_label\", \"sort\": [\"703.9\", \"703.9\", \"703.9\", \"703.9\", \"703.9\", \"917.2\", \"917.2\", \"917.2\", \"917.2\", \"917.2\", \"921.5\", \"921.5\", \"921.5\", \"921.5\", \"921.5\", \"925.5\", \" 925.5\", \"  925.5\", \"925.5\", \" 925.5\", \"  925.5\", \"925.5\", \" 925.5\", \"  925.5\", \"925.5\", \" 925.5\", \"  925.5\", \"925.5\", \" 925.5\", \"  925.5\", \"934.8\", \"934.8\", \"934.8\", \"934.8\", \"934.8\", \"948.7\", \"948.7\", \"948.7\", \"948.7\", \"948.7\", \"949.6\", \"949.6\", \"949.6\", \"949.6\", \"949.6\", \"949.9\", \"949.9\", \"949.9\", \"949.9\", \"949.9\"]}}, \"width\": 400}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Pathway tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"natural killer cell activation (GO:0030101)\", \"natural killer cell activation (GO:0030101)\", \"natural killer cell activation (GO:0030101)\", \"natural killer cell activation (GO:0030101)\", \"natural killer cell activation (GO:0030101)\", \"positive regulation of mRNA catabolic process (GO:0061014)\", \"positive regulation of mRNA catabolic process (GO:0061014)\", \"positive regulation of mRNA catabolic process (GO:0061014)\", \"positive regulation of mRNA catabolic process (GO:0061014)\", \"positive regulation of mRNA catabolic process (GO:0061014)\", \"regulation of Golgi organization (GO:1903358)\", \"regulation of Golgi organization (GO:1903358)\", \"regulation of Golgi organization (GO:1903358)\", \"regulation of Golgi organization (GO:1903358)\", \"regulation of Golgi organization (GO:1903358)\", \"positive regulation of interleukin-1 beta secretion (GO:0050718)\", \"positive regulation of interleukin-1 secretion (GO:0050716)\", \"regulation of interleukin-1 beta secretion (GO:0050706)\", \"positive regulation of interleukin-1 beta secretion (GO:0050718)\", \"positive regulation of interleukin-1 secretion (GO:0050716)\", \"regulation of interleukin-1 beta secretion (GO:0050706)\", \"positive regulation of interleukin-1 beta secretion (GO:0050718)\", \"positive regulation of interleukin-1 secretion (GO:0050716)\", \"regulation of interleukin-1 beta secretion (GO:0050706)\", \"positive regulation of interleukin-1 beta secretion (GO:0050718)\", \"positive regulation of interleukin-1 secretion (GO:0050716)\", \"regulation of interleukin-1 beta secretion (GO:0050706)\", \"positive regulation of interleukin-1 beta secretion (GO:0050718)\", \"positive regulation of interleukin-1 secretion (GO:0050716)\", \"regulation of interleukin-1 beta secretion (GO:0050706)\", \"production of miRNAs involved in gene silencing by miRNA (GO:0035196)\", \"production of miRNAs involved in gene silencing by miRNA (GO:0035196)\", \"production of miRNAs involved in gene silencing by miRNA (GO:0035196)\", \"production of miRNAs involved in gene silencing by miRNA (GO:0035196)\", \"production of miRNAs involved in gene silencing by miRNA (GO:0035196)\", \"negative regulation of catabolic process (GO:0009895)\", \"negative regulation of catabolic process (GO:0009895)\", \"negative regulation of catabolic process (GO:0009895)\", \"negative regulation of catabolic process (GO:0009895)\", \"negative regulation of catabolic process (GO:0009895)\", \"Wnt signaling pathway, calcium modulating pathway (GO:0007223)\", \"Wnt signaling pathway, calcium modulating pathway (GO:0007223)\", \"Wnt signaling pathway, calcium modulating pathway (GO:0007223)\", \"Wnt signaling pathway, calcium modulating pathway (GO:0007223)\", \"Wnt signaling pathway, calcium modulating pathway (GO:0007223)\", \"positive regulation of embryonic development (GO:0040019)\", \"positive regulation of embryonic development (GO:0040019)\", \"positive regulation of embryonic development (GO:0040019)\", \"positive regulation of embryonic development (GO:0040019)\", \"positive regulation of embryonic development (GO:0040019)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": \"Cancer type\", \"titleFontSize\": 12}, \"field\": \"cancer_type\"}}, \"height\": 300, \"width\": 400}], \"data\": {\"url\": \"altair_data/altair-data-62e145609e9dc9025a2c5762afe7939e.json\", \"format\": {\"type\": \"json\"}}, \"title\": \"GO without_gain data - rSEA, threshold = 0.99, sort by Comp_adjP, cutoff Comp_adjP\"}, {\"vconcat\": [{\"mark\": \"circle\", \"encoding\": {\"size\": {\"type\": \"quantitative\", \"field\": \"avg_rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"Overall rank of pathway\", \"titleFontSize\": 12}, \"field\": \"avg_rank_label\", \"sort\": [\"689.0\", \"689.0\", \"689.0\", \"689.0\", \"689.0\", \"862.0\", \"862.0\", \"862.0\", \"862.0\", \"862.0\", \"898.5\", \" 898.5\", \"898.5\", \" 898.5\", \"898.5\", \" 898.5\", \"898.5\", \" 898.5\", \"898.5\", \" 898.5\", \"911.5\", \"911.5\", \"911.5\", \"911.5\", \"911.5\", \"912.2\", \" 912.2\", \"912.2\", \" 912.2\", \"912.2\", \" 912.2\", \"912.2\", \" 912.2\", \"912.2\", \" 912.2\", \"924.6\", \"924.6\", \"924.6\", \"924.6\", \"924.6\", \"933.6\", \" 933.6\", \"933.6\", \" 933.6\", \"933.6\", \" 933.6\", \"933.6\", \" 933.6\", \"933.6\", \" 933.6\"]}}, \"width\": 400}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Pathway tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"regulation of phosphatidylinositol 3-kinase signaling (GO:0014066)\", \"regulation of phosphatidylinositol 3-kinase signaling (GO:0014066)\", \"regulation of phosphatidylinositol 3-kinase signaling (GO:0014066)\", \"regulation of phosphatidylinositol 3-kinase signaling (GO:0014066)\", \"regulation of phosphatidylinositol 3-kinase signaling (GO:0014066)\", \"positive regulation of substrate adhesion-dependent cell spreading (GO:1900026)\", \"positive regulation of substrate adhesion-dependent cell spreading (GO:1900026)\", \"positive regulation of substrate adhesion-dependent cell spreading (GO:1900026)\", \"positive regulation of substrate adhesion-dependent cell spreading (GO:1900026)\", \"positive regulation of substrate adhesion-dependent cell spreading (GO:1900026)\", \"regulation of exosomal secretion (GO:1903541)\", \"positive regulation of exocytosis (GO:0045921)\", \"regulation of exosomal secretion (GO:1903541)\", \"positive regulation of exocytosis (GO:0045921)\", \"regulation of exosomal secretion (GO:1903541)\", \"positive regulation of exocytosis (GO:0045921)\", \"regulation of exosomal secretion (GO:1903541)\", \"positive regulation of exocytosis (GO:0045921)\", \"regulation of exosomal secretion (GO:1903541)\", \"positive regulation of exocytosis (GO:0045921)\", \"regulation of bone mineralization (GO:0030500)\", \"regulation of bone mineralization (GO:0030500)\", \"regulation of bone mineralization (GO:0030500)\", \"regulation of bone mineralization (GO:0030500)\", \"regulation of bone mineralization (GO:0030500)\", \"negative regulation of sequestering of calcium ion (GO:0051283)\", \"release of sequestered calcium ion into cytosol (GO:0051209)\", \"negative regulation of sequestering of calcium ion (GO:0051283)\", \"release of sequestered calcium ion into cytosol (GO:0051209)\", \"negative regulation of sequestering of calcium ion (GO:0051283)\", \"release of sequestered calcium ion into cytosol (GO:0051209)\", \"negative regulation of sequestering of calcium ion (GO:0051283)\", \"release of sequestered calcium ion into cytosol (GO:0051209)\", \"negative regulation of sequestering of calcium ion (GO:0051283)\", \"release of sequestered calcium ion into cytosol (GO:0051209)\", \"positive regulation of DNA metabolic process (GO:0051054)\", \"positive regulation of DNA metabolic process (GO:0051054)\", \"positive regulation of DNA metabolic process (GO:0051054)\", \"positive regulation of DNA metabolic process (GO:0051054)\", \"positive regulation of DNA metabolic process (GO:0051054)\", \"regulation of plasma membrane bounded cell projection assembly (GO:0120032)\", \"regulation of actin filament bundle assembly (GO:0032231)\", \"regulation of plasma membrane bounded cell projection assembly (GO:0120032)\", \"regulation of actin filament bundle assembly (GO:0032231)\", \"regulation of plasma membrane bounded cell projection assembly (GO:0120032)\", \"regulation of actin filament bundle assembly (GO:0032231)\", \"regulation of plasma membrane bounded cell projection assembly (GO:0120032)\", \"regulation of actin filament bundle assembly (GO:0032231)\", \"regulation of plasma membrane bounded cell projection assembly (GO:0120032)\", \"regulation of actin filament bundle assembly (GO:0032231)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": \"Cancer type\", \"titleFontSize\": 12}, \"field\": \"cancer_type\"}}, \"height\": 300, \"width\": 400}], \"data\": {\"url\": \"altair_data/altair-data-b7cd34f82694d1f7d67a84a844957e7b.json\", \"format\": {\"type\": \"json\"}}, \"title\": \"GO with_loss data - rSEA, threshold = 0.99, sort by Comp_adjP, cutoff Comp_adjP\"}, {\"vconcat\": [{\"mark\": \"circle\", \"encoding\": {\"size\": {\"type\": \"quantitative\", \"field\": \"avg_rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"Overall rank of pathway\", \"titleFontSize\": 12}, \"field\": \"avg_rank_label\", \"sort\": [\"271.6\", \"271.6\", \"271.6\", \"271.6\", \"271.6\", \"684.8\", \" 684.8\", \"  684.8\", \"684.8\", \" 684.8\", \"  684.8\", \"684.8\", \" 684.8\", \"  684.8\", \"684.8\", \" 684.8\", \"  684.8\", \"684.8\", \" 684.8\", \"  684.8\", \"803.4\", \"803.4\", \"803.4\", \"803.4\", \"803.4\", \"840.3\", \"840.3\", \"840.3\", \"840.3\", \"840.3\", \"844.8\", \"844.8\", \"844.8\", \"844.8\", \"844.8\", \"873.0\", \"873.0\", \"873.0\", \"873.0\", \"873.0\", \"928.5\", \"928.5\", \"928.5\", \"928.5\", \"928.5\", \"931.9\", \"931.9\", \"931.9\", \"931.9\", \"931.9\"]}}, \"width\": 400}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Pathway tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"regulation of nuclear-transcribed mRNA catabolic process, deadenylation-dependent decay (GO:1900151)\", \"regulation of nuclear-transcribed mRNA catabolic process, deadenylation-dependent decay (GO:1900151)\", \"regulation of nuclear-transcribed mRNA catabolic process, deadenylation-dependent decay (GO:1900151)\", \"regulation of nuclear-transcribed mRNA catabolic process, deadenylation-dependent decay (GO:1900151)\", \"regulation of nuclear-transcribed mRNA catabolic process, deadenylation-dependent decay (GO:1900151)\", \"positive regulation of nuclear-transcribed mRNA catabolic process, deadenylation-dependent decay (GO:1900153)\", \"positive regulation of nuclear-transcribed mRNA poly(A) tail shortening (GO:0060213)\", \"regulation of nuclear-transcribed mRNA poly(A) tail shortening (GO:0060211)\", \"positive regulation of nuclear-transcribed mRNA catabolic process, deadenylation-dependent decay (GO:1900153)\", \"positive regulation of nuclear-transcribed mRNA poly(A) tail shortening (GO:0060213)\", \"regulation of nuclear-transcribed mRNA poly(A) tail shortening (GO:0060211)\", \"positive regulation of nuclear-transcribed mRNA catabolic process, deadenylation-dependent decay (GO:1900153)\", \"positive regulation of nuclear-transcribed mRNA poly(A) tail shortening (GO:0060213)\", \"regulation of nuclear-transcribed mRNA poly(A) tail shortening (GO:0060211)\", \"positive regulation of nuclear-transcribed mRNA catabolic process, deadenylation-dependent decay (GO:1900153)\", \"positive regulation of nuclear-transcribed mRNA poly(A) tail shortening (GO:0060213)\", \"regulation of nuclear-transcribed mRNA poly(A) tail shortening (GO:0060211)\", \"positive regulation of nuclear-transcribed mRNA catabolic process, deadenylation-dependent decay (GO:1900153)\", \"positive regulation of nuclear-transcribed mRNA poly(A) tail shortening (GO:0060213)\", \"regulation of nuclear-transcribed mRNA poly(A) tail shortening (GO:0060211)\", \"regulation of DNA damage checkpoint (GO:2000001)\", \"regulation of DNA damage checkpoint (GO:2000001)\", \"regulation of DNA damage checkpoint (GO:2000001)\", \"regulation of DNA damage checkpoint (GO:2000001)\", \"regulation of DNA damage checkpoint (GO:2000001)\", \"regulation of T cell receptor signaling pathway (GO:0050856)\", \"regulation of T cell receptor signaling pathway (GO:0050856)\", \"regulation of T cell receptor signaling pathway (GO:0050856)\", \"regulation of T cell receptor signaling pathway (GO:0050856)\", \"regulation of T cell receptor signaling pathway (GO:0050856)\", \"phosphatidylinositol phosphorylation (GO:0046854)\", \"phosphatidylinositol phosphorylation (GO:0046854)\", \"phosphatidylinositol phosphorylation (GO:0046854)\", \"phosphatidylinositol phosphorylation (GO:0046854)\", \"phosphatidylinositol phosphorylation (GO:0046854)\", \"positive regulation of mRNA catabolic process (GO:0061014)\", \"positive regulation of mRNA catabolic process (GO:0061014)\", \"positive regulation of mRNA catabolic process (GO:0061014)\", \"positive regulation of mRNA catabolic process (GO:0061014)\", \"positive regulation of mRNA catabolic process (GO:0061014)\", \"positive regulation of gene silencing by miRNA (GO:2000637)\", \"positive regulation of gene silencing by miRNA (GO:2000637)\", \"positive regulation of gene silencing by miRNA (GO:2000637)\", \"positive regulation of gene silencing by miRNA (GO:2000637)\", \"positive regulation of gene silencing by miRNA (GO:2000637)\", \"histone mRNA metabolic process (GO:0008334)\", \"histone mRNA metabolic process (GO:0008334)\", \"histone mRNA metabolic process (GO:0008334)\", \"histone mRNA metabolic process (GO:0008334)\", \"histone mRNA metabolic process (GO:0008334)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": \"Cancer type\", \"titleFontSize\": 12}, \"field\": \"cancer_type\"}}, \"height\": 300, \"width\": 400}], \"data\": {\"url\": \"altair_data/altair-data-07ab29ae517c17fb69b333f1ea6702c9.json\", \"format\": {\"type\": \"json\"}}, \"title\": \"GO without_loss data - rSEA, threshold = 0.99, sort by Comp_adjP, cutoff Comp_adjP\"}, {\"vconcat\": [{\"mark\": \"circle\", \"encoding\": {\"size\": {\"type\": \"quantitative\", \"field\": \"avg_rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"Overall rank of pathway\", \"titleFontSize\": 12}, \"field\": \"avg_rank_label\", \"sort\": [\"658.7\", \"658.7\", \"658.7\", \"658.7\", \"658.7\", \"684.5\", \"684.5\", \"684.5\", \"684.5\", \"684.5\", \"786.2\", \"786.2\", \"786.2\", \"786.2\", \"786.2\", \"813.9\", \" 813.9\", \"813.9\", \" 813.9\", \"813.9\", \" 813.9\", \"813.9\", \" 813.9\", \"813.9\", \" 813.9\", \"818.2\", \"818.2\", \"818.2\", \"818.2\", \"818.2\", \"820.6\", \"820.6\", \"820.6\", \"820.6\", \"820.6\", \"823.4\", \"823.4\", \"823.4\", \"823.4\", \"823.4\", \"828.5\", \"828.5\", \"828.5\", \"828.5\", \"828.5\", \"828.9\", \"828.9\", \"828.9\", \"828.9\", \"828.9\"]}}, \"width\": 400}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Pathway tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"nucleobase-containing small molecule interconversion (GO:0015949)\", \"nucleobase-containing small molecule interconversion (GO:0015949)\", \"nucleobase-containing small molecule interconversion (GO:0015949)\", \"nucleobase-containing small molecule interconversion (GO:0015949)\", \"nucleobase-containing small molecule interconversion (GO:0015949)\", \"regulation of cell division (GO:0051302)\", \"regulation of cell division (GO:0051302)\", \"regulation of cell division (GO:0051302)\", \"regulation of cell division (GO:0051302)\", \"regulation of cell division (GO:0051302)\", \"antibacterial humoral response (GO:0019731)\", \"antibacterial humoral response (GO:0019731)\", \"antibacterial humoral response (GO:0019731)\", \"antibacterial humoral response (GO:0019731)\", \"antibacterial humoral response (GO:0019731)\", \"negative regulation of secretion by cell (GO:1903531)\", \"lysosome localization (GO:0032418)\", \"negative regulation of secretion by cell (GO:1903531)\", \"lysosome localization (GO:0032418)\", \"negative regulation of secretion by cell (GO:1903531)\", \"lysosome localization (GO:0032418)\", \"negative regulation of secretion by cell (GO:1903531)\", \"lysosome localization (GO:0032418)\", \"negative regulation of secretion by cell (GO:1903531)\", \"lysosome localization (GO:0032418)\", \"ribosome assembly (GO:0042255)\", \"ribosome assembly (GO:0042255)\", \"ribosome assembly (GO:0042255)\", \"ribosome assembly (GO:0042255)\", \"ribosome assembly (GO:0042255)\", \"Rho protein signal transduction (GO:0007266)\", \"Rho protein signal transduction (GO:0007266)\", \"Rho protein signal transduction (GO:0007266)\", \"Rho protein signal transduction (GO:0007266)\", \"Rho protein signal transduction (GO:0007266)\", \"positive regulation of DNA metabolic process (GO:0051054)\", \"positive regulation of DNA metabolic process (GO:0051054)\", \"positive regulation of DNA metabolic process (GO:0051054)\", \"positive regulation of DNA metabolic process (GO:0051054)\", \"positive regulation of DNA metabolic process (GO:0051054)\", \"nuclear-transcribed mRNA catabolic process, nonsense-mediated decay (GO:0000184)\", \"nuclear-transcribed mRNA catabolic process, nonsense-mediated decay (GO:0000184)\", \"nuclear-transcribed mRNA catabolic process, nonsense-mediated decay (GO:0000184)\", \"nuclear-transcribed mRNA catabolic process, nonsense-mediated decay (GO:0000184)\", \"nuclear-transcribed mRNA catabolic process, nonsense-mediated decay (GO:0000184)\", \"cytoskeleton-dependent intracellular transport (GO:0030705)\", \"cytoskeleton-dependent intracellular transport (GO:0030705)\", \"cytoskeleton-dependent intracellular transport (GO:0030705)\", \"cytoskeleton-dependent intracellular transport (GO:0030705)\", \"cytoskeleton-dependent intracellular transport (GO:0030705)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": \"Cancer type\", \"titleFontSize\": 12}, \"field\": \"cancer_type\"}}, \"height\": 300, \"width\": 400}], \"data\": {\"url\": \"altair_data/altair-data-45bd9bd762fdedb39a024004f01f2e63.json\", \"format\": {\"type\": \"json\"}}, \"title\": \"GO with_both data - rSEA, threshold = 0.99, sort by Comp_adjP, cutoff Comp_adjP\"}, {\"vconcat\": [{\"mark\": \"circle\", \"encoding\": {\"size\": {\"type\": \"quantitative\", \"field\": \"avg_rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"Overall rank of pathway\", \"titleFontSize\": 12}, \"field\": \"avg_rank_label\", \"sort\": [\"505.0\", \"505.0\", \"505.0\", \"505.0\", \"505.0\", \"634.7\", \"634.7\", \"634.7\", \"634.7\", \"634.7\", \"788.1\", \"788.1\", \"788.1\", \"788.1\", \"788.1\", \"891.2\", \"891.2\", \"891.2\", \"891.2\", \"891.2\", \"894.6\", \" 894.6\", \"894.6\", \" 894.6\", \"894.6\", \" 894.6\", \"894.6\", \" 894.6\", \"894.6\", \" 894.6\", \"907.9\", \"907.9\", \"907.9\", \"907.9\", \"907.9\", \"919.6\", \" 919.6\", \"919.6\", \" 919.6\", \"919.6\", \" 919.6\", \"919.6\", \" 919.6\", \"919.6\", \" 919.6\", \"923.9\", \"923.9\", \"923.9\", \"923.9\", \"923.9\"]}}, \"width\": 400}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Pathway tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"GDP-mannose metabolic process (GO:0019673)\", \"GDP-mannose metabolic process (GO:0019673)\", \"GDP-mannose metabolic process (GO:0019673)\", \"GDP-mannose metabolic process (GO:0019673)\", \"GDP-mannose metabolic process (GO:0019673)\", \"glutamine family amino acid biosynthetic process (GO:0009084)\", \"glutamine family amino acid biosynthetic process (GO:0009084)\", \"glutamine family amino acid biosynthetic process (GO:0009084)\", \"glutamine family amino acid biosynthetic process (GO:0009084)\", \"glutamine family amino acid biosynthetic process (GO:0009084)\", \"AMP metabolic process (GO:0046033)\", \"AMP metabolic process (GO:0046033)\", \"AMP metabolic process (GO:0046033)\", \"AMP metabolic process (GO:0046033)\", \"AMP metabolic process (GO:0046033)\", \"regulation of nuclear-transcribed mRNA catabolic process, deadenylation-dependent decay (GO:1900151)\", \"regulation of nuclear-transcribed mRNA catabolic process, deadenylation-dependent decay (GO:1900151)\", \"regulation of nuclear-transcribed mRNA catabolic process, deadenylation-dependent decay (GO:1900151)\", \"regulation of nuclear-transcribed mRNA catabolic process, deadenylation-dependent decay (GO:1900151)\", \"regulation of nuclear-transcribed mRNA catabolic process, deadenylation-dependent decay (GO:1900151)\", \"positive regulation of mRNA metabolic process (GO:1903313)\", \"positive regulation of mRNA catabolic process (GO:0061014)\", \"positive regulation of mRNA metabolic process (GO:1903313)\", \"positive regulation of mRNA catabolic process (GO:0061014)\", \"positive regulation of mRNA metabolic process (GO:1903313)\", \"positive regulation of mRNA catabolic process (GO:0061014)\", \"positive regulation of mRNA metabolic process (GO:1903313)\", \"positive regulation of mRNA catabolic process (GO:0061014)\", \"positive regulation of mRNA metabolic process (GO:1903313)\", \"positive regulation of mRNA catabolic process (GO:0061014)\", \"neuroepithelial cell differentiation (GO:0060563)\", \"neuroepithelial cell differentiation (GO:0060563)\", \"neuroepithelial cell differentiation (GO:0060563)\", \"neuroepithelial cell differentiation (GO:0060563)\", \"neuroepithelial cell differentiation (GO:0060563)\", \"positive regulation of nuclear-transcribed mRNA poly(A) tail shortening (GO:0060213)\", \"regulation of nuclear-transcribed mRNA poly(A) tail shortening (GO:0060211)\", \"positive regulation of nuclear-transcribed mRNA poly(A) tail shortening (GO:0060213)\", \"regulation of nuclear-transcribed mRNA poly(A) tail shortening (GO:0060211)\", \"positive regulation of nuclear-transcribed mRNA poly(A) tail shortening (GO:0060213)\", \"regulation of nuclear-transcribed mRNA poly(A) tail shortening (GO:0060211)\", \"positive regulation of nuclear-transcribed mRNA poly(A) tail shortening (GO:0060213)\", \"regulation of nuclear-transcribed mRNA poly(A) tail shortening (GO:0060211)\", \"positive regulation of nuclear-transcribed mRNA poly(A) tail shortening (GO:0060213)\", \"regulation of nuclear-transcribed mRNA poly(A) tail shortening (GO:0060211)\", \"arginine biosynthetic process (GO:0006526)\", \"arginine biosynthetic process (GO:0006526)\", \"arginine biosynthetic process (GO:0006526)\", \"arginine biosynthetic process (GO:0006526)\", \"arginine biosynthetic process (GO:0006526)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": \"Cancer type\", \"titleFontSize\": 12}, \"field\": \"cancer_type\"}}, \"height\": 300, \"width\": 400}], \"data\": {\"url\": \"altair_data/altair-data-ef859f93c4b9746099a466655b89041a.json\", \"format\": {\"type\": \"json\"}}, \"title\": \"GO without_both data - rSEA, threshold = 0.99, sort by Comp_adjP, cutoff Comp_adjP\"}], \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.8.1.json\"}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.VConcatChart(...)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alt.vconcat(*go_plots).configure_axis(grid=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Look at which gene sets are most enriched for each cancer type in each group, and see if there are similarities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_single_cancer(\n",
    "    enrich_file_path, \n",
    "    expr_file_path, \n",
    "    gmt_file_path, \n",
    "    plot_cancer_type,\n",
    "    enrichment_group,\n",
    "    pathway_min_size, \n",
    "    sort_col, \n",
    "    sort_asc, \n",
    "    cutoff_col, \n",
    "    cutoff=0.05\n",
    "):\n",
    "    \"\"\"Make a bubble chart where the x axis is pathway within a cancer type, the y axis is cancer type \n",
    "    (only one value), circle size is enrichment rank in cancer type, and circle color is gene set mean \n",
    "    change in expression in tumor compared to normal.\n",
    "    \n",
    "    Parameters:\n",
    "    enrich_file_path (str): The path to the file with rSEA enrichment analysis results.\n",
    "    expr_file_path (str): The path to the file with differential expression analysis results.\n",
    "    gmt_file_path (str): The path to the .gmt file that was used for the rSEA analysis.\n",
    "    plot_cancer_type (str): Which cancer type to make the plot for.\n",
    "    enrichment_group (str): Which CNV event group to look at. Either 'with_gain', 'without_gain', \n",
    "        'with_loss', 'without_loss', 'with_both', or 'without_both'. \n",
    "    pathway_min_size (int): The minimum number of genes a pathway/gene set must have to be included\n",
    "        in the chart.\n",
    "    sort_col (str): Which column in the rSEA results dataframe to use for ranking pathways.\n",
    "    sort_asc (bool): Whether to rank the sort column in ascending order.\n",
    "    cutoff_col (str): Which column in the rSEA results to use for a significance cutoff.\n",
    "    cutoff (float, optional): What cutoff to use when deciding whether to count a pathway as enriched.\n",
    "    \n",
    "    Returns:\n",
    "    altair.Chart: The chart.\n",
    "    pandas.DataFrame: The enrichment data used for the chart\n",
    "    pandas.DataFrame: The expression data used for the chart\n",
    "    pandas.DataFrame: The summary of the enrichment data, created in the process of making the chart\n",
    "    \"\"\"\n",
    "    \n",
    "    # Read in the expression data, selecting only the data for the specified group\n",
    "    all_expression_data = pd.read_csv(expr_file_path, sep=\"\\t\")\n",
    "    \n",
    "    all_expression_data = all_expression_data[\n",
    "        (all_expression_data[\"cancer_type\"] == plot_cancer_type) &\n",
    "        (all_expression_data[\"group\"] == enrichment_group)\n",
    "    ]\n",
    "\n",
    "    # Make a column where all increases are +1 and all decreases \n",
    "    # are -1, because these are ratioed abundances, so we can't \n",
    "    # compare magnitudes between genes--we can only compare whether \n",
    "    # there was a change, and whether it was positive or negative\n",
    "    all_expression_data = all_expression_data.assign(simplified_change=np.nan)\n",
    "\n",
    "    # adj p < 0.05 and change > 1 => +1\n",
    "    all_expression_data.loc[\n",
    "        (all_expression_data[\"change\"] > 0) & (all_expression_data[\"adj_p\"] < 0.05), \n",
    "        \"simplified_change\"\n",
    "    ] = 1\n",
    "\n",
    "    # adj p >= 0.05 and change > 1 => +0.5\n",
    "    all_expression_data.loc[(all_expression_data[\"change\"] > 0) & (all_expression_data[\"adj_p\"] >= 0.05),\n",
    "        \"simplified_change\"\n",
    "    ] = MSV\n",
    "\n",
    "    # change == 0 => 0\n",
    "    all_expression_data.loc[\n",
    "        all_expression_data[\"change\"] == 0,\n",
    "        \"simplified_change\"\n",
    "    ] = 0\n",
    "\n",
    "    # adj p >= 0.05 and change < 1 => -0.5\n",
    "    all_expression_data.loc[(all_expression_data[\"change\"] < 0) & (all_expression_data[\"adj_p\"] >= 0.05), \n",
    "        \"simplified_change\"\n",
    "    ] = -MSV\n",
    "\n",
    "    # adj p < 0.05 and change < 1 => -1\n",
    "    all_expression_data.loc[\n",
    "        (all_expression_data[\"change\"] < 0) & (all_expression_data[\"adj_p\"] < 0.05),\n",
    "        \"simplified_change\"\n",
    "    ] = -1\n",
    "\n",
    "    # Select just the proteins where we chose to reject the null hypothesis of no change\n",
    "    # We'll use this later to calculate average change in expression for each pathway\n",
    "    expression_data = all_expression_data[all_expression_data[\"adj_p\"] <= 0.05]\n",
    "    \n",
    "    # Read in the GMT file so we can have a list of which genes are in each pathway, in\n",
    "    # order to calculate the average change in expression for the pathways\n",
    "    with open(gmt_file_path, \"r\") as fp:\n",
    "        gene_lists = fp.readlines()\n",
    "\n",
    "    # Take the newline off the end, and split on the tab character to create a list of lists\n",
    "    gene_lists = [l.strip().split(\"\\t\") for l in gene_lists]\n",
    "\n",
    "    # Create a dataframe mapping pathway name and ID to contained genes\n",
    "    pathway_names = [l[0] for l in gene_lists]\n",
    "    pathway_genes = [l[2:] for l in gene_lists] # We skip index 1 -- it's a blank unused field.\n",
    "    pathway_data = pd.DataFrame({\n",
    "        \"pathway_name\": pathway_names,\n",
    "        \"pathway_genes\": pathway_genes\n",
    "    })\n",
    "\n",
    "    # Read in the enrichment data\n",
    "    enrichment_data = pd.\\\n",
    "    read_csv(enrich_file_path, sep=\"\\t\", index_col=0).\\\n",
    "    rename(columns={\"Name\": \"pathway_name\"})\n",
    "    \n",
    "    # Split the cancer_type_group column, then select the specified group\n",
    "    # This assumes that there are no underscores in the cancer type names,\n",
    "    # and that the cancer_type_group column has the format [CANCER]_[GROUP]\n",
    "    grp_split = enrichment_data[\"cancer_type_group\"].str.split(\"_\", n=1, expand=True)\n",
    "    \n",
    "    enrichment_data = enrichment_data.assign(\n",
    "        cancer_type=grp_split[0],\n",
    "        group=grp_split[1]\n",
    "    ).\\\n",
    "    drop(columns=\"cancer_type_group\")\n",
    "    \n",
    "    enrichment_data = enrichment_data[\n",
    "        (enrichment_data[\"cancer_type\"] == plot_cancer_type) &\n",
    "        (enrichment_data[\"group\"] == enrichment_group)\n",
    "    ]\n",
    "    \n",
    "    # Merge pathway data into the enrichment data, so we know which pathways were enriched\n",
    "    enrichment_data = enrichment_data.merge(\n",
    "        pathway_data,\n",
    "        how=\"left\",\n",
    "        left_on=\"pathway_name\",\n",
    "        right_on=\"pathway_name\",\n",
    "        validate=\"many_to_one\"\n",
    "    )\n",
    "    \n",
    "    # Select enrichment data where pathways meet the minimum size and the\n",
    "    # p values for them being enriched pass the cutoff\n",
    "    enrichment_data = enrichment_data[\n",
    "        (enrichment_data[\"pathway_genes\"].apply(len) >= pathway_min_size) &\n",
    "        (enrichment_data[cutoff_col] <= cutoff)\n",
    "    ]\n",
    "    \n",
    "    # Assign pathway ranks within each cancer type based on the sort_col.\n",
    "    enrichment_data = enrichment_data.\\\n",
    "    assign(\n",
    "        cancer_rank=enrichment_data.\\\n",
    "        groupby(\"cancer_type\")[sort_col].\\\n",
    "        rank(ascending=sort_asc)\n",
    "    ).\\\n",
    "    sort_values(by=[\"cancer_type\", \"cancer_rank\"]).\\\n",
    "    reset_index(drop=True)\n",
    "\n",
    "    # Select top 10 for our plot\n",
    "    sel_enrich = enrichment_data[:10]\n",
    "\n",
    "    # Calculate the mean expression for each pathway in each cancer type\n",
    "    mean_exprs = []\n",
    "\n",
    "    for idx in sel_enrich.index:\n",
    "        genes = sel_enrich.loc[idx, \"pathway_genes\"]\n",
    "        cancer_type = sel_enrich.loc[idx, \"cancer_type\"]\n",
    "\n",
    "        genes_expr = expression_data.\\\n",
    "        loc[\n",
    "            expression_data[\"protein_str\"].isin(genes) &\n",
    "            (expression_data[\"cancer_type\"] == cancer_type),\n",
    "            \"simplified_change\"\n",
    "        ].\\\n",
    "        mean()\n",
    "\n",
    "        mean_exprs.append(genes_expr)\n",
    "\n",
    "    sel_enrich = sel_enrich.assign(mean_expr=mean_exprs)\n",
    "\n",
    "    sel_enrich = sel_enrich.assign(\n",
    "        rank_size=1 / sel_enrich[\"cancer_rank\"],\n",
    "        rank_label=sel_enrich[\"cancer_rank\"].apply(lambda x: round(x, 2)).astype(str))\n",
    "    \n",
    "    # Take care of duplicates for the upper plot\n",
    "    for i in range(10):\n",
    "        sel_enrich[\"rank_label\"] = sel_enrich[\"rank_label\"].where(\n",
    "            cond=~(sel_enrich.duplicated(subset=[\"cancer_type\", \"rank_label\"], keep=\"first\")),\n",
    "            other=\" \" + sel_enrich[\"rank_label\"])\n",
    "    \n",
    "    # Make our plots!\n",
    "    base = alt.Chart(sel_enrich).encode(\n",
    "         y=alt.Y(\n",
    "            \"cancer_type:N\",\n",
    "            axis=alt.Axis(\n",
    "                title=None\n",
    "            ),\n",
    "        ),\n",
    "        size=alt.Size(\n",
    "            \"rank_size:Q\",\n",
    "            legend=None\n",
    "        ),\n",
    "        color=alt.Color(\n",
    "            \"mean_expr:Q\",\n",
    "            scale=alt.Scale(\n",
    "                scheme=\"blueorange\",\n",
    "                domain=[-1, 1]\n",
    "            ),\n",
    "            legend=alt.Legend(\n",
    "                title=\"Tumor expression\"\n",
    "            )\n",
    "        )\n",
    "    )\n",
    "\n",
    "    bubbles = base.mark_circle().encode(\n",
    "        x=alt.X(\n",
    "            \"pathway_name:N\",\n",
    "            sort=sel_enrich[\"pathway_name\"].values,\n",
    "            axis=alt.Axis(\n",
    "                labelAngle=-30,\n",
    "                labelFontSize=12,\n",
    "                labelLimit=500,\n",
    "                title=\"\",\n",
    "                titleFontSize=16\n",
    "            )\n",
    "        )\n",
    "    )\n",
    "    \n",
    "    numbers = base.mark_circle().encode(\n",
    "        x=alt.X(\n",
    "            \"rank_label:N\",\n",
    "            sort=sel_enrich[\"pathway_name\"].values,\n",
    "            axis=alt.Axis(\n",
    "                labelAngle=0,\n",
    "                labelFontSize=12,\n",
    "                labelLimit=500,\n",
    "                labelPadding=15,\n",
    "                title=\"Pathway rank\",\n",
    "                titleFontSize=10\n",
    "            )\n",
    "        )\n",
    "    )\n",
    "    \n",
    "    chart = alt.layer(bubbles, numbers).resolve_scale(\n",
    "        x=\"independent\"\n",
    "    ).properties(\n",
    "        width=400,\n",
    "        height=30\n",
    "    )\n",
    "    \n",
    "    return chart, enrichment_data, all_expression_data, sel_enrich"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def group_chart_by_cancer(group):\n",
    "    cancer_types = [\n",
    "    #     \"brca\",\n",
    "    #     \"ccrcc\",\n",
    "        \"colon\",\n",
    "    #     \"endometrial\",\n",
    "    #     \"gbm\",\n",
    "        \"hnscc\",\n",
    "        \"lscc\",\n",
    "        \"luad\",\n",
    "        \"ovarian\"\n",
    "    ]\n",
    "\n",
    "    charts = [    \n",
    "        plot_single_cancer(\n",
    "            enrich_file_path=reactome_results_file, \n",
    "            expr_file_path=diff_expr_file, \n",
    "            gmt_file_path=reactome_gmt,\n",
    "            plot_cancer_type=ct,\n",
    "            enrichment_group=group,\n",
    "            pathway_min_size=5,\n",
    "            sort_col=\"Comp.0.99.adjP\",\n",
    "            sort_asc=True,\n",
    "            cutoff_col=\"Comp.0.99.adjP\",\n",
    "            cutoff=0.05\n",
    "        )[0]\n",
    "        for ct in cancer_types\n",
    "    ]\n",
    "\n",
    "    return alt.vconcat(*charts).properties(\n",
    "        title=f\"Reactome data: {group} group\"\n",
    "    ).configure_title(\n",
    "        anchor=\"middle\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-62db351c0d9d476493e453e6c802da42\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-62db351c0d9d476493e453e6c802da42\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-62db351c0d9d476493e453e6c802da42\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.8.1?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function loadScript(lib) {\n",
       "      return new Promise(function(resolve, reject) {\n",
       "        var s = document.createElement('script');\n",
       "        s.src = paths[lib];\n",
       "        s.async = true;\n",
       "        s.onload = () => resolve(paths[lib]);\n",
       "        s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "        document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "      });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else if (typeof vegaEmbed === \"function\") {\n",
       "      displayChart(vegaEmbed);\n",
       "    } else {\n",
       "      loadScript(\"vega\")\n",
       "        .then(() => loadScript(\"vega-lite\"))\n",
       "        .then(() => loadScript(\"vega-embed\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}, \"title\": {\"anchor\": \"middle\"}}, \"vconcat\": [{\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"GP1b-IX-V activation signalling (R-HSA-430116)\", \"KSRP (KHSRP) binds and destabilizes mRNA (R-HSA-450604)\", \"Negative regulation of NOTCH4 signaling (R-HSA-9604323)\", \"Rap1 signalling (R-HSA-392517)\", \"Cell recruitment (pro-inflammatory response) (R-HSA-9664424)\", \"Inflammasomes (R-HSA-622312)\", \"Purinergic signaling in leishmaniasis infection (R-HSA-9660826)\", \"The NLRP3 inflammasome (R-HSA-844456)\", \"Activation of IRF3/IRF7 mediated by TBK1/IKK epsilon (R-HSA-936964)\", \"TICAM1-dependent activation of IRF3/IRF7 (R-HSA-9013973)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"GP1b-IX-V activation signalling (R-HSA-430116)\", \"KSRP (KHSRP) binds and destabilizes mRNA (R-HSA-450604)\", \"Negative regulation of NOTCH4 signaling (R-HSA-9604323)\", \"Rap1 signalling (R-HSA-392517)\", \"Cell recruitment (pro-inflammatory response) (R-HSA-9664424)\", \"Inflammasomes (R-HSA-622312)\", \"Purinergic signaling in leishmaniasis infection (R-HSA-9660826)\", \"The NLRP3 inflammasome (R-HSA-844456)\", \"Activation of IRF3/IRF7 mediated by TBK1/IKK epsilon (R-HSA-936964)\", \"TICAM1-dependent activation of IRF3/IRF7 (R-HSA-9013973)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-3ffda9740a4a73b4aa78aeed66d4b8ca.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}, {\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"TNF receptor superfamily (TNFSF) members mediating non-canonical NF-kB pathway (R-HSA-5676594)\", \"tRNA processing in the mitochondrion (R-HSA-6785470)\", \"Abortive elongation of HIV-1 transcript in the absence of Tat (R-HSA-167242)\", \"Dual incision in TC-NER (R-HSA-6782135)\", \"FGFR2 mutant receptor activation (R-HSA-1839126)\", \"Formation of the Early Elongation Complex (R-HSA-113418)\", \"Formation of the HIV-1 Early Elongation Complex (R-HSA-167158)\", \"Gap-filling DNA repair synthesis and ligation in TC-NER (R-HSA-6782210)\", \"MicroRNA (miRNA) biogenesis (R-HSA-203927)\", \"PIWI-interacting RNA (piRNA) biogenesis (R-HSA-5601884)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"TNF receptor superfamily (TNFSF) members mediating non-canonical NF-kB pathway (R-HSA-5676594)\", \"tRNA processing in the mitochondrion (R-HSA-6785470)\", \"Abortive elongation of HIV-1 transcript in the absence of Tat (R-HSA-167242)\", \"Dual incision in TC-NER (R-HSA-6782135)\", \"FGFR2 mutant receptor activation (R-HSA-1839126)\", \"Formation of the Early Elongation Complex (R-HSA-113418)\", \"Formation of the HIV-1 Early Elongation Complex (R-HSA-167158)\", \"Gap-filling DNA repair synthesis and ligation in TC-NER (R-HSA-6782210)\", \"MicroRNA (miRNA) biogenesis (R-HSA-203927)\", \"PIWI-interacting RNA (piRNA) biogenesis (R-HSA-5601884)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-8770b163c2a2967e8031f6497c397e73.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}, {\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Translation of Replicase and Assembly of the Replication Transcription Complex (R-HSA-9679504)\", \"TP53 regulates transcription of several additional cell death genes whose specific roles in p53-dependent apoptosis remain uncertain (R-HSA-6803205)\", \"HDACs deacetylate histones (R-HSA-3214815)\", \"Negative epigenetic regulation of rRNA expression (R-HSA-5250941)\", \"NoRC negatively regulates rRNA expression (R-HSA-427413)\", \"Synthesis of PE (R-HSA-1483213)\", \"Triglyceride biosynthesis (R-HSA-75109)\", \"Amino acid transport across the plasma membrane (R-HSA-352230)\", \"HDL remodeling (R-HSA-8964058)\", \"Axonal growth inhibition (RHOA activation) (R-HSA-193634)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"Translation of Replicase and Assembly of the Replication Transcription Complex (R-HSA-9679504)\", \"TP53 regulates transcription of several additional cell death genes whose specific roles in p53-dependent apoptosis remain uncertain (R-HSA-6803205)\", \"HDACs deacetylate histones (R-HSA-3214815)\", \"Negative epigenetic regulation of rRNA expression (R-HSA-5250941)\", \"NoRC negatively regulates rRNA expression (R-HSA-427413)\", \"Synthesis of PE (R-HSA-1483213)\", \"Triglyceride biosynthesis (R-HSA-75109)\", \"Amino acid transport across the plasma membrane (R-HSA-352230)\", \"HDL remodeling (R-HSA-8964058)\", \"Axonal growth inhibition (RHOA activation) (R-HSA-193634)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-415f35b7208808ed40436fa2bf57943a.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}, {\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Activation of RAC1 (R-HSA-428540)\", \"Nef and signal transduction (R-HSA-164944)\", \"RHO GTPases activate PAKs (R-HSA-5627123)\", \"Sema3A PAK dependent Axon repulsion (R-HSA-399954)\", \"MECP2 regulates transcription factors (R-HSA-9022707)\", \"Activated PKN1 stimulates transcription of AR (androgen receptor) regulated genes KLK2 and KLK3 (R-HSA-5625886)\", \"BMAL1:CLOCK,NPAS2 activates circadian gene expression (R-HSA-1368108)\", \"Mitochondrial biogenesis (R-HSA-1592230)\", \"RORA activates gene expression (R-HSA-1368082)\", \"Recycling of bile acids and salts (R-HSA-159418)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"Activation of RAC1 (R-HSA-428540)\", \"Nef and signal transduction (R-HSA-164944)\", \"RHO GTPases activate PAKs (R-HSA-5627123)\", \"Sema3A PAK dependent Axon repulsion (R-HSA-399954)\", \"MECP2 regulates transcription factors (R-HSA-9022707)\", \"Activated PKN1 stimulates transcription of AR (androgen receptor) regulated genes KLK2 and KLK3 (R-HSA-5625886)\", \"BMAL1:CLOCK,NPAS2 activates circadian gene expression (R-HSA-1368108)\", \"Mitochondrial biogenesis (R-HSA-1592230)\", \"RORA activates gene expression (R-HSA-1368082)\", \"Recycling of bile acids and salts (R-HSA-159418)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-6076070fa94c206a25d0a8e04d3b7a67.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}, {\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Glutathione conjugation (R-HSA-156590)\", \"Glutathione synthesis and recycling (R-HSA-174403)\", \"Metabolic disorders of biological oxidation enzymes (R-HSA-5579029)\", \"Aggrephagy (R-HSA-9646399)\", \"Activation of BAD and translocation to mitochondria (R-HSA-111447)\", \"Activation of BH3-only proteins (R-HSA-114452)\", \"Beta-oxidation of pristanoyl-CoA (R-HSA-389887)\", \"Competing endogenous RNAs (ceRNAs) regulate PTEN translation (R-HSA-8948700)\", \"Dissolution of Fibrin Clot (R-HSA-75205)\", \"MicroRNA (miRNA) biogenesis (R-HSA-203927)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"Glutathione conjugation (R-HSA-156590)\", \"Glutathione synthesis and recycling (R-HSA-174403)\", \"Metabolic disorders of biological oxidation enzymes (R-HSA-5579029)\", \"Aggrephagy (R-HSA-9646399)\", \"Activation of BAD and translocation to mitochondria (R-HSA-111447)\", \"Activation of BH3-only proteins (R-HSA-114452)\", \"Beta-oxidation of pristanoyl-CoA (R-HSA-389887)\", \"Competing endogenous RNAs (ceRNAs) regulate PTEN translation (R-HSA-8948700)\", \"Dissolution of Fibrin Clot (R-HSA-75205)\", \"MicroRNA (miRNA) biogenesis (R-HSA-203927)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-d34e806c714f84b7627ec98c051194af.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}], \"title\": \"Reactome data: with_gain group\", \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.8.1.json\"}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.VConcatChart(...)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "group_chart_by_cancer(\"with_gain\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-ae756e3c72844544b18d1d04778b0e67\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-ae756e3c72844544b18d1d04778b0e67\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-ae756e3c72844544b18d1d04778b0e67\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.8.1?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function loadScript(lib) {\n",
       "      return new Promise(function(resolve, reject) {\n",
       "        var s = document.createElement('script');\n",
       "        s.src = paths[lib];\n",
       "        s.async = true;\n",
       "        s.onload = () => resolve(paths[lib]);\n",
       "        s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "        document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "      });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else if (typeof vegaEmbed === \"function\") {\n",
       "      displayChart(vegaEmbed);\n",
       "    } else {\n",
       "      loadScript(\"vega\")\n",
       "        .then(() => loadScript(\"vega-lite\"))\n",
       "        .then(() => loadScript(\"vega-embed\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}, \"title\": {\"anchor\": \"middle\"}}, \"vconcat\": [{\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Complex I biogenesis (R-HSA-6799198)\", \"Ubiquinol biosynthesis (R-HSA-2142789)\", \"Post-transcriptional silencing by small RNAs (R-HSA-426496)\", \"Advanced glycosylation endproduct receptor signaling (R-HSA-879415)\", \"ATF6 (ATF6-alpha) activates chaperone genes (R-HSA-381183)\", \"ATF6 (ATF6-alpha) activates chaperones (R-HSA-381033)\", \"Assembly of Viral Components at the Budding Site (R-HSA-168316)\", \"Scavenging by Class F Receptors (R-HSA-3000484)\", \"Virus Assembly and Release (R-HSA-168268)\", \"Competing endogenous RNAs (ceRNAs) regulate PTEN translation (R-HSA-8948700)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"Complex I biogenesis (R-HSA-6799198)\", \"Ubiquinol biosynthesis (R-HSA-2142789)\", \"Post-transcriptional silencing by small RNAs (R-HSA-426496)\", \"Advanced glycosylation endproduct receptor signaling (R-HSA-879415)\", \"ATF6 (ATF6-alpha) activates chaperone genes (R-HSA-381183)\", \"ATF6 (ATF6-alpha) activates chaperones (R-HSA-381033)\", \"Assembly of Viral Components at the Budding Site (R-HSA-168316)\", \"Scavenging by Class F Receptors (R-HSA-3000484)\", \"Virus Assembly and Release (R-HSA-168268)\", \"Competing endogenous RNAs (ceRNAs) regulate PTEN translation (R-HSA-8948700)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-8f551a65d68875aa22612a9fd0c12ad7.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}, {\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Diseases associated with visual transduction (R-HSA-2474795)\", \"Diseases of the neuronal system (R-HSA-9675143)\", \"Retinoid cycle disease events (R-HSA-2453864)\", \"The canonical retinoid cycle in rods (twilight vision) (R-HSA-2453902)\", \"Voltage gated Potassium channels (R-HSA-1296072)\", \"Interleukin-37 signaling (R-HSA-9008059)\", \"The IPAF inflammasome (R-HSA-844623)\", \"FGFR2 alternative splicing (R-HSA-6803529)\", \"Acetylcholine regulates insulin secretion (R-HSA-399997)\", \"Fatty Acids bound to GPR40 (FFAR1) regulate insulin secretion (R-HSA-434316)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"Diseases associated with visual transduction (R-HSA-2474795)\", \"Diseases of the neuronal system (R-HSA-9675143)\", \"Retinoid cycle disease events (R-HSA-2453864)\", \"The canonical retinoid cycle in rods (twilight vision) (R-HSA-2453902)\", \"Voltage gated Potassium channels (R-HSA-1296072)\", \"Interleukin-37 signaling (R-HSA-9008059)\", \"The IPAF inflammasome (R-HSA-844623)\", \"FGFR2 alternative splicing (R-HSA-6803529)\", \"Acetylcholine regulates insulin secretion (R-HSA-399997)\", \"Fatty Acids bound to GPR40 (FFAR1) regulate insulin secretion (R-HSA-434316)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-34ec47394b56a273ac46bc0d24f1dc50.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}, {\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"TNFR1-mediated ceramide production (R-HSA-5626978)\", \"POU5F1 (OCT4), SOX2, NANOG repress genes related to differentiation (R-HSA-2892245)\", \"Nucleotide salvage (R-HSA-8956321)\", \"Pyrimidine catabolism (R-HSA-73621)\", \"Pyrimidine salvage (R-HSA-73614)\", \"Fibronectin matrix formation (R-HSA-1566977)\", \"RUNX2 regulates genes involved in cell migration (R-HSA-8941332)\", \"Signal transduction by L1 (R-HSA-445144)\", \"LDL clearance (R-HSA-8964038)\", \"Plasma lipoprotein clearance (R-HSA-8964043)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"TNFR1-mediated ceramide production (R-HSA-5626978)\", \"POU5F1 (OCT4), SOX2, NANOG repress genes related to differentiation (R-HSA-2892245)\", \"Nucleotide salvage (R-HSA-8956321)\", \"Pyrimidine catabolism (R-HSA-73621)\", \"Pyrimidine salvage (R-HSA-73614)\", \"Fibronectin matrix formation (R-HSA-1566977)\", \"RUNX2 regulates genes involved in cell migration (R-HSA-8941332)\", \"Signal transduction by L1 (R-HSA-445144)\", \"LDL clearance (R-HSA-8964038)\", \"Plasma lipoprotein clearance (R-HSA-8964043)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-e6da24e92bd591ab8b8fbcaea8a9deb8.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}, {\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Tight junction interactions (R-HSA-420029)\", \"Negative regulation of activity of TFAP2 (AP-2) family transcription factors (R-HSA-8866904)\", \"Postmitotic nuclear pore complex (NPC) reformation (R-HSA-9615933)\", \"Processing and activation of SUMO (R-HSA-3215018)\", \"SUMO is transferred from E1 to E2 (UBE2I, UBC9) (R-HSA-3065678)\", \"SUMOylation of DNA methylation proteins (R-HSA-4655427)\", \"SUMOylation of chromatin organization proteins (R-HSA-4551638)\", \"Transcriptional regulation by the AP-2 (TFAP2) family of transcription factors (R-HSA-8864260)\", \"Transcriptional regulation by small RNAs (R-HSA-5578749)\", \"RNF mutants show enhanced WNT signaling and proliferation (R-HSA-5340588)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"Tight junction interactions (R-HSA-420029)\", \"Negative regulation of activity of TFAP2 (AP-2) family transcription factors (R-HSA-8866904)\", \"Postmitotic nuclear pore complex (NPC) reformation (R-HSA-9615933)\", \"Processing and activation of SUMO (R-HSA-3215018)\", \"SUMO is transferred from E1 to E2 (UBE2I, UBC9) (R-HSA-3065678)\", \"SUMOylation of DNA methylation proteins (R-HSA-4655427)\", \"SUMOylation of chromatin organization proteins (R-HSA-4551638)\", \"Transcriptional regulation by the AP-2 (TFAP2) family of transcription factors (R-HSA-8864260)\", \"Transcriptional regulation by small RNAs (R-HSA-5578749)\", \"RNF mutants show enhanced WNT signaling and proliferation (R-HSA-5340588)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-144358e0507765185376c7021b408b87.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}, {\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Insulin-like Growth Factor-2 mRNA Binding Proteins (IGF2BPs/IMPs/VICKZs) bind RNA (R-HSA-428359)\", \"Vitamin B5 (pantothenate) metabolism (R-HSA-199220)\", \"Ephrin signaling (R-HSA-3928664)\", \"Competing endogenous RNAs (ceRNAs) regulate PTEN translation (R-HSA-8948700)\", \"Post-transcriptional silencing by small RNAs (R-HSA-426496)\", \"Regulation of PTEN mRNA translation (R-HSA-8943723)\", \"Regulation of RUNX1 Expression and Activity (R-HSA-8934593)\", \"Acetylcholine Neurotransmitter Release Cycle (R-HSA-264642)\", \"Basigin interactions (R-HSA-210991)\", \"Diseases associated with the TLR signaling cascade (R-HSA-5602358)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"Insulin-like Growth Factor-2 mRNA Binding Proteins (IGF2BPs/IMPs/VICKZs) bind RNA (R-HSA-428359)\", \"Vitamin B5 (pantothenate) metabolism (R-HSA-199220)\", \"Ephrin signaling (R-HSA-3928664)\", \"Competing endogenous RNAs (ceRNAs) regulate PTEN translation (R-HSA-8948700)\", \"Post-transcriptional silencing by small RNAs (R-HSA-426496)\", \"Regulation of PTEN mRNA translation (R-HSA-8943723)\", \"Regulation of RUNX1 Expression and Activity (R-HSA-8934593)\", \"Acetylcholine Neurotransmitter Release Cycle (R-HSA-264642)\", \"Basigin interactions (R-HSA-210991)\", \"Diseases associated with the TLR signaling cascade (R-HSA-5602358)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-eee5e317cf46465c5e84d62476aa1be2.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}], \"title\": \"Reactome data: without_gain group\", \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.8.1.json\"}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.VConcatChart(...)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "group_chart_by_cancer(\"without_gain\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-d1308b3f4a5f4aafb2d75d4ac25f4e43\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-d1308b3f4a5f4aafb2d75d4ac25f4e43\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-d1308b3f4a5f4aafb2d75d4ac25f4e43\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.8.1?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function loadScript(lib) {\n",
       "      return new Promise(function(resolve, reject) {\n",
       "        var s = document.createElement('script');\n",
       "        s.src = paths[lib];\n",
       "        s.async = true;\n",
       "        s.onload = () => resolve(paths[lib]);\n",
       "        s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "        document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "      });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else if (typeof vegaEmbed === \"function\") {\n",
       "      displayChart(vegaEmbed);\n",
       "    } else {\n",
       "      loadScript(\"vega\")\n",
       "        .then(() => loadScript(\"vega-lite\"))\n",
       "        .then(() => loadScript(\"vega-embed\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}, \"title\": {\"anchor\": \"middle\"}}, \"vconcat\": [{\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Nef Mediated CD4 Down-regulation (R-HSA-167590)\", \"Nef Mediated CD8 Down-regulation (R-HSA-182218)\", \"Response of Mtb to phagocytosis (R-HSA-9637690)\", \"Suppression of phagosomal maturation (R-HSA-9637687)\", \"Activation of the TFAP2 (AP-2) family of transcription factors (R-HSA-8866907)\", \"Negative regulation of activity of TFAP2 (AP-2) family transcription factors (R-HSA-8866904)\", \"Transcriptional regulation by the AP-2 (TFAP2) family of transcription factors (R-HSA-8864260)\", \"Advanced glycosylation endproduct receptor signaling (R-HSA-879415)\", \"COPI-independent Golgi-to-ER retrograde traffic (R-HSA-6811436)\", \"MHC class II antigen presentation (R-HSA-2132295)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"Nef Mediated CD4 Down-regulation (R-HSA-167590)\", \"Nef Mediated CD8 Down-regulation (R-HSA-182218)\", \"Response of Mtb to phagocytosis (R-HSA-9637690)\", \"Suppression of phagosomal maturation (R-HSA-9637687)\", \"Activation of the TFAP2 (AP-2) family of transcription factors (R-HSA-8866907)\", \"Negative regulation of activity of TFAP2 (AP-2) family transcription factors (R-HSA-8866904)\", \"Transcriptional regulation by the AP-2 (TFAP2) family of transcription factors (R-HSA-8864260)\", \"Advanced glycosylation endproduct receptor signaling (R-HSA-879415)\", \"COPI-independent Golgi-to-ER retrograde traffic (R-HSA-6811436)\", \"MHC class II antigen presentation (R-HSA-2132295)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-84ee9f4be08c7b86329c5b488cc7be15.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}, {\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Creatine metabolism (R-HSA-71288)\", \"GABA synthesis, release, reuptake and degradation (R-HSA-888590)\", \"Na+/Cl- dependent neurotransmitter transporters (R-HSA-442660)\", \"Transcriptional regulation of granulopoiesis (R-HSA-9616222)\", \"Synthesis of PE (R-HSA-1483213)\", \"A tetrasaccharide linker sequence is required for GAG synthesis (R-HSA-1971475)\", \"ABC transporter disorders (R-HSA-5619084)\", \"ABC-family proteins mediated transport (R-HSA-382556)\", \"ADORA2B mediated anti-inflammatory cytokines production (R-HSA-9660821)\", \"AKT phosphorylates targets in the nucleus (R-HSA-198693)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"Creatine metabolism (R-HSA-71288)\", \"GABA synthesis, release, reuptake and degradation (R-HSA-888590)\", \"Na+/Cl- dependent neurotransmitter transporters (R-HSA-442660)\", \"Transcriptional regulation of granulopoiesis (R-HSA-9616222)\", \"Synthesis of PE (R-HSA-1483213)\", \"A tetrasaccharide linker sequence is required for GAG synthesis (R-HSA-1971475)\", \"ABC transporter disorders (R-HSA-5619084)\", \"ABC-family proteins mediated transport (R-HSA-382556)\", \"ADORA2B mediated anti-inflammatory cytokines production (R-HSA-9660821)\", \"AKT phosphorylates targets in the nucleus (R-HSA-198693)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-535bce1212b3d28f4686855379d9f7f8.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}, {\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Cyclin A/B1/B2 associated events during G2/M transition (R-HSA-69273)\", \"Chylomicron clearance (R-HSA-8964026)\", \"Calcineurin activates NFAT (R-HSA-2025928)\", \"Formation of the ternary complex, and subsequently, the 43S complex (R-HSA-72695)\", \"EGR2 and SOX10-mediated initiation of Schwann cell myelination (R-HSA-9619665)\", \"LDL clearance (R-HSA-8964038)\", \"Plasma lipoprotein clearance (R-HSA-8964043)\", \"VLDLR internalisation and degradation (R-HSA-8866427)\", \"Physiological factors (R-HSA-5578768)\", \"Iron uptake and transport (R-HSA-917937)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"Cyclin A/B1/B2 associated events during G2/M transition (R-HSA-69273)\", \"Chylomicron clearance (R-HSA-8964026)\", \"Calcineurin activates NFAT (R-HSA-2025928)\", \"Formation of the ternary complex, and subsequently, the 43S complex (R-HSA-72695)\", \"EGR2 and SOX10-mediated initiation of Schwann cell myelination (R-HSA-9619665)\", \"LDL clearance (R-HSA-8964038)\", \"Plasma lipoprotein clearance (R-HSA-8964043)\", \"VLDLR internalisation and degradation (R-HSA-8866427)\", \"Physiological factors (R-HSA-5578768)\", \"Iron uptake and transport (R-HSA-917937)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-98e785e087b57c797c805ad729f19ac4.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}, {\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Cyclin A/B1/B2 associated events during G2/M transition (R-HSA-69273)\", \"Cyclin D associated events in G1 (R-HSA-69231)\", \"G1 Phase (R-HSA-69236)\", \"p130Cas linkage to MAPK signaling for integrins (R-HSA-372708)\", \"Aflatoxin activation and detoxification (R-HSA-5423646)\", \"Other semaphorin interactions (R-HSA-416700)\", \"Lysine catabolism (R-HSA-71064)\", \"Heme biosynthesis (R-HSA-189451)\", \"Urea cycle (R-HSA-70635)\", \"IKK complex recruitment mediated by RIP1 (R-HSA-937041)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"Cyclin A/B1/B2 associated events during G2/M transition (R-HSA-69273)\", \"Cyclin D associated events in G1 (R-HSA-69231)\", \"G1 Phase (R-HSA-69236)\", \"p130Cas linkage to MAPK signaling for integrins (R-HSA-372708)\", \"Aflatoxin activation and detoxification (R-HSA-5423646)\", \"Other semaphorin interactions (R-HSA-416700)\", \"Lysine catabolism (R-HSA-71064)\", \"Heme biosynthesis (R-HSA-189451)\", \"Urea cycle (R-HSA-70635)\", \"IKK complex recruitment mediated by RIP1 (R-HSA-937041)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-497608937b6226e0217c26084978db65.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}, {\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"OAS antiviral response (R-HSA-8983711)\", \"Infection with Mycobacterium tuberculosis (R-HSA-9635486)\", \"Response of Mtb to phagocytosis (R-HSA-9637690)\", \"Suppression of apoptosis (R-HSA-9635465)\", \"FGFR1 ligand binding and activation (R-HSA-190242)\", \"FGFR1c ligand binding and activation (R-HSA-190373)\", \"FGFR3 mutant receptor activation (R-HSA-2033514)\", \"FGFR3b ligand binding and activation (R-HSA-190371)\", \"FGFR4 ligand binding and activation (R-HSA-190322)\", \"FGFRL1 modulation of FGFR1 signaling (R-HSA-5658623)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"OAS antiviral response (R-HSA-8983711)\", \"Infection with Mycobacterium tuberculosis (R-HSA-9635486)\", \"Response of Mtb to phagocytosis (R-HSA-9637690)\", \"Suppression of apoptosis (R-HSA-9635465)\", \"FGFR1 ligand binding and activation (R-HSA-190242)\", \"FGFR1c ligand binding and activation (R-HSA-190373)\", \"FGFR3 mutant receptor activation (R-HSA-2033514)\", \"FGFR3b ligand binding and activation (R-HSA-190371)\", \"FGFR4 ligand binding and activation (R-HSA-190322)\", \"FGFRL1 modulation of FGFR1 signaling (R-HSA-5658623)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-64d61bcfb6a4f59e663642ddd737777d.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}], \"title\": \"Reactome data: with_loss group\", \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.8.1.json\"}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.VConcatChart(...)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "group_chart_by_cancer(\"with_loss\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-f3f0aca61a774d66b70d0d04045e5bba\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-f3f0aca61a774d66b70d0d04045e5bba\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-f3f0aca61a774d66b70d0d04045e5bba\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.8.1?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function loadScript(lib) {\n",
       "      return new Promise(function(resolve, reject) {\n",
       "        var s = document.createElement('script');\n",
       "        s.src = paths[lib];\n",
       "        s.async = true;\n",
       "        s.onload = () => resolve(paths[lib]);\n",
       "        s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "        document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "      });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else if (typeof vegaEmbed === \"function\") {\n",
       "      displayChart(vegaEmbed);\n",
       "    } else {\n",
       "      loadScript(\"vega\")\n",
       "        .then(() => loadScript(\"vega-lite\"))\n",
       "        .then(() => loadScript(\"vega-embed\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}, \"title\": {\"anchor\": \"middle\"}}, \"vconcat\": [{\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"RORA activates gene expression (R-HSA-1368082)\", \"Insulin receptor recycling (R-HSA-77387)\", \"Transferrin endocytosis and recycling (R-HSA-917977)\", \"Maturation of nucleoprotein (R-HSA-9683610)\", \"Negative regulation of activity of TFAP2 (AP-2) family transcription factors (R-HSA-8866904)\", \"Postmitotic nuclear pore complex (NPC) reformation (R-HSA-9615933)\", \"Processing and activation of SUMO (R-HSA-3215018)\", \"Regulation of IFNG signaling (R-HSA-877312)\", \"SUMO is conjugated to E1 (UBA2:SAE1) (R-HSA-3065676)\", \"SUMO is proteolytically processed (R-HSA-3065679)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"RORA activates gene expression (R-HSA-1368082)\", \"Insulin receptor recycling (R-HSA-77387)\", \"Transferrin endocytosis and recycling (R-HSA-917977)\", \"Maturation of nucleoprotein (R-HSA-9683610)\", \"Negative regulation of activity of TFAP2 (AP-2) family transcription factors (R-HSA-8866904)\", \"Postmitotic nuclear pore complex (NPC) reformation (R-HSA-9615933)\", \"Processing and activation of SUMO (R-HSA-3215018)\", \"Regulation of IFNG signaling (R-HSA-877312)\", \"SUMO is conjugated to E1 (UBA2:SAE1) (R-HSA-3065676)\", \"SUMO is proteolytically processed (R-HSA-3065679)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-d7cb4d4ca57680afcfcdbe3ad454da51.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}, {\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Activation of SMO (R-HSA-5635838)\", \"Disinhibition of SNARE formation (R-HSA-114516)\", \"Activated PKN1 stimulates transcription of AR (androgen receptor) regulated genes KLK2 and KLK3 (R-HSA-5625886)\", \"Glutamate and glutamine metabolism (R-HSA-8964539)\", \"Metabolism of non-coding RNA (R-HSA-194441)\", \"snRNP Assembly (R-HSA-191859)\", \"Fertilization (R-HSA-1187000)\", \"Sperm Motility And Taxes (R-HSA-1300642)\", \"Interleukin-23 signaling (R-HSA-9020933)\", \"Interleukin-35 Signalling (R-HSA-8984722)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"Activation of SMO (R-HSA-5635838)\", \"Disinhibition of SNARE formation (R-HSA-114516)\", \"Activated PKN1 stimulates transcription of AR (androgen receptor) regulated genes KLK2 and KLK3 (R-HSA-5625886)\", \"Glutamate and glutamine metabolism (R-HSA-8964539)\", \"Metabolism of non-coding RNA (R-HSA-194441)\", \"snRNP Assembly (R-HSA-191859)\", \"Fertilization (R-HSA-1187000)\", \"Sperm Motility And Taxes (R-HSA-1300642)\", \"Interleukin-23 signaling (R-HSA-9020933)\", \"Interleukin-35 Signalling (R-HSA-8984722)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-f8025bdc429d37df6c6022a540cc8f36.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}, {\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Chromosome Maintenance (R-HSA-73886)\", \"Extension of Telomeres (R-HSA-180786)\", \"Processive synthesis on the C-strand of the telomere (R-HSA-174414)\", \"Removal of the Flap Intermediate from the C-strand (R-HSA-174437)\", \"Resolution of D-Loop Structures (R-HSA-5693537)\", \"Resolution of D-loop Structures through Holliday Junction Intermediates (R-HSA-5693568)\", \"Resolution of D-loop Structures through Synthesis-Dependent Strand Annealing (SDSA) (R-HSA-5693554)\", \"Telomere C-strand (Lagging Strand) Synthesis (R-HSA-174417)\", \"Telomere Maintenance (R-HSA-157579)\", \"Aberrant regulation of mitotic G1/S transition in cancer due to RB1 defects (R-HSA-9659787)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"Chromosome Maintenance (R-HSA-73886)\", \"Extension of Telomeres (R-HSA-180786)\", \"Processive synthesis on the C-strand of the telomere (R-HSA-174414)\", \"Removal of the Flap Intermediate from the C-strand (R-HSA-174437)\", \"Resolution of D-Loop Structures (R-HSA-5693537)\", \"Resolution of D-loop Structures through Holliday Junction Intermediates (R-HSA-5693568)\", \"Resolution of D-loop Structures through Synthesis-Dependent Strand Annealing (SDSA) (R-HSA-5693554)\", \"Telomere C-strand (Lagging Strand) Synthesis (R-HSA-174417)\", \"Telomere Maintenance (R-HSA-157579)\", \"Aberrant regulation of mitotic G1/S transition in cancer due to RB1 defects (R-HSA-9659787)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-47a9c3951ab9c7587518ff6327e406f6.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}, {\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Diseases associated with the TLR signaling cascade (R-HSA-5602358)\", \"Diseases of Immune System (R-HSA-5260271)\", \"TICAM1, RIP1-mediated IKK complex recruitment (R-HSA-168927)\", \"TICAM1,TRAF6-dependent induction of TAK1 complex (R-HSA-9014325)\", \"TLR3-mediated TICAM1-dependent programmed cell death (R-HSA-9013957)\", \"Deadenylation of mRNA (R-HSA-429947)\", \"TP53 regulates transcription of additional cell cycle genes whose exact role in the p53 pathway remain uncertain (R-HSA-6804115)\", \"CD22 mediated BCR regulation (R-HSA-5690714)\", \"Ca2+ pathway (R-HSA-4086398)\", \"RNF mutants show enhanced WNT signaling and proliferation (R-HSA-5340588)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"Diseases associated with the TLR signaling cascade (R-HSA-5602358)\", \"Diseases of Immune System (R-HSA-5260271)\", \"TICAM1, RIP1-mediated IKK complex recruitment (R-HSA-168927)\", \"TICAM1,TRAF6-dependent induction of TAK1 complex (R-HSA-9014325)\", \"TLR3-mediated TICAM1-dependent programmed cell death (R-HSA-9013957)\", \"Deadenylation of mRNA (R-HSA-429947)\", \"TP53 regulates transcription of additional cell cycle genes whose exact role in the p53 pathway remain uncertain (R-HSA-6804115)\", \"CD22 mediated BCR regulation (R-HSA-5690714)\", \"Ca2+ pathway (R-HSA-4086398)\", \"RNF mutants show enhanced WNT signaling and proliferation (R-HSA-5340588)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-3c6809113b2ea009eb264ed4140139e5.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}, {\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Assembly of collagen fibrils and other multimeric structures (R-HSA-2022090)\", \"Collagen degradation (R-HSA-1442490)\", \"Trafficking and processing of endosomal TLR (R-HSA-1679131)\", \"IRE1alpha activates chaperones (R-HSA-381070)\", \"Signaling by Hippo (R-HSA-2028269)\", \"XBP1(S) activates chaperone genes (R-HSA-381038)\", \"ADORA2B mediated anti-inflammatory cytokines production (R-HSA-9660821)\", \"ADP signalling through P2Y purinoceptor 1 (R-HSA-418592)\", \"AKT phosphorylates targets in the cytosol (R-HSA-198323)\", \"AMER1 mutants destabilize the destruction complex (R-HSA-4839748)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"Assembly of collagen fibrils and other multimeric structures (R-HSA-2022090)\", \"Collagen degradation (R-HSA-1442490)\", \"Trafficking and processing of endosomal TLR (R-HSA-1679131)\", \"IRE1alpha activates chaperones (R-HSA-381070)\", \"Signaling by Hippo (R-HSA-2028269)\", \"XBP1(S) activates chaperone genes (R-HSA-381038)\", \"ADORA2B mediated anti-inflammatory cytokines production (R-HSA-9660821)\", \"ADP signalling through P2Y purinoceptor 1 (R-HSA-418592)\", \"AKT phosphorylates targets in the cytosol (R-HSA-198323)\", \"AMER1 mutants destabilize the destruction complex (R-HSA-4839748)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-4409b0ce18865b2a2a0f195c7fcfd1a5.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}], \"title\": \"Reactome data: without_loss group\", \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.8.1.json\"}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.VConcatChart(...)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "group_chart_by_cancer(\"without_loss\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-e95ddfb22cc44bd2baf4730f9d40e712\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-e95ddfb22cc44bd2baf4730f9d40e712\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-e95ddfb22cc44bd2baf4730f9d40e712\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.8.1?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function loadScript(lib) {\n",
       "      return new Promise(function(resolve, reject) {\n",
       "        var s = document.createElement('script');\n",
       "        s.src = paths[lib];\n",
       "        s.async = true;\n",
       "        s.onload = () => resolve(paths[lib]);\n",
       "        s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "        document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "      });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else if (typeof vegaEmbed === \"function\") {\n",
       "      displayChart(vegaEmbed);\n",
       "    } else {\n",
       "      loadScript(\"vega\")\n",
       "        .then(() => loadScript(\"vega-lite\"))\n",
       "        .then(() => loadScript(\"vega-embed\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}, \"title\": {\"anchor\": \"middle\"}}, \"vconcat\": [{\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Diseases of carbohydrate metabolism (R-HSA-5663084)\", \"Glycogen storage diseases (R-HSA-3229121)\", \"Glycogen synthesis (R-HSA-3322077)\", \"Myoclonic epilepsy of Lafora (R-HSA-3785653)\", \"Infection with Mycobacterium tuberculosis (R-HSA-9635486)\", \"Nef Mediated CD4 Down-regulation (R-HSA-167590)\", \"Nef Mediated CD8 Down-regulation (R-HSA-182218)\", \"Nef-mediates down modulation of cell surface receptors by recruiting them to clathrin adapters (R-HSA-164938)\", \"Response of Mtb to phagocytosis (R-HSA-9637690)\", \"Suppression of phagosomal maturation (R-HSA-9637687)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"Diseases of carbohydrate metabolism (R-HSA-5663084)\", \"Glycogen storage diseases (R-HSA-3229121)\", \"Glycogen synthesis (R-HSA-3322077)\", \"Myoclonic epilepsy of Lafora (R-HSA-3785653)\", \"Infection with Mycobacterium tuberculosis (R-HSA-9635486)\", \"Nef Mediated CD4 Down-regulation (R-HSA-167590)\", \"Nef Mediated CD8 Down-regulation (R-HSA-182218)\", \"Nef-mediates down modulation of cell surface receptors by recruiting them to clathrin adapters (R-HSA-164938)\", \"Response of Mtb to phagocytosis (R-HSA-9637690)\", \"Suppression of phagosomal maturation (R-HSA-9637687)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-709d44742fb5116d5f10b63b75af631e.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}, {\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Creatine metabolism (R-HSA-71288)\", \"Na+/Cl- dependent neurotransmitter transporters (R-HSA-442660)\", \"DARPP-32 events (R-HSA-180024)\", \"Synthesis of PE (R-HSA-1483213)\", \"Butyrophilin (BTN) family interactions (R-HSA-8851680)\", \"Signaling by Retinoic Acid (R-HSA-5362517)\", \"Triglyceride catabolism (R-HSA-163560)\", \"Triglyceride metabolism (R-HSA-8979227)\", \"Base-Excision Repair, AP Site Formation (R-HSA-73929)\", \"Cleavage of the damaged purine (R-HSA-110331)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"Creatine metabolism (R-HSA-71288)\", \"Na+/Cl- dependent neurotransmitter transporters (R-HSA-442660)\", \"DARPP-32 events (R-HSA-180024)\", \"Synthesis of PE (R-HSA-1483213)\", \"Butyrophilin (BTN) family interactions (R-HSA-8851680)\", \"Signaling by Retinoic Acid (R-HSA-5362517)\", \"Triglyceride catabolism (R-HSA-163560)\", \"Triglyceride metabolism (R-HSA-8979227)\", \"Base-Excision Repair, AP Site Formation (R-HSA-73929)\", \"Cleavage of the damaged purine (R-HSA-110331)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-d0ebaf7da577a191300bceba44989a15.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}, {\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Cyclin A/B1/B2 associated events during G2/M transition (R-HSA-69273)\", \"Interconversion of nucleotide di- and triphosphates (R-HSA-499943)\", \"TP53 Regulates Metabolic Genes (R-HSA-5628897)\", \"HIV Transcription Initiation (R-HSA-167161)\", \"RNA Polymerase II HIV Promoter Escape (R-HSA-167162)\", \"RNA Polymerase II Pre-transcription Events (R-HSA-674695)\", \"RNA Polymerase II Promoter Escape (R-HSA-73776)\", \"RNA Polymerase II Transcription Initiation (R-HSA-75953)\", \"RNA Polymerase II Transcription Initiation And Promoter Clearance (R-HSA-76042)\", \"RNA Polymerase II Transcription Pre-Initiation And Promoter Opening (R-HSA-73779)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"Cyclin A/B1/B2 associated events during G2/M transition (R-HSA-69273)\", \"Interconversion of nucleotide di- and triphosphates (R-HSA-499943)\", \"TP53 Regulates Metabolic Genes (R-HSA-5628897)\", \"HIV Transcription Initiation (R-HSA-167161)\", \"RNA Polymerase II HIV Promoter Escape (R-HSA-167162)\", \"RNA Polymerase II Pre-transcription Events (R-HSA-674695)\", \"RNA Polymerase II Promoter Escape (R-HSA-73776)\", \"RNA Polymerase II Transcription Initiation (R-HSA-75953)\", \"RNA Polymerase II Transcription Initiation And Promoter Clearance (R-HSA-76042)\", \"RNA Polymerase II Transcription Pre-Initiation And Promoter Opening (R-HSA-73779)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-adccd2336f7b50cdd709f8214cb1b21b.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}, {\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Cyclin A/B1/B2 associated events during G2/M transition (R-HSA-69273)\", \"Cyclin D associated events in G1 (R-HSA-69231)\", \"G1 Phase (R-HSA-69236)\", \"Mitotic G1 phase and G1/S transition (R-HSA-453279)\", \"Nonsense Mediated Decay (NMD) enhanced by the Exon Junction Complex (EJC) (R-HSA-975957)\", \"Nonsense-Mediated Decay (NMD) (R-HSA-927802)\", \"MicroRNA (miRNA) biogenesis (R-HSA-203927)\", \"Small interfering RNA (siRNA) biogenesis (R-HSA-426486)\", \"MECP2 regulates transcription factors (R-HSA-9022707)\", \"Transcriptional Regulation by MECP2 (R-HSA-8986944)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"Cyclin A/B1/B2 associated events during G2/M transition (R-HSA-69273)\", \"Cyclin D associated events in G1 (R-HSA-69231)\", \"G1 Phase (R-HSA-69236)\", \"Mitotic G1 phase and G1/S transition (R-HSA-453279)\", \"Nonsense Mediated Decay (NMD) enhanced by the Exon Junction Complex (EJC) (R-HSA-975957)\", \"Nonsense-Mediated Decay (NMD) (R-HSA-927802)\", \"MicroRNA (miRNA) biogenesis (R-HSA-203927)\", \"Small interfering RNA (siRNA) biogenesis (R-HSA-426486)\", \"MECP2 regulates transcription factors (R-HSA-9022707)\", \"Transcriptional Regulation by MECP2 (R-HSA-8986944)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-56cb0a1d6410a69ffcdb74a1b6ecf6ac.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}, {\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"ADORA2B mediated anti-inflammatory cytokines production (R-HSA-9660821)\", \"ADP signalling through P2Y purinoceptor 1 (R-HSA-418592)\", \"ADP signalling through P2Y purinoceptor 12 (R-HSA-392170)\", \"Aberrant regulation of mitotic G1/S transition in cancer due to RB1 defects (R-HSA-9659787)\", \"Aberrant regulation of mitotic cell cycle due to RB1 defects (R-HSA-9687139)\", \"Activation of G protein gated Potassium channels (R-HSA-1296041)\", \"Activation of GABAB receptors (R-HSA-991365)\", \"Activation of the pre-replicative complex (R-HSA-68962)\", \"Adrenaline,noradrenaline inhibits insulin secretion (R-HSA-400042)\", \"Aquaporin-mediated transport (R-HSA-445717)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"ADORA2B mediated anti-inflammatory cytokines production (R-HSA-9660821)\", \"ADP signalling through P2Y purinoceptor 1 (R-HSA-418592)\", \"ADP signalling through P2Y purinoceptor 12 (R-HSA-392170)\", \"Aberrant regulation of mitotic G1/S transition in cancer due to RB1 defects (R-HSA-9659787)\", \"Aberrant regulation of mitotic cell cycle due to RB1 defects (R-HSA-9687139)\", \"Activation of G protein gated Potassium channels (R-HSA-1296041)\", \"Activation of GABAB receptors (R-HSA-991365)\", \"Activation of the pre-replicative complex (R-HSA-68962)\", \"Adrenaline,noradrenaline inhibits insulin secretion (R-HSA-400042)\", \"Aquaporin-mediated transport (R-HSA-445717)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-fab6993e37770df898539710313f6e14.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}], \"title\": \"Reactome data: with_both group\", \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.8.1.json\"}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.VConcatChart(...)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "group_chart_by_cancer(\"with_both\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-62d7e179102c4c86b4160799fab85a5b\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-62d7e179102c4c86b4160799fab85a5b\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-62d7e179102c4c86b4160799fab85a5b\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.8.1?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function loadScript(lib) {\n",
       "      return new Promise(function(resolve, reject) {\n",
       "        var s = document.createElement('script');\n",
       "        s.src = paths[lib];\n",
       "        s.async = true;\n",
       "        s.onload = () => resolve(paths[lib]);\n",
       "        s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "        document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "      });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else if (typeof vegaEmbed === \"function\") {\n",
       "      displayChart(vegaEmbed);\n",
       "    } else {\n",
       "      loadScript(\"vega\")\n",
       "        .then(() => loadScript(\"vega-lite\"))\n",
       "        .then(() => loadScript(\"vega-embed\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}, \"title\": {\"anchor\": \"middle\"}}, \"vconcat\": [{\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"RORA activates gene expression (R-HSA-1368082)\", \"Disinhibition of SNARE formation (R-HSA-114516)\", \"Multifunctional anion exchangers (R-HSA-427601)\", \"Trafficking and processing of endosomal TLR (R-HSA-1679131)\", \"Histidine catabolism (R-HSA-70921)\", \"The activation of arylsulfatases (R-HSA-1663150)\", \"TYSND1 cleaves peroxisomal proteins (R-HSA-9033500)\", \"Assembly of Viral Components at the Budding Site (R-HSA-168316)\", \"Virus Assembly and Release (R-HSA-168268)\", \"Carboxyterminal post-translational modifications of tubulin (R-HSA-8955332)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"RORA activates gene expression (R-HSA-1368082)\", \"Disinhibition of SNARE formation (R-HSA-114516)\", \"Multifunctional anion exchangers (R-HSA-427601)\", \"Trafficking and processing of endosomal TLR (R-HSA-1679131)\", \"Histidine catabolism (R-HSA-70921)\", \"The activation of arylsulfatases (R-HSA-1663150)\", \"TYSND1 cleaves peroxisomal proteins (R-HSA-9033500)\", \"Assembly of Viral Components at the Budding Site (R-HSA-168316)\", \"Virus Assembly and Release (R-HSA-168268)\", \"Carboxyterminal post-translational modifications of tubulin (R-HSA-8955332)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-d0ea68683b41af535c38c3eae99bf296.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}, {\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Cation-coupled Chloride cotransporters (R-HSA-426117)\", \"Defective TPR may confer susceptibility towards thyroid papillary carcinoma (TPC) (R-HSA-5619107)\", \"Export of Viral Ribonucleoproteins from Nucleus (R-HSA-168274)\", \"Interactions of Rev with host cellular proteins (R-HSA-177243)\", \"Interactions of Vpr with host cellular proteins (R-HSA-176033)\", \"NEP/NS2 Interacts with the Cellular Export Machinery (R-HSA-168333)\", \"Nuclear import of Rev protein (R-HSA-180746)\", \"Regulation of Glucokinase by Glucokinase Regulatory Protein (R-HSA-170822)\", \"Rev-mediated nuclear export of HIV RNA (R-HSA-165054)\", \"Transport of Mature mRNA Derived from an Intronless Transcript (R-HSA-159231)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"Cation-coupled Chloride cotransporters (R-HSA-426117)\", \"Defective TPR may confer susceptibility towards thyroid papillary carcinoma (TPC) (R-HSA-5619107)\", \"Export of Viral Ribonucleoproteins from Nucleus (R-HSA-168274)\", \"Interactions of Rev with host cellular proteins (R-HSA-177243)\", \"Interactions of Vpr with host cellular proteins (R-HSA-176033)\", \"NEP/NS2 Interacts with the Cellular Export Machinery (R-HSA-168333)\", \"Nuclear import of Rev protein (R-HSA-180746)\", \"Regulation of Glucokinase by Glucokinase Regulatory Protein (R-HSA-170822)\", \"Rev-mediated nuclear export of HIV RNA (R-HSA-165054)\", \"Transport of Mature mRNA Derived from an Intronless Transcript (R-HSA-159231)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-e8aab02f43ac25ead0cd3d565b7e18ed.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}, {\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Cellular hexose transport (R-HSA-189200)\", \"Vitamin B5 (pantothenate) metabolism (R-HSA-199220)\", \"Inactivation, recovery and regulation of the phototransduction cascade (R-HSA-2514859)\", \"The phototransduction cascade (R-HSA-2514856)\", \"PIWI-interacting RNA (piRNA) biogenesis (R-HSA-5601884)\", \"Transcriptional regulation by small RNAs (R-HSA-5578749)\", \"Viral Messenger RNA Synthesis (R-HSA-168325)\", \"Synthesis of epoxy (EET) and dihydroxyeicosatrienoic acids (DHET) (R-HSA-2142670)\", \"Abortive elongation of HIV-1 transcript in the absence of Tat (R-HSA-167242)\", \"RNA polymerase II transcribes snRNA genes (R-HSA-6807505)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"Cellular hexose transport (R-HSA-189200)\", \"Vitamin B5 (pantothenate) metabolism (R-HSA-199220)\", \"Inactivation, recovery and regulation of the phototransduction cascade (R-HSA-2514859)\", \"The phototransduction cascade (R-HSA-2514856)\", \"PIWI-interacting RNA (piRNA) biogenesis (R-HSA-5601884)\", \"Transcriptional regulation by small RNAs (R-HSA-5578749)\", \"Viral Messenger RNA Synthesis (R-HSA-168325)\", \"Synthesis of epoxy (EET) and dihydroxyeicosatrienoic acids (DHET) (R-HSA-2142670)\", \"Abortive elongation of HIV-1 transcript in the absence of Tat (R-HSA-167242)\", \"RNA polymerase II transcribes snRNA genes (R-HSA-6807505)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-31716a4913caed7ff5b68e05f6117b80.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}, {\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Maturation of nucleoprotein (R-HSA-9683610)\", \"Negative regulation of activity of TFAP2 (AP-2) family transcription factors (R-HSA-8866904)\", \"SUMOylation of DNA methylation proteins (R-HSA-4655427)\", \"Translation of Replicase and Assembly of the Replication Transcription Complex (R-HSA-9679504)\", \"DEx/H-box helicases activate type I IFN and inflammatory cytokines production (R-HSA-3134963)\", \"Interleukin-1 processing (R-HSA-448706)\", \"HSP90 chaperone cycle for steroid hormone receptors (SHR) (R-HSA-3371497)\", \"Viral RNP Complexes in the Host Cell Nucleus (R-HSA-168330)\", \"CHL1 interactions (R-HSA-447041)\", \"CREB1 phosphorylation through the activation of Adenylate Cyclase (R-HSA-442720)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"Maturation of nucleoprotein (R-HSA-9683610)\", \"Negative regulation of activity of TFAP2 (AP-2) family transcription factors (R-HSA-8866904)\", \"SUMOylation of DNA methylation proteins (R-HSA-4655427)\", \"Translation of Replicase and Assembly of the Replication Transcription Complex (R-HSA-9679504)\", \"DEx/H-box helicases activate type I IFN and inflammatory cytokines production (R-HSA-3134963)\", \"Interleukin-1 processing (R-HSA-448706)\", \"HSP90 chaperone cycle for steroid hormone receptors (SHR) (R-HSA-3371497)\", \"Viral RNP Complexes in the Host Cell Nucleus (R-HSA-168330)\", \"CHL1 interactions (R-HSA-447041)\", \"CREB1 phosphorylation through the activation of Adenylate Cyclase (R-HSA-442720)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-f5d1cc2a8779dc96849d5cfafbf1f531.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}, {\"layer\": [{\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": -30, \"labelFontSize\": 12, \"labelLimit\": 500, \"title\": \"\", \"titleFontSize\": 16}, \"field\": \"pathway_name\", \"sort\": [\"Diseases associated with surfactant metabolism (R-HSA-5687613)\", \"Sodium-coupled phosphate cotransporters (R-HSA-427652)\", \"Assembly of Viral Components at the Budding Site (R-HSA-168316)\", \"Interleukin-27 signaling (R-HSA-9020956)\", \"Interleukin-35 Signalling (R-HSA-8984722)\", \"Maturation of spike protein (R-HSA-9683686)\", \"Virus Assembly and Release (R-HSA-168268)\", \"ABC transporters in lipid homeostasis (R-HSA-1369062)\", \"Assembly of the ORC complex at the origin of replication (R-HSA-68616)\", \"CDC6 association with the ORC:origin complex (R-HSA-68689)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}, {\"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"quantitative\", \"field\": \"mean_expr\", \"legend\": {\"title\": \"Tumor expression\"}, \"scale\": {\"domain\": [-1, 1], \"scheme\": \"blueorange\"}}, \"size\": {\"type\": \"quantitative\", \"field\": \"rank_size\", \"legend\": null}, \"x\": {\"type\": \"nominal\", \"axis\": {\"labelAngle\": 0, \"labelFontSize\": 12, \"labelLimit\": 500, \"labelPadding\": 15, \"title\": \"Pathway rank\", \"titleFontSize\": 10}, \"field\": \"rank_label\", \"sort\": [\"Diseases associated with surfactant metabolism (R-HSA-5687613)\", \"Sodium-coupled phosphate cotransporters (R-HSA-427652)\", \"Assembly of Viral Components at the Budding Site (R-HSA-168316)\", \"Interleukin-27 signaling (R-HSA-9020956)\", \"Interleukin-35 Signalling (R-HSA-8984722)\", \"Maturation of spike protein (R-HSA-9683686)\", \"Virus Assembly and Release (R-HSA-168268)\", \"ABC transporters in lipid homeostasis (R-HSA-1369062)\", \"Assembly of the ORC complex at the origin of replication (R-HSA-68616)\", \"CDC6 association with the ORC:origin complex (R-HSA-68689)\"]}, \"y\": {\"type\": \"nominal\", \"axis\": {\"title\": null}, \"field\": \"cancer_type\"}}}], \"data\": {\"url\": \"altair_data/altair-data-601ad560dec804f29ca8644bf08870af.json\", \"format\": {\"type\": \"json\"}}, \"height\": 30, \"resolve\": {\"scale\": {\"x\": \"independent\"}}, \"width\": 400}], \"title\": \"Reactome data: without_both group\", \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.8.1.json\"}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.VConcatChart(...)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "group_chart_by_cancer(\"without_both\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
