{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import cptac\n",
    "import pyensembl\n",
    "ensembl = pyensembl.EnsemblRelease()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                \r"
     ]
    },
    {
     "ename": "NoInternetError",
     "evalue": "Insufficient internet. Check your internet connection.",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNoInternetError\u001b[0m                           Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-2-ca7065a67193>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mcptac\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdownload\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdataset\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m\"endometrial\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      2\u001b[0m \u001b[0mcptac\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdownload\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdataset\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m\"brca\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[0mcptac\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdownload\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdataset\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m\"ccrcc\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[0mcptac\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdownload\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdataset\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m\"gbm\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[0mcptac\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdownload\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdataset\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m\"hnscc\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\cptac\\file_download.py\u001b[0m in \u001b[0;36mdownload\u001b[1;34m(dataset, version, redownload)\u001b[0m\n\u001b[0;32m     54\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     55\u001b[0m     \u001b[1;31m# Update the index\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 56\u001b[1;33m     \u001b[0mupdate_index\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdataset\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     57\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     58\u001b[0m     \u001b[1;31m# Load the index\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\cptac\\file_download.py\u001b[0m in \u001b[0;36mupdate_index\u001b[1;34m(dataset)\u001b[0m\n\u001b[0;32m    146\u001b[0m     \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mchecking_msg\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mend\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'\\r'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    147\u001b[0m     \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 148\u001b[1;33m         \u001b[0mserver_index_hash\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mdownload_text\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mindex_hash_url\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    149\u001b[0m     \u001b[1;32mfinally\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    150\u001b[0m         \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\" \"\u001b[0m \u001b[1;33m*\u001b[0m \u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mchecking_msg\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mend\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'\\r'\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;31m# Erase the checking message, even if there was an internet error\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\cptac\\file_download.py\u001b[0m in \u001b[0;36mdownload_text\u001b[1;34m(url)\u001b[0m\n\u001b[0;32m    180\u001b[0m         \u001b[0mresponse\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mraise_for_status\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;31m# Raises a requests HTTPError if the response code was unsuccessful\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    181\u001b[0m     \u001b[1;32mexcept\u001b[0m \u001b[0mrequests\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mRequestException\u001b[0m\u001b[1;33m:\u001b[0m \u001b[1;31m# Parent class for all exceptions in the requests module\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 182\u001b[1;33m         \u001b[1;32mraise\u001b[0m \u001b[0mNoInternetError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"Insufficient internet. Check your internet connection.\"\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    183\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    184\u001b[0m     \u001b[0mtext\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mresponse\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtext\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mstrip\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNoInternetError\u001b[0m: Insufficient internet. Check your internet connection."
     ]
    }
   ],
   "source": [
    "cptac.download(dataset=\"endometrial\")\n",
    "cptac.download(dataset=\"brca\")\n",
    "cptac.download(dataset=\"ccrcc\")\n",
    "cptac.download(dataset=\"gbm\")\n",
    "cptac.download(dataset=\"hnscc\")\n",
    "cptac.download(dataset=\"luad\")\n",
    "cptac.download(dataset=\"ovarian\")\n",
    "cptac.download(dataset=\"colon\")\n",
    "cptac.download(dataset=\"lscc\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Formatting dataframes...                        \r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking that hnscc index is up-to-date...\r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cptac warning: The GBM dataset is under publication embargo until March 01, 2021. CPTAC is a community resource project and data are made available rapidly after generation for community research use. The embargo allows exploring and utilizing the data, but analysis may not be published until after the embargo date. Please see https://proteomics.cancer.gov/data-portal/about/data-use-agreement or enter cptac.embargo() to open the webpage for more details. (C:\\Users\\chels\\anaconda3\\lib\\site-packages\\ipykernel_launcher.py, line 4)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking that luad index is up-to-date... \r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cptac warning: The HNSCC data is currently strictly reserved for CPTAC investigators. Otherwise, you are not authorized to access these data. Additionally, even after these data become publicly available, they will be subject to a publication embargo (see https://proteomics.cancer.gov/data-portal/about/data-use-agreement or enter cptac.embargo() to open the webpage for more details). (C:\\Users\\chels\\anaconda3\\lib\\site-packages\\ipykernel_launcher.py, line 5)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "version 3scc v3.2.......                    \n",
      "                            \r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cptac warning: The LSCC data is currently strictly reserved for CPTAC investigators. Otherwise, you are not authorized to access these data. Additionally, even after these data become publicly available, they will be subject to a publication embargo (see https://proteomics.cancer.gov/data-portal/about/data-use-agreement or enter cptac.embargo() to open the webpage for more details). (C:\\Users\\chels\\anaconda3\\lib\\site-packages\\ipykernel_launcher.py, line 9)\n"
     ]
    }
   ],
   "source": [
    "en = cptac.Endometrial()\n",
    "brca = cptac.Brca()\n",
    "ccrcc = cptac.Ccrcc()\n",
    "gbm = cptac.Gbm()\n",
    "hnscc = cptac.Hnscc()\n",
    "luad = cptac.Luad()\n",
    "ovarian = cptac.Ovarian()\n",
    "colon = cptac.Colon()\n",
    "lscc = cptac.Lscc()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "en_cnv = en.get_CNV()\n",
    "brca_cnv = brca.get_CNV()\n",
    "ccrcc_cnv = ccrcc.get_CNV()\n",
    "gbm_cnv = gbm.get_CNV()\n",
    "hnscc_cnv = hnscc.get_CNV()\n",
    "luad_cnv = luad.get_CNV()\n",
    "ovarian_cnv = ovarian.get_CNV()\n",
    "colon_cnv = colon.get_CNV()\n",
    "lscc_cnv = lscc.get_CNV()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Database_ID</th>\n",
       "      <th>ENSG00000121410.11</th>\n",
       "      <th>ENSG00000148584.14</th>\n",
       "      <th>ENSG00000175899.14</th>\n",
       "      <th>ENSG00000166535.19</th>\n",
       "      <th>ENSG00000184389.9</th>\n",
       "      <th>ENSG00000128274.15</th>\n",
       "      <th>ENSG00000118017.3</th>\n",
       "      <th>ENSG00000094914.12</th>\n",
       "      <th>ENSG00000081760.16</th>\n",
       "      <th>ENSG00000114771.13</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000174442.11</th>\n",
       "      <th>ENSG00000122952.16</th>\n",
       "      <th>ENSG00000198205.6</th>\n",
       "      <th>ENSG00000198455.4</th>\n",
       "      <th>ENSG00000070476.14</th>\n",
       "      <th>ENSG00000203995.9</th>\n",
       "      <th>ENSG00000162378.12</th>\n",
       "      <th>ENSG00000159840.15</th>\n",
       "      <th>ENSG00000074755.14</th>\n",
       "      <th>ENSG00000036549.12</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C3L-00004</th>\n",
       "      <td>0.020366</td>\n",
       "      <td>0.013001</td>\n",
       "      <td>0.015820</td>\n",
       "      <td>0.015820</td>\n",
       "      <td>0.013783</td>\n",
       "      <td>0.016040</td>\n",
       "      <td>0.012262</td>\n",
       "      <td>0.015433</td>\n",
       "      <td>0.015433</td>\n",
       "      <td>0.012262</td>\n",
       "      <td>...</td>\n",
       "      <td>0.016381</td>\n",
       "      <td>0.013001</td>\n",
       "      <td>0.013331</td>\n",
       "      <td>0.013331</td>\n",
       "      <td>0.012262</td>\n",
       "      <td>0.013783</td>\n",
       "      <td>0.013783</td>\n",
       "      <td>0.022275</td>\n",
       "      <td>0.020393</td>\n",
       "      <td>0.013783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00010</th>\n",
       "      <td>0.027739</td>\n",
       "      <td>0.019186</td>\n",
       "      <td>0.019790</td>\n",
       "      <td>0.019790</td>\n",
       "      <td>0.021678</td>\n",
       "      <td>0.022887</td>\n",
       "      <td>0.019669</td>\n",
       "      <td>0.022677</td>\n",
       "      <td>0.022677</td>\n",
       "      <td>0.019669</td>\n",
       "      <td>...</td>\n",
       "      <td>0.022946</td>\n",
       "      <td>0.019186</td>\n",
       "      <td>0.011710</td>\n",
       "      <td>0.011710</td>\n",
       "      <td>0.019669</td>\n",
       "      <td>0.021678</td>\n",
       "      <td>0.021678</td>\n",
       "      <td>0.021366</td>\n",
       "      <td>0.025461</td>\n",
       "      <td>0.021678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00011</th>\n",
       "      <td>0.089568</td>\n",
       "      <td>0.121480</td>\n",
       "      <td>0.134582</td>\n",
       "      <td>0.134582</td>\n",
       "      <td>0.104257</td>\n",
       "      <td>0.066205</td>\n",
       "      <td>0.125552</td>\n",
       "      <td>0.126908</td>\n",
       "      <td>0.091427</td>\n",
       "      <td>0.125552</td>\n",
       "      <td>...</td>\n",
       "      <td>0.121135</td>\n",
       "      <td>0.121480</td>\n",
       "      <td>-0.470122</td>\n",
       "      <td>-0.470122</td>\n",
       "      <td>0.125552</td>\n",
       "      <td>0.104257</td>\n",
       "      <td>0.104257</td>\n",
       "      <td>0.116052</td>\n",
       "      <td>-0.488441</td>\n",
       "      <td>0.131674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00026</th>\n",
       "      <td>-0.021259</td>\n",
       "      <td>-0.021430</td>\n",
       "      <td>-0.015553</td>\n",
       "      <td>-0.015553</td>\n",
       "      <td>-0.020833</td>\n",
       "      <td>-0.116570</td>\n",
       "      <td>-0.024959</td>\n",
       "      <td>-0.018876</td>\n",
       "      <td>-0.018876</td>\n",
       "      <td>-0.024959</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.021340</td>\n",
       "      <td>-0.021430</td>\n",
       "      <td>-0.550917</td>\n",
       "      <td>-0.550917</td>\n",
       "      <td>-0.026006</td>\n",
       "      <td>-0.020833</td>\n",
       "      <td>-0.020833</td>\n",
       "      <td>0.394042</td>\n",
       "      <td>-0.017815</td>\n",
       "      <td>-0.020833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00079</th>\n",
       "      <td>0.061714</td>\n",
       "      <td>0.123367</td>\n",
       "      <td>-0.048778</td>\n",
       "      <td>-0.048778</td>\n",
       "      <td>0.048854</td>\n",
       "      <td>-0.178020</td>\n",
       "      <td>0.160586</td>\n",
       "      <td>-0.062128</td>\n",
       "      <td>-0.062128</td>\n",
       "      <td>0.160586</td>\n",
       "      <td>...</td>\n",
       "      <td>0.041436</td>\n",
       "      <td>0.123367</td>\n",
       "      <td>0.086172</td>\n",
       "      <td>0.086172</td>\n",
       "      <td>0.160586</td>\n",
       "      <td>0.048854</td>\n",
       "      <td>0.048854</td>\n",
       "      <td>0.049045</td>\n",
       "      <td>-0.033958</td>\n",
       "      <td>0.039913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N-01646</th>\n",
       "      <td>0.026528</td>\n",
       "      <td>0.065469</td>\n",
       "      <td>0.074286</td>\n",
       "      <td>0.074286</td>\n",
       "      <td>0.031000</td>\n",
       "      <td>0.031598</td>\n",
       "      <td>0.288325</td>\n",
       "      <td>0.062137</td>\n",
       "      <td>0.063978</td>\n",
       "      <td>0.288325</td>\n",
       "      <td>...</td>\n",
       "      <td>0.033744</td>\n",
       "      <td>0.030586</td>\n",
       "      <td>0.027502</td>\n",
       "      <td>0.027502</td>\n",
       "      <td>0.288325</td>\n",
       "      <td>0.031000</td>\n",
       "      <td>0.031000</td>\n",
       "      <td>0.033042</td>\n",
       "      <td>0.024613</td>\n",
       "      <td>0.031000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N-01648</th>\n",
       "      <td>0.003589</td>\n",
       "      <td>-0.017667</td>\n",
       "      <td>-0.009598</td>\n",
       "      <td>-0.009598</td>\n",
       "      <td>-0.003569</td>\n",
       "      <td>-0.000926</td>\n",
       "      <td>-0.009770</td>\n",
       "      <td>-0.004919</td>\n",
       "      <td>-0.004919</td>\n",
       "      <td>-0.009770</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.007744</td>\n",
       "      <td>-0.008758</td>\n",
       "      <td>-0.009086</td>\n",
       "      <td>-0.009086</td>\n",
       "      <td>-0.009770</td>\n",
       "      <td>-0.003569</td>\n",
       "      <td>-0.003569</td>\n",
       "      <td>-0.014371</td>\n",
       "      <td>0.003618</td>\n",
       "      <td>-0.003569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N-01649</th>\n",
       "      <td>0.017956</td>\n",
       "      <td>0.019480</td>\n",
       "      <td>0.026971</td>\n",
       "      <td>0.026971</td>\n",
       "      <td>0.017902</td>\n",
       "      <td>0.020458</td>\n",
       "      <td>0.020200</td>\n",
       "      <td>0.028800</td>\n",
       "      <td>0.028800</td>\n",
       "      <td>0.020200</td>\n",
       "      <td>...</td>\n",
       "      <td>0.017537</td>\n",
       "      <td>0.019480</td>\n",
       "      <td>0.020941</td>\n",
       "      <td>0.020941</td>\n",
       "      <td>0.020200</td>\n",
       "      <td>0.017902</td>\n",
       "      <td>0.017902</td>\n",
       "      <td>0.328963</td>\n",
       "      <td>0.021064</td>\n",
       "      <td>0.017902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N-01651</th>\n",
       "      <td>0.163116</td>\n",
       "      <td>-0.331378</td>\n",
       "      <td>0.140222</td>\n",
       "      <td>0.140222</td>\n",
       "      <td>-0.055388</td>\n",
       "      <td>-0.049226</td>\n",
       "      <td>0.134608</td>\n",
       "      <td>0.142253</td>\n",
       "      <td>0.142253</td>\n",
       "      <td>0.134608</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.315716</td>\n",
       "      <td>-0.321503</td>\n",
       "      <td>0.114801</td>\n",
       "      <td>0.114801</td>\n",
       "      <td>0.134608</td>\n",
       "      <td>-0.055388</td>\n",
       "      <td>-0.055388</td>\n",
       "      <td>0.145273</td>\n",
       "      <td>-0.051065</td>\n",
       "      <td>-0.078956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N-01808</th>\n",
       "      <td>-0.025767</td>\n",
       "      <td>-0.022834</td>\n",
       "      <td>-0.019618</td>\n",
       "      <td>-0.019618</td>\n",
       "      <td>-0.026103</td>\n",
       "      <td>-0.023638</td>\n",
       "      <td>0.382217</td>\n",
       "      <td>-0.019048</td>\n",
       "      <td>-0.019048</td>\n",
       "      <td>0.382012</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.024736</td>\n",
       "      <td>-0.022834</td>\n",
       "      <td>-0.027958</td>\n",
       "      <td>-0.027958</td>\n",
       "      <td>-0.593051</td>\n",
       "      <td>-0.026103</td>\n",
       "      <td>-0.026103</td>\n",
       "      <td>0.380146</td>\n",
       "      <td>-0.023639</td>\n",
       "      <td>-0.026103</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>110 rows × 19285 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name                      A1BG               A1CF                A2M  \\\n",
       "Database_ID ENSG00000121410.11 ENSG00000148584.14 ENSG00000175899.14   \n",
       "Patient_ID                                                             \n",
       "C3L-00004             0.020366           0.013001           0.015820   \n",
       "C3L-00010             0.027739           0.019186           0.019790   \n",
       "C3L-00011             0.089568           0.121480           0.134582   \n",
       "C3L-00026            -0.021259          -0.021430          -0.015553   \n",
       "C3L-00079             0.061714           0.123367          -0.048778   \n",
       "...                        ...                ...                ...   \n",
       "C3N-01646             0.026528           0.065469           0.074286   \n",
       "C3N-01648             0.003589          -0.017667          -0.009598   \n",
       "C3N-01649             0.017956           0.019480           0.026971   \n",
       "C3N-01651             0.163116          -0.331378           0.140222   \n",
       "C3N-01808            -0.025767          -0.022834          -0.019618   \n",
       "\n",
       "Name                     A2ML1           A3GALT2             A4GALT  \\\n",
       "Database_ID ENSG00000166535.19 ENSG00000184389.9 ENSG00000128274.15   \n",
       "Patient_ID                                                            \n",
       "C3L-00004             0.015820          0.013783           0.016040   \n",
       "C3L-00010             0.019790          0.021678           0.022887   \n",
       "C3L-00011             0.134582          0.104257           0.066205   \n",
       "C3L-00026            -0.015553         -0.020833          -0.116570   \n",
       "C3L-00079            -0.048778          0.048854          -0.178020   \n",
       "...                        ...               ...                ...   \n",
       "C3N-01646             0.074286          0.031000           0.031598   \n",
       "C3N-01648            -0.009598         -0.003569          -0.000926   \n",
       "C3N-01649             0.026971          0.017902           0.020458   \n",
       "C3N-01651             0.140222         -0.055388          -0.049226   \n",
       "C3N-01808            -0.019618         -0.026103          -0.023638   \n",
       "\n",
       "Name                    A4GNT               AAAS               AACS  \\\n",
       "Database_ID ENSG00000118017.3 ENSG00000094914.12 ENSG00000081760.16   \n",
       "Patient_ID                                                            \n",
       "C3L-00004            0.012262           0.015433           0.015433   \n",
       "C3L-00010            0.019669           0.022677           0.022677   \n",
       "C3L-00011            0.125552           0.126908           0.091427   \n",
       "C3L-00026           -0.024959          -0.018876          -0.018876   \n",
       "C3L-00079            0.160586          -0.062128          -0.062128   \n",
       "...                       ...                ...                ...   \n",
       "C3N-01646            0.288325           0.062137           0.063978   \n",
       "C3N-01648           -0.009770          -0.004919          -0.004919   \n",
       "C3N-01649            0.020200           0.028800           0.028800   \n",
       "C3N-01651            0.134608           0.142253           0.142253   \n",
       "C3N-01808            0.382217          -0.019048          -0.019048   \n",
       "\n",
       "Name                     AADAC  ...             ZWILCH              ZWINT  \\\n",
       "Database_ID ENSG00000114771.13  ... ENSG00000174442.11 ENSG00000122952.16   \n",
       "Patient_ID                      ...                                         \n",
       "C3L-00004             0.012262  ...           0.016381           0.013001   \n",
       "C3L-00010             0.019669  ...           0.022946           0.019186   \n",
       "C3L-00011             0.125552  ...           0.121135           0.121480   \n",
       "C3L-00026            -0.024959  ...          -0.021340          -0.021430   \n",
       "C3L-00079             0.160586  ...           0.041436           0.123367   \n",
       "...                        ...  ...                ...                ...   \n",
       "C3N-01646             0.288325  ...           0.033744           0.030586   \n",
       "C3N-01648            -0.009770  ...          -0.007744          -0.008758   \n",
       "C3N-01649             0.020200  ...           0.017537           0.019480   \n",
       "C3N-01651             0.134608  ...          -0.315716          -0.321503   \n",
       "C3N-01808             0.382012  ...          -0.024736          -0.022834   \n",
       "\n",
       "Name                     ZXDA              ZXDB               ZXDC  \\\n",
       "Database_ID ENSG00000198205.6 ENSG00000198455.4 ENSG00000070476.14   \n",
       "Patient_ID                                                           \n",
       "C3L-00004            0.013331          0.013331           0.012262   \n",
       "C3L-00010            0.011710          0.011710           0.019669   \n",
       "C3L-00011           -0.470122         -0.470122           0.125552   \n",
       "C3L-00026           -0.550917         -0.550917          -0.026006   \n",
       "C3L-00079            0.086172          0.086172           0.160586   \n",
       "...                       ...               ...                ...   \n",
       "C3N-01646            0.027502          0.027502           0.288325   \n",
       "C3N-01648           -0.009086         -0.009086          -0.009770   \n",
       "C3N-01649            0.020941          0.020941           0.020200   \n",
       "C3N-01651            0.114801          0.114801           0.134608   \n",
       "C3N-01808           -0.027958         -0.027958          -0.593051   \n",
       "\n",
       "Name                   ZYG11A             ZYG11B                ZYX  \\\n",
       "Database_ID ENSG00000203995.9 ENSG00000162378.12 ENSG00000159840.15   \n",
       "Patient_ID                                                            \n",
       "C3L-00004            0.013783           0.013783           0.022275   \n",
       "C3L-00010            0.021678           0.021678           0.021366   \n",
       "C3L-00011            0.104257           0.104257           0.116052   \n",
       "C3L-00026           -0.020833          -0.020833           0.394042   \n",
       "C3L-00079            0.048854           0.048854           0.049045   \n",
       "...                       ...                ...                ...   \n",
       "C3N-01646            0.031000           0.031000           0.033042   \n",
       "C3N-01648           -0.003569          -0.003569          -0.014371   \n",
       "C3N-01649            0.017902           0.017902           0.328963   \n",
       "C3N-01651           -0.055388          -0.055388           0.145273   \n",
       "C3N-01808           -0.026103          -0.026103           0.380146   \n",
       "\n",
       "Name                     ZZEF1               ZZZ3  \n",
       "Database_ID ENSG00000074755.14 ENSG00000036549.12  \n",
       "Patient_ID                                         \n",
       "C3L-00004             0.020393           0.013783  \n",
       "C3L-00010             0.025461           0.021678  \n",
       "C3L-00011            -0.488441           0.131674  \n",
       "C3L-00026            -0.017815          -0.020833  \n",
       "C3L-00079            -0.033958           0.039913  \n",
       "...                        ...                ...  \n",
       "C3N-01646             0.024613           0.031000  \n",
       "C3N-01648             0.003618          -0.003569  \n",
       "C3N-01649             0.021064           0.017902  \n",
       "C3N-01651            -0.051065          -0.078956  \n",
       "C3N-01808            -0.023639          -0.026103  \n",
       "\n",
       "[110 rows x 19285 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccrcc_cnv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def separate(gain, loss, data):\n",
    "  \"\"\"\n",
    "  Counts the Number of patients with high and low values for each gene.\n",
    "  \"\"\"\n",
    "  sepDict = dict()\n",
    "  for column in data:\n",
    "    numGain = 0\n",
    "    numLoss = 0\n",
    "    numNeu = 0\n",
    "    for i in data[column]:\n",
    "        if i >= gain:\n",
    "            numGain += 1\n",
    "        elif i <= loss:\n",
    "            numLoss += 1\n",
    "        else:\n",
    "            numNeu += 1 \n",
    "    sepDict[column] = [numGain, numNeu, numLoss]\n",
    "  return sepDict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_counts_and_average(data):\n",
    "  \"\"\"\n",
    "  Takes a CNV dataframe and generates a dataframe with the counts of gains, losses and neutral values\n",
    "  \"\"\"\n",
    "  df = pd.DataFrame(data = separate(.2,-.2,data), index = [\"gain\", \"neutral\", \"loss\"])\n",
    "  return df.transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_chromo_and_loc(my_dict):\n",
    "    chromo = list()\n",
    "    locations = list()\n",
    "    not_found = list()\n",
    "    count = 0\n",
    "    for gene in list(my_dict.index.get_level_values(0)):\n",
    "        count += 1\n",
    "        try:\n",
    "            e_gene = ensembl.genes_by_name(gene)\n",
    "            chromo.append(int(e_gene[0].contig))\n",
    "            locations.append((e_gene[0].start + e_gene[0].end)/2)\n",
    "        except:\n",
    "            not_found.append(gene)\n",
    "            chromo.append(None)\n",
    "            locations.append(None)\n",
    "    my_dict['chromo'] = chromo\n",
    "    my_dict['location'] = locations\n",
    "    print(len(not_found)/count)\n",
    "    return my_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "en_counts = get_counts_and_average(en_cnv)\n",
    "gbm_counts = get_counts_and_average(gbm_cnv)\n",
    "brca_counts = get_counts_and_average(brca_cnv)\n",
    "hnscc_counts = get_counts_and_average(hnscc_cnv)\n",
    "luad_counts = get_counts_and_average(luad_cnv)\n",
    "ovarian_counts = get_counts_and_average(ovarian_cnv)\n",
    "ccrcc_counts = get_counts_and_average(ccrcc_cnv)\n",
    "colon_counts = get_counts_and_average(colon_cnv)\n",
    "lscc_counts = get_counts_and_average(lscc_cnv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.09352389777952026\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gain</th>\n",
       "      <th>neutral</th>\n",
       "      <th>loss</th>\n",
       "      <th>chromo</th>\n",
       "      <th>location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>4</td>\n",
       "      <td>85</td>\n",
       "      <td>6</td>\n",
       "      <td>19</td>\n",
       "      <td>58349335.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>4</td>\n",
       "      <td>85</td>\n",
       "      <td>6</td>\n",
       "      <td>19</td>\n",
       "      <td>58351586.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>17</td>\n",
       "      <td>75</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>50842542.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>7</td>\n",
       "      <td>87</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>9091946.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M-AS1</th>\n",
       "      <td>7</td>\n",
       "      <td>87</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>9066926.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>3</td>\n",
       "      <td>90</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>52868754.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>3</td>\n",
       "      <td>90</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>52776894.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>12</td>\n",
       "      <td>81</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>143386203.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>3</td>\n",
       "      <td>82</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "      <td>4073737.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>4</td>\n",
       "      <td>89</td>\n",
       "      <td>2</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28057 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gain  neutral  loss chromo     location\n",
       "A1BG         4       85     6     19   58349335.0\n",
       "A1BG-AS1     4       85     6     19   58351586.5\n",
       "A1CF        17       75     3     10   50842542.0\n",
       "A2M          7       87     1     12    9091946.5\n",
       "A2M-AS1      7       87     1     12    9066926.0\n",
       "...        ...      ...   ...    ...          ...\n",
       "ZYG11A       3       90     2      1   52868754.5\n",
       "ZYG11B       3       90     2      1   52776894.5\n",
       "ZYX         12       81     2      7  143386203.0\n",
       "ZZEF1        3       82    10     17    4073737.5\n",
       "ZZZ3         4       89     2   None          NaN\n",
       "\n",
       "[28057 rows x 5 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "add_chromo_and_loc(en_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.09352389777952026\n",
      "0.016476616265635204\n",
      "0.09678372446395407\n",
      "0.019470286208123633\n",
      "0.037577204546634144\n",
      "0.05881324318022849\n",
      "0.03754213119004408\n",
      "0.1250645174097749\n",
      "0.1323070139986038\n"
     ]
    }
   ],
   "source": [
    "en_counts = add_chromo_and_loc(en_counts)\n",
    "gbm_counts = add_chromo_and_loc(gbm_counts)\n",
    "brca_counts = add_chromo_and_loc(brca_counts)\n",
    "hnscc_counts = add_chromo_and_loc(hnscc_counts)\n",
    "luad_counts = add_chromo_and_loc(luad_counts)\n",
    "ovarian_counts = add_chromo_and_loc(ovarian_counts)\n",
    "ccrcc_counts = add_chromo_and_loc(ccrcc_counts)\n",
    "colon_counts = add_chromo_and_loc(colon_counts)\n",
    "lscc_counts = add_chromo_and_loc(lscc_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "en_counts.reset_index(inplace=True)\n",
    "gbm_counts.reset_index(inplace=True)\n",
    "brca_counts.reset_index(inplace=True)\n",
    "hnscc_counts.reset_index(inplace=True)\n",
    "luad_counts.reset_index(inplace=True)\n",
    "ovarian_counts.reset_index(inplace=True)\n",
    "ccrcc_counts.reset_index(inplace=True)\n",
    "colon_counts.reset_index(inplace=True)\n",
    "lscc_counts.reset_index(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "en_counts.to_csv(\"data/endo_counts.csv\")\n",
    "gbm_counts.to_csv(\"data/gbm_counts.csv\")\n",
    "brca_counts.to_csv(\"data/brca_counts.csv\")\n",
    "hnscc_counts.to_csv(\"data/hnscc_counts.csv\")\n",
    "luad_counts.to_csv(\"data/luad_counts.csv\")\n",
    "ovarian_counts.to_csv(\"data/ovarian_counts.csv\")\n",
    "ccrcc_counts.to_csv(\"data/ccrcc_counts.csv\")\n",
    "colon_counts.to_csv(\"data/colon_counts.csv\")\n",
    "lscc_counts.to_csv(\"data/lscc_counts.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>level_0</th>\n",
       "      <th>level_1</th>\n",
       "      <th>gain</th>\n",
       "      <th>neutral</th>\n",
       "      <th>loss</th>\n",
       "      <th>chromo</th>\n",
       "      <th>location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7SK</td>\n",
       "      <td>ENSG00000232512.2</td>\n",
       "      <td>15</td>\n",
       "      <td>89</td>\n",
       "      <td>18</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7SK</td>\n",
       "      <td>ENSG00000249352.3</td>\n",
       "      <td>11</td>\n",
       "      <td>80</td>\n",
       "      <td>31</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7SK</td>\n",
       "      <td>ENSG00000254144.2</td>\n",
       "      <td>66</td>\n",
       "      <td>49</td>\n",
       "      <td>7</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7SK</td>\n",
       "      <td>ENSG00000260682.2</td>\n",
       "      <td>9</td>\n",
       "      <td>50</td>\n",
       "      <td>63</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7SK</td>\n",
       "      <td>ENSG00000271765.1</td>\n",
       "      <td>13</td>\n",
       "      <td>86</td>\n",
       "      <td>23</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23687</th>\n",
       "      <td>snoZ6</td>\n",
       "      <td>ENSG00000252200.1</td>\n",
       "      <td>12</td>\n",
       "      <td>93</td>\n",
       "      <td>17</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23688</th>\n",
       "      <td>snoZ6</td>\n",
       "      <td>ENSG00000253067.1</td>\n",
       "      <td>9</td>\n",
       "      <td>89</td>\n",
       "      <td>24</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23689</th>\n",
       "      <td>snoZ6</td>\n",
       "      <td>ENSG00000264452.1</td>\n",
       "      <td>21</td>\n",
       "      <td>86</td>\n",
       "      <td>15</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23690</th>\n",
       "      <td>snoZ6</td>\n",
       "      <td>ENSG00000266692.1</td>\n",
       "      <td>21</td>\n",
       "      <td>86</td>\n",
       "      <td>15</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23691</th>\n",
       "      <td>snosnR66</td>\n",
       "      <td>ENSG00000212397.1</td>\n",
       "      <td>5</td>\n",
       "      <td>69</td>\n",
       "      <td>48</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23692 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        level_0            level_1  gain  neutral  loss chromo  location\n",
       "0           7SK  ENSG00000232512.2    15       89    18   None       NaN\n",
       "1           7SK  ENSG00000249352.3    11       80    31   None       NaN\n",
       "2           7SK  ENSG00000254144.2    66       49     7   None       NaN\n",
       "3           7SK  ENSG00000260682.2     9       50    63   None       NaN\n",
       "4           7SK  ENSG00000271765.1    13       86    23   None       NaN\n",
       "...         ...                ...   ...      ...   ...    ...       ...\n",
       "23687     snoZ6  ENSG00000252200.1    12       93    17   None       NaN\n",
       "23688     snoZ6  ENSG00000253067.1     9       89    24   None       NaN\n",
       "23689     snoZ6  ENSG00000264452.1    21       86    15   None       NaN\n",
       "23690     snoZ6  ENSG00000266692.1    21       86    15   None       NaN\n",
       "23691  snosnR66  ENSG00000212397.1     5       69    48   None       NaN\n",
       "\n",
       "[23692 rows x 7 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "brca_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
